// llmz/dist/chunk-UOPTKXLE.js
import {
  isPlainObject_default,
  isString_default
} from "./chunk-5XC6ZMS5.js";

// src/jsx.ts
import { z } from "@bpinternal/zui";
function isAnyJsxComponent(component) {
  return typeof component === "object" && component !== null && "__jsx" in component && component.__jsx === true && typeof component.type === "string";
}
function isJsxComponent(type, component) {
  return typeof component === "object" && component !== null && "__jsx" in component && component.__jsx === true && typeof component.type === "string" && typeof type === "string" && component.type.toUpperCase().trim() === type.toUpperCase().trim();
}
var createJsxComponent = (props) => ({
  __jsx: true,
  type: isString_default(props.type) ? props.type.toUpperCase() : "__unknown__",
  children: Array.isArray(props.children) ? props.children : props.children ? [props.children] : [],
  props: isPlainObject_default(props.props) ? props.props : {}
});
var Jsx = z.custom(
  (value) => isAnyJsxComponent(value),
  (value) => ({
    params: { value },
    message: `Invalid JSX component: ${value}`
  })
);

export {
  isAnyJsxComponent,
  isJsxComponent,
  createJsxComponent,
  Jsx
};
