// llmz/dist/chunk-F4OZHUMS.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } var _class;



var _chunk4VYDQ3VFcjs = require('./chunk-4VYDQ3VF.cjs');


var _chunkRKKRSBX3cjs = require('./chunk-RKKRSBX3.cjs');

// src/exit.ts
var _zui = require('@bpinternal/zui');
var Exit = (_class = class _Exit {
  /** The primary name of the exit (used in return statements) */
  
  /** Alternative names that can be used to reference this exit */
  __init() {this.aliases = []}
  /** Human-readable description of when this exit should be used */
  
  /** Additional metadata for orchestration and custom logic */
  
  /** JSON Schema for validating exit result data */
  
  /**
   * Returns the Zod schema equivalent of the JSON schema (if available).
   * Used internally for validation and type inference.
   */
  get zSchema() {
    return this.schema ? _zui.transforms.fromJSONSchemaLegacy(this.schema) : void 0;
  }
  /**
   * Renames the exit and updates aliases accordingly.
   *
   * @param name - The new name for the exit (must be a valid identifier)
   * @returns This exit instance for chaining
   *
   * @example
   * ```typescript
   * const exit = new Exit({ name: 'old_name', description: 'Test exit' })
   * exit.rename('new_name')
   * console.log(exit.name) // 'new_name'
   * ```
   */
  rename(name) {
    const before = this.name;
    if (!_chunk4VYDQ3VFcjs.isValidIdentifier.call(void 0, name)) {
      throw new Error(
        `Invalid name for exit ${name}. An exit name must start with a letter and contain only letters, numbers, and underscores. It must be 1-50 characters long.`
      );
    }
    this.name = name;
    this.aliases = _chunkRKKRSBX3cjs.uniq_default.call(void 0, [name, ...this.aliases.map((alias) => alias === before ? name : alias)]);
    return this;
  }
  /**
   * Creates a deep copy of this exit.
   *
   * The clone is completely independent and can be modified without affecting
   * the original exit. This is useful for creating variations of existing exits.
   *
   * @returns A new Exit instance with the same configuration
   *
   * @example
   * ```typescript
   * const originalExit = new Exit({
   *   name: 'base',
   *   description: 'Base exit',
   *   schema: z.object({ status: z.string() }),
   * })
   *
   * const customExit = originalExit.clone().rename('custom')
   * // customExit is independent of originalExit
   * ```
   */
  clone() {
    return new _Exit({
      name: this.name,
      aliases: [...this.aliases],
      description: this.description,
      metadata: JSON.parse(JSON.stringify(this.metadata)),
      schema: this.zSchema
    });
  }
  /**
   * Type guard to check if this exit matches another exit by name.
   *
   * Used internally for type narrowing and exit comparison.
   *
   * @param exit - The exit to compare against
   * @returns True if the exits have the same name
   */
  is(exit) {
    return this.name === exit.name;
  }
  /**
   * Type guard to check if an ExitResult matches this exit.
   *
   * @param result - The exit result to check
   * @returns True if the result was created by this exit
   */
  match(result) {
    return result.exit instanceof _Exit && this.name === result.exit.name;
  }
  /**
   * Serializes this exit to a JSON-compatible object.
   *
   * @returns JSON representation of the exit
   *
   * @example
   * ```typescript
   * const exit = new Exit({
   *   name: 'complete',
   *   description: 'Task completed successfully',
   * })
   *
   * console.log(exit.toJSON())
   * // { name: 'complete', aliases: [], description: 'Task completed successfully', metadata: {}, schema: undefined }
   * ```
   */
  toJSON() {
    return {
      name: this.name,
      aliases: [...this.aliases],
      description: this.description,
      metadata: { ...this.metadata },
      schema: this.schema
    };
  }
  /**
   * Creates a new Exit instance.
   *
   * @param props - Exit configuration
   * @param props.name - Primary name for the exit (must be valid identifier)
   * @param props.description - Human-readable description of the exit's purpose
   * @param props.aliases - Alternative names that can be used to reference this exit
   * @param props.metadata - Additional data for orchestration and custom logic
   * @param props.schema - Zod schema for validating exit result data
   *
   * @example
   * ```typescript
   * // Simple exit without data validation
   * const exit = new Exit({
   *   name: 'complete',
   *   description: 'Task completed successfully',
   * })
   * ```
   *
   * @example
   * ```typescript
   * // Exit with typed result data
   * const approval = new Exit({
   *   name: 'approved',
   *   description: 'Request approved by system',
   *   schema: z.object({
   *     amount: z.number().positive(),
   *     reference: z.string().min(1),
   *     timestamp: z.date().default(() => new Date()),
   *   }),
   * })
   * ```
   *
   * @example
   * ```typescript
   * // Exit with aliases and metadata
   * const handoff = new Exit({
   *   name: 'handoff_support',
   *   aliases: ['escalate', 'transfer'],
   *   description: 'Transfer to human support agent',
   *   metadata: {
   *     department: 'customer_service',
   *     priority: 'high',
   *   },
   *   schema: z.object({
   *     reason: z.string(),
   *     customerData: z.record(z.any()),
   *   }),
   * })
   * ```
   */
  constructor(props) {;_class.prototype.__init.call(this);
    if (!_chunk4VYDQ3VFcjs.isValidIdentifier.call(void 0, props.name)) {
      throw new Error(
        `Invalid name for exit ${props.name}. A exit name must start with a letter and contain only letters, numbers, and underscores. It must be 1-50 characters long.`
      );
    }
    if (typeof props.description !== "string" || props.description.trim().length === 0) {
      throw new Error(
        `Invalid description for exit ${props.name}. Expected a non-empty string, but got type "${typeof props.description}"`
      );
    }
    if (props.metadata !== void 0 && typeof props.metadata !== "object") {
      throw new Error(
        `Invalid metadata for exit ${props.name}. Expected an object, but got type "${typeof props.metadata}"`
      );
    }
    if (props.aliases !== void 0 && !Array.isArray(props.aliases)) {
      throw new Error(
        `Invalid aliases for exit ${props.name}. Expected an array, but got type "${typeof props.aliases}"`
      );
    }
    if (props.aliases && props.aliases.some((alias) => !_chunk4VYDQ3VFcjs.isValidIdentifier.call(void 0, alias))) {
      throw new Error(`Invalid aliases for exit ${props.name}. Expected an array of valid identifiers.`);
    }
    if (typeof props.schema !== "undefined") {
      if (_chunk4VYDQ3VFcjs.isZuiSchema.call(void 0, props.schema)) {
        this.schema = _zui.transforms.toJSONSchemaLegacy(props.schema);
      } else if (_chunk4VYDQ3VFcjs.isJsonSchema.call(void 0, props.schema)) {
        this.schema = props.schema;
      } else {
        throw new Error(
          `Invalid input schema for exit ${props.name}. Expected a ZodType or JSONSchema, but got type "${typeof props.schema}"`
        );
      }
    }
    this.name = props.name;
    this.aliases = _chunkRKKRSBX3cjs.uniq_default.call(void 0, [props.name, ..._nullishCoalesce(props.aliases, () => ( []))]);
    this.description = props.description;
    this.metadata = _nullishCoalesce(props.metadata, () => ( {}));
  }
  /**
   * Ensures all exits in an array have unique names by renaming duplicates.
   *
   * When multiple exits have the same name, this method appends numbers to
   * create unique names (e.g., 'exit1', 'exit2'). This prevents naming conflicts
   * in execution contexts with multiple exits.
   *
   * @param exits - Array of exits that may have duplicate names
   * @returns Array of exits with guaranteed unique names
   *
   * @example
   * ```typescript
   * const exit1 = new Exit({ name: 'done', description: 'First done' })
   * const exit2 = new Exit({ name: 'done', description: 'Second done' })
   *
   * const uniqueExits = Exit.withUniqueNames([exit1, exit2])
   * // Result: [{ name: 'done' }, { name: 'done1' }]
   * ```
   */
  static __initStatic() {this.withUniqueNames = (exits) => {
    const names = /* @__PURE__ */ new Set();
    return exits.map((exit) => {
      if (exits.filter((t) => t.name === exit.name).length === 1) {
        return exit;
      }
      let counter = 1;
      let exitName = exit.name + counter;
      while (names.has(exitName)) {
        exitName = `${exit.name}${++counter}`;
      }
      return exit.rename(exitName);
    });
  }}
}, _class.__initStatic(), _class);



exports.Exit = Exit;
