// llmz/dist/chunk-K2B5YEGJ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _chunk54G63644cjs = require('./chunk-54G63644.cjs');

// src/component.ts
var _zui = require('@bpinternal/zui');
function assertValidComponent(component) {
  if (!component.name) {
    throw new Error("Component must have a name");
  }
  const nameRegex = /^[A-Z]{3,50}$/i;
  if (!nameRegex.test(component.name)) {
    throw new Error(
      `Component name "${component.name}" must be 3-50 characters long and start with an uppercase letter`
    );
  }
  if (component.aliases) {
    for (const alias of component.aliases) {
      if (!nameRegex.test(alias)) {
        throw new Error(`Alias "${alias}" must be 3-50 characters long and start with an uppercase letter`);
      }
    }
  }
  if (!component.description) {
    throw new Error("Component must have a description");
  }
  if (!component.examples || component.examples.length === 0) {
    throw new Error("Component must have at least one example");
  }
  if (component.type === "default" && !component.default) {
    throw new Error("Default component must have default props and children");
  }
  if (component.type === "leaf" && !component.leaf) {
    throw new Error("Leaf component must have leaf props");
  }
  if (component.type === "container" && !component.container) {
    throw new Error("Container component must have container props and children");
  }
}
var getDefaultValue = (schema) => {
  if (schema._def.defaultValue !== void 0) {
    if (typeof schema._def.defaultValue === "function") {
      return String(schema._def.defaultValue()).toString();
    } else {
      return String(schema._def.defaultValue);
    }
  }
  return "";
};
function getComponentReference(component) {
  let doc = `### <${component.name}>

`;
  doc += `${component.description}

`;
  const getPropsDoc = (props) => {
    const shape = props.shape;
    if (Object.keys(shape).length === 0)
      return "_No props._\n\n";
    const zodTypeToTsType = {
      ZodString: "string",
      ZodNumber: "number",
      ZodBoolean: "boolean",
      ZodEnum: "enum",
      ZodArray: "array",
      ZodObject: "object",
      ZodDate: "date",
      ZodBigInt: "bigint",
      ZodSymbol: "symbol",
      ZodUndefined: "undefined",
      ZodNull: "null",
      ZodVoid: "void",
      ZodNever: "never",
      ZodUnknown: "unknown",
      ZodAny: "any"
    };
    return Object.entries(shape).map(([name, schema]) => {
      const naked = schema.naked();
      const zodType = naked._def.typeName;
      const defValue = getDefaultValue(schema);
      const typings = _zui.z.is.zuiEnum(naked) ? naked._def.values.map((x) => `"${x}"`).join(" | ") : zodTypeToTsType[zodType] || zodType;
      const required = !schema.isOptional() ? "**(required)**" : "(optional)";
      const def = defValue ? ` _Default: \`${defValue}\`_` : "";
      const description = schema.description || schema.naked().description || (schema == null ? void 0 : schema._def.description) || "";
      return `- \`${name}: ${typings}\` ${required} \u2014 ${description}${def}`;
    }).join("\n") + "\n\n";
  };
  const getChildrenDoc = (children) => {
    if (children.length === 0)
      return "_None allowed._\n\n";
    return "Can contain:\n" + children.map((child) => `- ${child.description} \u2014 \`<${child.component.name}>\``).join("\n") + "\n\n";
  };
  const getExamplesDoc = (examples) => {
    if (!examples.length)
      return "";
    return "**Examples:**\n\n" + examples.map((example) => `**${example.name}** \u2014 ${example.description}

\`\`\`tsx
${example.code.trim()}
\`\`\`
`).join("\n");
  };
  switch (component.type) {
    case "leaf":
      doc += "**Props:**\n\n";
      doc += getPropsDoc(component.leaf.props);
      doc += "**Children:**\n\n";
      doc += getChildrenDoc([]);
      doc += getExamplesDoc(component.examples);
      break;
    case "container":
      doc += "**Props:**\n\n";
      doc += getPropsDoc(component.container.props);
      doc += "**Children:**\n\n";
      doc += getChildrenDoc(component.container.children);
      doc += getExamplesDoc(component.examples);
      break;
    case "default":
    default:
      doc += "**Props:**\n\n";
      doc += getPropsDoc(component.default.props);
      doc += getExamplesDoc(component.examples);
      break;
  }
  return doc.trim();
}
var Component = class {
  
  
  // phantom type for inference
  constructor(definition) {
    assertValidComponent(definition);
    this.definition = definition;
  }
};
function isComponent(rendered, component) {
  return _chunk54G63644cjs.isJsxComponent.call(void 0, component.definition.name, rendered);
}
function isAnyComponent(message) {
  return _chunk54G63644cjs.isAnyJsxComponent.call(void 0, message);
}
function renderToTsx(component) {
  const props = Object.entries(component.props).map(([key, value]) => {
    if (typeof value === "string") {
      return `${key}="${value}"`;
    }
    if (typeof value === "boolean") {
      return value ? key : "";
    }
    if (typeof value === "number") {
      return `${key}={${value}}`;
    }
    if (value === null || value === void 0) {
      return "";
    }
    if (typeof value === "object") {
      return `${key}={${JSON.stringify(value)}}`;
    }
    return `${key}={${String(value)}}`;
  }).filter(Boolean).join(" ");
  const children = component.children.map((child) => {
    if (typeof child === "string") {
      return child;
    }
    if (isAnyComponent(child)) {
      return renderToTsx(child);
    }
    return String(child);
  }).join("");
  return `<${component.type}${props ? " " + props : ""}>${children}</${component.type}>`;
}








exports.assertValidComponent = assertValidComponent; exports.getComponentReference = getComponentReference; exports.Component = Component; exports.isComponent = isComponent; exports.isAnyComponent = isAnyComponent; exports.renderToTsx = renderToTsx;
