// llmz/dist/chunk-W4HLZNF4.js
import {
  CodeExecutionError,
  InvalidCodeError,
  Signals,
  SnapshotSignal,
  VMSignal
} from "./chunk-WSVDMGMR.js";
import {
  cleanStackTrace
} from "./chunk-YEAWWJSJ.js";
import {
  createJsxComponent
} from "./chunk-UOPTKXLE.js";
import {
  __commonJS,
  __toESM,
  isFunction_default,
  mapValues_default,
  maxBy_default
} from "./chunk-5XC6ZMS5.js";

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/base64.js
var require_base64 = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/base64.js"(exports) {
    "use strict";
    var intToCharMap = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".split("");
    exports.encode = function(number) {
      if (0 <= number && number < intToCharMap.length) {
        return intToCharMap[number];
      }
      throw new TypeError("Must be between 0 and 63: " + number);
    };
    exports.decode = function(charCode) {
      var bigA = 65;
      var bigZ = 90;
      var littleA = 97;
      var littleZ = 122;
      var zero = 48;
      var nine = 57;
      var plus = 43;
      var slash = 47;
      var littleOffset = 26;
      var numberOffset = 52;
      if (bigA <= charCode && charCode <= bigZ) {
        return charCode - bigA;
      }
      if (littleA <= charCode && charCode <= littleZ) {
        return charCode - littleA + littleOffset;
      }
      if (zero <= charCode && charCode <= nine) {
        return charCode - zero + numberOffset;
      }
      if (charCode == plus) {
        return 62;
      }
      if (charCode == slash) {
        return 63;
      }
      return -1;
    };
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/base64-vlq.js
var require_base64_vlq = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/base64-vlq.js"(exports) {
    "use strict";
    var base64 = require_base64();
    var VLQ_BASE_SHIFT = 5;
    var VLQ_BASE = 1 << VLQ_BASE_SHIFT;
    var VLQ_BASE_MASK = VLQ_BASE - 1;
    var VLQ_CONTINUATION_BIT = VLQ_BASE;
    function toVLQSigned(aValue) {
      return aValue < 0 ? (-aValue << 1) + 1 : (aValue << 1) + 0;
    }
    function fromVLQSigned(aValue) {
      var isNegative = (aValue & 1) === 1;
      var shifted = aValue >> 1;
      return isNegative ? -shifted : shifted;
    }
    exports.encode = function base64VLQ_encode(aValue) {
      var encoded = "";
      var digit;
      var vlq = toVLQSigned(aValue);
      do {
        digit = vlq & VLQ_BASE_MASK;
        vlq >>>= VLQ_BASE_SHIFT;
        if (vlq > 0) {
          digit |= VLQ_CONTINUATION_BIT;
        }
        encoded += base64.encode(digit);
      } while (vlq > 0);
      return encoded;
    };
    exports.decode = function base64VLQ_decode(aStr, aIndex, aOutParam) {
      var strLen = aStr.length;
      var result = 0;
      var shift = 0;
      var continuation, digit;
      do {
        if (aIndex >= strLen) {
          throw new Error("Expected more digits in base 64 VLQ value.");
        }
        digit = base64.decode(aStr.charCodeAt(aIndex++));
        if (digit === -1) {
          throw new Error("Invalid base64 digit: " + aStr.charAt(aIndex - 1));
        }
        continuation = !!(digit & VLQ_CONTINUATION_BIT);
        digit &= VLQ_BASE_MASK;
        result = result + (digit << shift);
        shift += VLQ_BASE_SHIFT;
      } while (continuation);
      aOutParam.value = fromVLQSigned(result);
      aOutParam.rest = aIndex;
    };
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/util.js
var require_util = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/util.js"(exports) {
    "use strict";
    function getArg(aArgs, aName, aDefaultValue) {
      if (aName in aArgs) {
        return aArgs[aName];
      } else if (arguments.length === 3) {
        return aDefaultValue;
      } else {
        throw new Error('"' + aName + '" is a required argument.');
      }
    }
    exports.getArg = getArg;
    var urlRegexp = /^(?:([\w+\-.]+):)?\/\/(?:(\w+:\w+)@)?([\w.-]*)(?::(\d+))?(.*)$/;
    var dataUrlRegexp = /^data:.+\,.+$/;
    function urlParse(aUrl) {
      var match = aUrl.match(urlRegexp);
      if (!match) {
        return null;
      }
      return {
        scheme: match[1],
        auth: match[2],
        host: match[3],
        port: match[4],
        path: match[5]
      };
    }
    exports.urlParse = urlParse;
    function urlGenerate(aParsedUrl) {
      var url = "";
      if (aParsedUrl.scheme) {
        url += aParsedUrl.scheme + ":";
      }
      url += "//";
      if (aParsedUrl.auth) {
        url += aParsedUrl.auth + "@";
      }
      if (aParsedUrl.host) {
        url += aParsedUrl.host;
      }
      if (aParsedUrl.port) {
        url += ":" + aParsedUrl.port;
      }
      if (aParsedUrl.path) {
        url += aParsedUrl.path;
      }
      return url;
    }
    exports.urlGenerate = urlGenerate;
    var MAX_CACHED_INPUTS = 32;
    function lruMemoize(f) {
      var cache = [];
      return function(input) {
        for (var i = 0; i < cache.length; i++) {
          if (cache[i].input === input) {
            var temp = cache[0];
            cache[0] = cache[i];
            cache[i] = temp;
            return cache[0].result;
          }
        }
        var result = f(input);
        cache.unshift({
          input,
          result
        });
        if (cache.length > MAX_CACHED_INPUTS) {
          cache.pop();
        }
        return result;
      };
    }
    var normalize = lruMemoize(function normalize2(aPath) {
      var path = aPath;
      var url = urlParse(aPath);
      if (url) {
        if (!url.path) {
          return aPath;
        }
        path = url.path;
      }
      var isAbsolute = exports.isAbsolute(path);
      var parts = [];
      var start = 0;
      var i = 0;
      while (true) {
        start = i;
        i = path.indexOf("/", start);
        if (i === -1) {
          parts.push(path.slice(start));
          break;
        } else {
          parts.push(path.slice(start, i));
          while (i < path.length && path[i] === "/") {
            i++;
          }
        }
      }
      for (var part, up = 0, i = parts.length - 1; i >= 0; i--) {
        part = parts[i];
        if (part === ".") {
          parts.splice(i, 1);
        } else if (part === "..") {
          up++;
        } else if (up > 0) {
          if (part === "") {
            parts.splice(i + 1, up);
            up = 0;
          } else {
            parts.splice(i, 2);
            up--;
          }
        }
      }
      path = parts.join("/");
      if (path === "") {
        path = isAbsolute ? "/" : ".";
      }
      if (url) {
        url.path = path;
        return urlGenerate(url);
      }
      return path;
    });
    exports.normalize = normalize;
    function join(aRoot, aPath) {
      if (aRoot === "") {
        aRoot = ".";
      }
      if (aPath === "") {
        aPath = ".";
      }
      var aPathUrl = urlParse(aPath);
      var aRootUrl = urlParse(aRoot);
      if (aRootUrl) {
        aRoot = aRootUrl.path || "/";
      }
      if (aPathUrl && !aPathUrl.scheme) {
        if (aRootUrl) {
          aPathUrl.scheme = aRootUrl.scheme;
        }
        return urlGenerate(aPathUrl);
      }
      if (aPathUrl || aPath.match(dataUrlRegexp)) {
        return aPath;
      }
      if (aRootUrl && !aRootUrl.host && !aRootUrl.path) {
        aRootUrl.host = aPath;
        return urlGenerate(aRootUrl);
      }
      var joined = aPath.charAt(0) === "/" ? aPath : normalize(aRoot.replace(/\/+$/, "") + "/" + aPath);
      if (aRootUrl) {
        aRootUrl.path = joined;
        return urlGenerate(aRootUrl);
      }
      return joined;
    }
    exports.join = join;
    exports.isAbsolute = function(aPath) {
      return aPath.charAt(0) === "/" || urlRegexp.test(aPath);
    };
    function relative(aRoot, aPath) {
      if (aRoot === "") {
        aRoot = ".";
      }
      aRoot = aRoot.replace(/\/$/, "");
      var level = 0;
      while (aPath.indexOf(aRoot + "/") !== 0) {
        var index = aRoot.lastIndexOf("/");
        if (index < 0) {
          return aPath;
        }
        aRoot = aRoot.slice(0, index);
        if (aRoot.match(/^([^\/]+:\/)?\/*$/)) {
          return aPath;
        }
        ++level;
      }
      return Array(level + 1).join("../") + aPath.substr(aRoot.length + 1);
    }
    exports.relative = relative;
    var supportsNullProto = function() {
      var obj = /* @__PURE__ */ Object.create(null);
      return !("__proto__" in obj);
    }();
    function identity(s) {
      return s;
    }
    function toSetString(aStr) {
      if (isProtoString(aStr)) {
        return "$" + aStr;
      }
      return aStr;
    }
    exports.toSetString = supportsNullProto ? identity : toSetString;
    function fromSetString(aStr) {
      if (isProtoString(aStr)) {
        return aStr.slice(1);
      }
      return aStr;
    }
    exports.fromSetString = supportsNullProto ? identity : fromSetString;
    function isProtoString(s) {
      if (!s) {
        return false;
      }
      var length = s.length;
      if (length < 9) {
        return false;
      }
      if (s.charCodeAt(length - 1) !== 95 || s.charCodeAt(length - 2) !== 95 || s.charCodeAt(length - 3) !== 111 || s.charCodeAt(length - 4) !== 116 || s.charCodeAt(length - 5) !== 111 || s.charCodeAt(length - 6) !== 114 || s.charCodeAt(length - 7) !== 112 || s.charCodeAt(length - 8) !== 95 || s.charCodeAt(length - 9) !== 95) {
        return false;
      }
      for (var i = length - 10; i >= 0; i--) {
        if (s.charCodeAt(i) !== 36) {
          return false;
        }
      }
      return true;
    }
    function compareByOriginalPositions(mappingA, mappingB, onlyCompareOriginal) {
      var cmp = strcmp(mappingA.source, mappingB.source);
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalLine - mappingB.originalLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalColumn - mappingB.originalColumn;
      if (cmp !== 0 || onlyCompareOriginal) {
        return cmp;
      }
      cmp = mappingA.generatedColumn - mappingB.generatedColumn;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.generatedLine - mappingB.generatedLine;
      if (cmp !== 0) {
        return cmp;
      }
      return strcmp(mappingA.name, mappingB.name);
    }
    exports.compareByOriginalPositions = compareByOriginalPositions;
    function compareByOriginalPositionsNoSource(mappingA, mappingB, onlyCompareOriginal) {
      var cmp;
      cmp = mappingA.originalLine - mappingB.originalLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalColumn - mappingB.originalColumn;
      if (cmp !== 0 || onlyCompareOriginal) {
        return cmp;
      }
      cmp = mappingA.generatedColumn - mappingB.generatedColumn;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.generatedLine - mappingB.generatedLine;
      if (cmp !== 0) {
        return cmp;
      }
      return strcmp(mappingA.name, mappingB.name);
    }
    exports.compareByOriginalPositionsNoSource = compareByOriginalPositionsNoSource;
    function compareByGeneratedPositionsDeflated(mappingA, mappingB, onlyCompareGenerated) {
      var cmp = mappingA.generatedLine - mappingB.generatedLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.generatedColumn - mappingB.generatedColumn;
      if (cmp !== 0 || onlyCompareGenerated) {
        return cmp;
      }
      cmp = strcmp(mappingA.source, mappingB.source);
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalLine - mappingB.originalLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalColumn - mappingB.originalColumn;
      if (cmp !== 0) {
        return cmp;
      }
      return strcmp(mappingA.name, mappingB.name);
    }
    exports.compareByGeneratedPositionsDeflated = compareByGeneratedPositionsDeflated;
    function compareByGeneratedPositionsDeflatedNoLine(mappingA, mappingB, onlyCompareGenerated) {
      var cmp = mappingA.generatedColumn - mappingB.generatedColumn;
      if (cmp !== 0 || onlyCompareGenerated) {
        return cmp;
      }
      cmp = strcmp(mappingA.source, mappingB.source);
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalLine - mappingB.originalLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalColumn - mappingB.originalColumn;
      if (cmp !== 0) {
        return cmp;
      }
      return strcmp(mappingA.name, mappingB.name);
    }
    exports.compareByGeneratedPositionsDeflatedNoLine = compareByGeneratedPositionsDeflatedNoLine;
    function strcmp(aStr1, aStr2) {
      if (aStr1 === aStr2) {
        return 0;
      }
      if (aStr1 === null) {
        return 1;
      }
      if (aStr2 === null) {
        return -1;
      }
      if (aStr1 > aStr2) {
        return 1;
      }
      return -1;
    }
    function compareByGeneratedPositionsInflated(mappingA, mappingB) {
      var cmp = mappingA.generatedLine - mappingB.generatedLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.generatedColumn - mappingB.generatedColumn;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = strcmp(mappingA.source, mappingB.source);
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalLine - mappingB.originalLine;
      if (cmp !== 0) {
        return cmp;
      }
      cmp = mappingA.originalColumn - mappingB.originalColumn;
      if (cmp !== 0) {
        return cmp;
      }
      return strcmp(mappingA.name, mappingB.name);
    }
    exports.compareByGeneratedPositionsInflated = compareByGeneratedPositionsInflated;
    function parseSourceMapInput(str) {
      return JSON.parse(str.replace(/^\)]}'[^\n]*\n/, ""));
    }
    exports.parseSourceMapInput = parseSourceMapInput;
    function computeSourceURL(sourceRoot, sourceURL, sourceMapURL) {
      sourceURL = sourceURL || "";
      if (sourceRoot) {
        if (sourceRoot[sourceRoot.length - 1] !== "/" && sourceURL[0] !== "/") {
          sourceRoot += "/";
        }
        sourceURL = sourceRoot + sourceURL;
      }
      if (sourceMapURL) {
        var parsed = urlParse(sourceMapURL);
        if (!parsed) {
          throw new Error("sourceMapURL could not be parsed");
        }
        if (parsed.path) {
          var index = parsed.path.lastIndexOf("/");
          if (index >= 0) {
            parsed.path = parsed.path.substring(0, index + 1);
          }
        }
        sourceURL = join(urlGenerate(parsed), sourceURL);
      }
      return normalize(sourceURL);
    }
    exports.computeSourceURL = computeSourceURL;
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/array-set.js
var require_array_set = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/array-set.js"(exports) {
    "use strict";
    var util = require_util();
    var has = Object.prototype.hasOwnProperty;
    var hasNativeMap = typeof Map !== "undefined";
    function ArraySet() {
      this._array = [];
      this._set = hasNativeMap ? /* @__PURE__ */ new Map() : /* @__PURE__ */ Object.create(null);
    }
    ArraySet.fromArray = function ArraySet_fromArray(aArray, aAllowDuplicates) {
      var set = new ArraySet();
      for (var i = 0, len = aArray.length; i < len; i++) {
        set.add(aArray[i], aAllowDuplicates);
      }
      return set;
    };
    ArraySet.prototype.size = function ArraySet_size() {
      return hasNativeMap ? this._set.size : Object.getOwnPropertyNames(this._set).length;
    };
    ArraySet.prototype.add = function ArraySet_add(aStr, aAllowDuplicates) {
      var sStr = hasNativeMap ? aStr : util.toSetString(aStr);
      var isDuplicate = hasNativeMap ? this.has(aStr) : has.call(this._set, sStr);
      var idx = this._array.length;
      if (!isDuplicate || aAllowDuplicates) {
        this._array.push(aStr);
      }
      if (!isDuplicate) {
        if (hasNativeMap) {
          this._set.set(aStr, idx);
        } else {
          this._set[sStr] = idx;
        }
      }
    };
    ArraySet.prototype.has = function ArraySet_has(aStr) {
      if (hasNativeMap) {
        return this._set.has(aStr);
      } else {
        var sStr = util.toSetString(aStr);
        return has.call(this._set, sStr);
      }
    };
    ArraySet.prototype.indexOf = function ArraySet_indexOf(aStr) {
      if (hasNativeMap) {
        var idx = this._set.get(aStr);
        if (idx >= 0) {
          return idx;
        }
      } else {
        var sStr = util.toSetString(aStr);
        if (has.call(this._set, sStr)) {
          return this._set[sStr];
        }
      }
      throw new Error('"' + aStr + '" is not in the set.');
    };
    ArraySet.prototype.at = function ArraySet_at(aIdx) {
      if (aIdx >= 0 && aIdx < this._array.length) {
        return this._array[aIdx];
      }
      throw new Error("No element indexed by " + aIdx);
    };
    ArraySet.prototype.toArray = function ArraySet_toArray() {
      return this._array.slice();
    };
    exports.ArraySet = ArraySet;
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/mapping-list.js
var require_mapping_list = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/mapping-list.js"(exports) {
    "use strict";
    var util = require_util();
    function generatedPositionAfter(mappingA, mappingB) {
      var lineA = mappingA.generatedLine;
      var lineB = mappingB.generatedLine;
      var columnA = mappingA.generatedColumn;
      var columnB = mappingB.generatedColumn;
      return lineB > lineA || lineB == lineA && columnB >= columnA || util.compareByGeneratedPositionsInflated(mappingA, mappingB) <= 0;
    }
    function MappingList() {
      this._array = [];
      this._sorted = true;
      this._last = { generatedLine: -1, generatedColumn: 0 };
    }
    MappingList.prototype.unsortedForEach = function MappingList_forEach(aCallback, aThisArg) {
      this._array.forEach(aCallback, aThisArg);
    };
    MappingList.prototype.add = function MappingList_add(aMapping) {
      if (generatedPositionAfter(this._last, aMapping)) {
        this._last = aMapping;
        this._array.push(aMapping);
      } else {
        this._sorted = false;
        this._array.push(aMapping);
      }
    };
    MappingList.prototype.toArray = function MappingList_toArray() {
      if (!this._sorted) {
        this._array.sort(util.compareByGeneratedPositionsInflated);
        this._sorted = true;
      }
      return this._array;
    };
    exports.MappingList = MappingList;
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/source-map-generator.js
var require_source_map_generator = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/source-map-generator.js"(exports) {
    "use strict";
    var base64VLQ = require_base64_vlq();
    var util = require_util();
    var ArraySet = require_array_set().ArraySet;
    var MappingList = require_mapping_list().MappingList;
    function SourceMapGenerator(aArgs) {
      if (!aArgs) {
        aArgs = {};
      }
      this._file = util.getArg(aArgs, "file", null);
      this._sourceRoot = util.getArg(aArgs, "sourceRoot", null);
      this._skipValidation = util.getArg(aArgs, "skipValidation", false);
      this._ignoreInvalidMapping = util.getArg(aArgs, "ignoreInvalidMapping", false);
      this._sources = new ArraySet();
      this._names = new ArraySet();
      this._mappings = new MappingList();
      this._sourcesContents = null;
    }
    SourceMapGenerator.prototype._version = 3;
    SourceMapGenerator.fromSourceMap = function SourceMapGenerator_fromSourceMap(aSourceMapConsumer, generatorOps) {
      var sourceRoot = aSourceMapConsumer.sourceRoot;
      var generator = new SourceMapGenerator(Object.assign(generatorOps || {}, {
        file: aSourceMapConsumer.file,
        sourceRoot
      }));
      aSourceMapConsumer.eachMapping(function(mapping) {
        var newMapping = {
          generated: {
            line: mapping.generatedLine,
            column: mapping.generatedColumn
          }
        };
        if (mapping.source != null) {
          newMapping.source = mapping.source;
          if (sourceRoot != null) {
            newMapping.source = util.relative(sourceRoot, newMapping.source);
          }
          newMapping.original = {
            line: mapping.originalLine,
            column: mapping.originalColumn
          };
          if (mapping.name != null) {
            newMapping.name = mapping.name;
          }
        }
        generator.addMapping(newMapping);
      });
      aSourceMapConsumer.sources.forEach(function(sourceFile) {
        var sourceRelative = sourceFile;
        if (sourceRoot !== null) {
          sourceRelative = util.relative(sourceRoot, sourceFile);
        }
        if (!generator._sources.has(sourceRelative)) {
          generator._sources.add(sourceRelative);
        }
        var content = aSourceMapConsumer.sourceContentFor(sourceFile);
        if (content != null) {
          generator.setSourceContent(sourceFile, content);
        }
      });
      return generator;
    };
    SourceMapGenerator.prototype.addMapping = function SourceMapGenerator_addMapping(aArgs) {
      var generated = util.getArg(aArgs, "generated");
      var original = util.getArg(aArgs, "original", null);
      var source = util.getArg(aArgs, "source", null);
      var name = util.getArg(aArgs, "name", null);
      if (!this._skipValidation) {
        if (this._validateMapping(generated, original, source, name) === false) {
          return;
        }
      }
      if (source != null) {
        source = String(source);
        if (!this._sources.has(source)) {
          this._sources.add(source);
        }
      }
      if (name != null) {
        name = String(name);
        if (!this._names.has(name)) {
          this._names.add(name);
        }
      }
      this._mappings.add({
        generatedLine: generated.line,
        generatedColumn: generated.column,
        originalLine: original != null && original.line,
        originalColumn: original != null && original.column,
        source,
        name
      });
    };
    SourceMapGenerator.prototype.setSourceContent = function SourceMapGenerator_setSourceContent(aSourceFile, aSourceContent) {
      var source = aSourceFile;
      if (this._sourceRoot != null) {
        source = util.relative(this._sourceRoot, source);
      }
      if (aSourceContent != null) {
        if (!this._sourcesContents) {
          this._sourcesContents = /* @__PURE__ */ Object.create(null);
        }
        this._sourcesContents[util.toSetString(source)] = aSourceContent;
      } else if (this._sourcesContents) {
        delete this._sourcesContents[util.toSetString(source)];
        if (Object.keys(this._sourcesContents).length === 0) {
          this._sourcesContents = null;
        }
      }
    };
    SourceMapGenerator.prototype.applySourceMap = function SourceMapGenerator_applySourceMap(aSourceMapConsumer, aSourceFile, aSourceMapPath) {
      var sourceFile = aSourceFile;
      if (aSourceFile == null) {
        if (aSourceMapConsumer.file == null) {
          throw new Error(
            `SourceMapGenerator.prototype.applySourceMap requires either an explicit source file, or the source map's "file" property. Both were omitted.`
          );
        }
        sourceFile = aSourceMapConsumer.file;
      }
      var sourceRoot = this._sourceRoot;
      if (sourceRoot != null) {
        sourceFile = util.relative(sourceRoot, sourceFile);
      }
      var newSources = new ArraySet();
      var newNames = new ArraySet();
      this._mappings.unsortedForEach(function(mapping) {
        if (mapping.source === sourceFile && mapping.originalLine != null) {
          var original = aSourceMapConsumer.originalPositionFor({
            line: mapping.originalLine,
            column: mapping.originalColumn
          });
          if (original.source != null) {
            mapping.source = original.source;
            if (aSourceMapPath != null) {
              mapping.source = util.join(aSourceMapPath, mapping.source);
            }
            if (sourceRoot != null) {
              mapping.source = util.relative(sourceRoot, mapping.source);
            }
            mapping.originalLine = original.line;
            mapping.originalColumn = original.column;
            if (original.name != null) {
              mapping.name = original.name;
            }
          }
        }
        var source = mapping.source;
        if (source != null && !newSources.has(source)) {
          newSources.add(source);
        }
        var name = mapping.name;
        if (name != null && !newNames.has(name)) {
          newNames.add(name);
        }
      }, this);
      this._sources = newSources;
      this._names = newNames;
      aSourceMapConsumer.sources.forEach(function(sourceFile2) {
        var content = aSourceMapConsumer.sourceContentFor(sourceFile2);
        if (content != null) {
          if (aSourceMapPath != null) {
            sourceFile2 = util.join(aSourceMapPath, sourceFile2);
          }
          if (sourceRoot != null) {
            sourceFile2 = util.relative(sourceRoot, sourceFile2);
          }
          this.setSourceContent(sourceFile2, content);
        }
      }, this);
    };
    SourceMapGenerator.prototype._validateMapping = function SourceMapGenerator_validateMapping(aGenerated, aOriginal, aSource, aName) {
      if (aOriginal && typeof aOriginal.line !== "number" && typeof aOriginal.column !== "number") {
        var message = "original.line and original.column are not numbers -- you probably meant to omit the original mapping entirely and only map the generated position. If so, pass null for the original mapping instead of an object with empty or null values.";
        if (this._ignoreInvalidMapping) {
          if (typeof console !== "undefined" && console.warn) {
            console.warn(message);
          }
          return false;
        } else {
          throw new Error(message);
        }
      }
      if (aGenerated && "line" in aGenerated && "column" in aGenerated && aGenerated.line > 0 && aGenerated.column >= 0 && !aOriginal && !aSource && !aName) {
        return;
      } else if (aGenerated && "line" in aGenerated && "column" in aGenerated && aOriginal && "line" in aOriginal && "column" in aOriginal && aGenerated.line > 0 && aGenerated.column >= 0 && aOriginal.line > 0 && aOriginal.column >= 0 && aSource) {
        return;
      } else {
        var message = "Invalid mapping: " + JSON.stringify({
          generated: aGenerated,
          source: aSource,
          original: aOriginal,
          name: aName
        });
        if (this._ignoreInvalidMapping) {
          if (typeof console !== "undefined" && console.warn) {
            console.warn(message);
          }
          return false;
        } else {
          throw new Error(message);
        }
      }
    };
    SourceMapGenerator.prototype._serializeMappings = function SourceMapGenerator_serializeMappings() {
      var previousGeneratedColumn = 0;
      var previousGeneratedLine = 1;
      var previousOriginalColumn = 0;
      var previousOriginalLine = 0;
      var previousName = 0;
      var previousSource = 0;
      var result = "";
      var next;
      var mapping;
      var nameIdx;
      var sourceIdx;
      var mappings = this._mappings.toArray();
      for (var i = 0, len = mappings.length; i < len; i++) {
        mapping = mappings[i];
        next = "";
        if (mapping.generatedLine !== previousGeneratedLine) {
          previousGeneratedColumn = 0;
          while (mapping.generatedLine !== previousGeneratedLine) {
            next += ";";
            previousGeneratedLine++;
          }
        } else {
          if (i > 0) {
            if (!util.compareByGeneratedPositionsInflated(mapping, mappings[i - 1])) {
              continue;
            }
            next += ",";
          }
        }
        next += base64VLQ.encode(mapping.generatedColumn - previousGeneratedColumn);
        previousGeneratedColumn = mapping.generatedColumn;
        if (mapping.source != null) {
          sourceIdx = this._sources.indexOf(mapping.source);
          next += base64VLQ.encode(sourceIdx - previousSource);
          previousSource = sourceIdx;
          next += base64VLQ.encode(mapping.originalLine - 1 - previousOriginalLine);
          previousOriginalLine = mapping.originalLine - 1;
          next += base64VLQ.encode(mapping.originalColumn - previousOriginalColumn);
          previousOriginalColumn = mapping.originalColumn;
          if (mapping.name != null) {
            nameIdx = this._names.indexOf(mapping.name);
            next += base64VLQ.encode(nameIdx - previousName);
            previousName = nameIdx;
          }
        }
        result += next;
      }
      return result;
    };
    SourceMapGenerator.prototype._generateSourcesContent = function SourceMapGenerator_generateSourcesContent(aSources, aSourceRoot) {
      return aSources.map(function(source) {
        if (!this._sourcesContents) {
          return null;
        }
        if (aSourceRoot != null) {
          source = util.relative(aSourceRoot, source);
        }
        var key = util.toSetString(source);
        return Object.prototype.hasOwnProperty.call(this._sourcesContents, key) ? this._sourcesContents[key] : null;
      }, this);
    };
    SourceMapGenerator.prototype.toJSON = function SourceMapGenerator_toJSON() {
      var map = {
        version: this._version,
        sources: this._sources.toArray(),
        names: this._names.toArray(),
        mappings: this._serializeMappings()
      };
      if (this._file != null) {
        map.file = this._file;
      }
      if (this._sourceRoot != null) {
        map.sourceRoot = this._sourceRoot;
      }
      if (this._sourcesContents) {
        map.sourcesContent = this._generateSourcesContent(map.sources, map.sourceRoot);
      }
      return map;
    };
    SourceMapGenerator.prototype.toString = function SourceMapGenerator_toString() {
      return JSON.stringify(this.toJSON());
    };
    exports.SourceMapGenerator = SourceMapGenerator;
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/binary-search.js
var require_binary_search = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/binary-search.js"(exports) {
    "use strict";
    exports.GREATEST_LOWER_BOUND = 1;
    exports.LEAST_UPPER_BOUND = 2;
    function recursiveSearch(aLow, aHigh, aNeedle, aHaystack, aCompare, aBias) {
      var mid = Math.floor((aHigh - aLow) / 2) + aLow;
      var cmp = aCompare(aNeedle, aHaystack[mid], true);
      if (cmp === 0) {
        return mid;
      } else if (cmp > 0) {
        if (aHigh - mid > 1) {
          return recursiveSearch(mid, aHigh, aNeedle, aHaystack, aCompare, aBias);
        }
        if (aBias == exports.LEAST_UPPER_BOUND) {
          return aHigh < aHaystack.length ? aHigh : -1;
        } else {
          return mid;
        }
      } else {
        if (mid - aLow > 1) {
          return recursiveSearch(aLow, mid, aNeedle, aHaystack, aCompare, aBias);
        }
        if (aBias == exports.LEAST_UPPER_BOUND) {
          return mid;
        } else {
          return aLow < 0 ? -1 : aLow;
        }
      }
    }
    exports.search = function search(aNeedle, aHaystack, aCompare, aBias) {
      if (aHaystack.length === 0) {
        return -1;
      }
      var index = recursiveSearch(
        -1,
        aHaystack.length,
        aNeedle,
        aHaystack,
        aCompare,
        aBias || exports.GREATEST_LOWER_BOUND
      );
      if (index < 0) {
        return -1;
      }
      while (index - 1 >= 0) {
        if (aCompare(aHaystack[index], aHaystack[index - 1], true) !== 0) {
          break;
        }
        --index;
      }
      return index;
    };
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/quick-sort.js
var require_quick_sort = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/quick-sort.js"(exports) {
    "use strict";
    function SortTemplate(comparator) {
      function swap(ary, x, y) {
        var temp = ary[x];
        ary[x] = ary[y];
        ary[y] = temp;
      }
      function randomIntInRange(low, high) {
        return Math.round(low + Math.random() * (high - low));
      }
      function doQuickSort(ary, comparator2, p, r) {
        if (p < r) {
          var pivotIndex = randomIntInRange(p, r);
          var i = p - 1;
          swap(ary, pivotIndex, r);
          var pivot = ary[r];
          for (var j = p; j < r; j++) {
            if (comparator2(ary[j], pivot, false) <= 0) {
              i += 1;
              swap(ary, i, j);
            }
          }
          swap(ary, i + 1, j);
          var q = i + 1;
          doQuickSort(ary, comparator2, p, q - 1);
          doQuickSort(ary, comparator2, q + 1, r);
        }
      }
      return doQuickSort;
    }
    function cloneSort(comparator) {
      let templateFn = (...args) => SortTemplate(...args);
      return templateFn(comparator);
    }
    var sortCache = /* @__PURE__ */ new WeakMap();
    exports.quickSort = function(ary, comparator, start = 0) {
      let doQuickSort = sortCache.get(comparator);
      if (doQuickSort === void 0) {
        doQuickSort = cloneSort(comparator);
        sortCache.set(comparator, doQuickSort);
      }
      doQuickSort(ary, comparator, start, ary.length - 1);
    };
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/source-map-consumer.js
var require_source_map_consumer = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/source-map-consumer.js"(exports) {
    "use strict";
    var util = require_util();
    var binarySearch = require_binary_search();
    var ArraySet = require_array_set().ArraySet;
    var base64VLQ = require_base64_vlq();
    var quickSort = require_quick_sort().quickSort;
    function SourceMapConsumer2(aSourceMap, aSourceMapURL) {
      var sourceMap = aSourceMap;
      if (typeof aSourceMap === "string") {
        sourceMap = util.parseSourceMapInput(aSourceMap);
      }
      return sourceMap.sections != null ? new IndexedSourceMapConsumer(sourceMap, aSourceMapURL) : new BasicSourceMapConsumer(sourceMap, aSourceMapURL);
    }
    SourceMapConsumer2.fromSourceMap = function(aSourceMap, aSourceMapURL) {
      return BasicSourceMapConsumer.fromSourceMap(aSourceMap, aSourceMapURL);
    };
    SourceMapConsumer2.prototype._version = 3;
    SourceMapConsumer2.prototype.__generatedMappings = null;
    Object.defineProperty(SourceMapConsumer2.prototype, "_generatedMappings", {
      configurable: true,
      enumerable: true,
      get: function() {
        if (!this.__generatedMappings) {
          this._parseMappings(this._mappings, this.sourceRoot);
        }
        return this.__generatedMappings;
      }
    });
    SourceMapConsumer2.prototype.__originalMappings = null;
    Object.defineProperty(SourceMapConsumer2.prototype, "_originalMappings", {
      configurable: true,
      enumerable: true,
      get: function() {
        if (!this.__originalMappings) {
          this._parseMappings(this._mappings, this.sourceRoot);
        }
        return this.__originalMappings;
      }
    });
    SourceMapConsumer2.prototype._charIsMappingSeparator = function SourceMapConsumer_charIsMappingSeparator(aStr, index) {
      var c = aStr.charAt(index);
      return c === ";" || c === ",";
    };
    SourceMapConsumer2.prototype._parseMappings = function SourceMapConsumer_parseMappings(aStr, aSourceRoot) {
      throw new Error("Subclasses must implement _parseMappings");
    };
    SourceMapConsumer2.GENERATED_ORDER = 1;
    SourceMapConsumer2.ORIGINAL_ORDER = 2;
    SourceMapConsumer2.GREATEST_LOWER_BOUND = 1;
    SourceMapConsumer2.LEAST_UPPER_BOUND = 2;
    SourceMapConsumer2.prototype.eachMapping = function SourceMapConsumer_eachMapping(aCallback, aContext, aOrder) {
      var context = aContext || null;
      var order = aOrder || SourceMapConsumer2.GENERATED_ORDER;
      var mappings;
      switch (order) {
        case SourceMapConsumer2.GENERATED_ORDER:
          mappings = this._generatedMappings;
          break;
        case SourceMapConsumer2.ORIGINAL_ORDER:
          mappings = this._originalMappings;
          break;
        default:
          throw new Error("Unknown order of iteration.");
      }
      var sourceRoot = this.sourceRoot;
      var boundCallback = aCallback.bind(context);
      var names = this._names;
      var sources = this._sources;
      var sourceMapURL = this._sourceMapURL;
      for (var i = 0, n = mappings.length; i < n; i++) {
        var mapping = mappings[i];
        var source = mapping.source === null ? null : sources.at(mapping.source);
        if (source !== null) {
          source = util.computeSourceURL(sourceRoot, source, sourceMapURL);
        }
        boundCallback({
          source,
          generatedLine: mapping.generatedLine,
          generatedColumn: mapping.generatedColumn,
          originalLine: mapping.originalLine,
          originalColumn: mapping.originalColumn,
          name: mapping.name === null ? null : names.at(mapping.name)
        });
      }
    };
    SourceMapConsumer2.prototype.allGeneratedPositionsFor = function SourceMapConsumer_allGeneratedPositionsFor(aArgs) {
      var line = util.getArg(aArgs, "line");
      var needle = {
        source: util.getArg(aArgs, "source"),
        originalLine: line,
        originalColumn: util.getArg(aArgs, "column", 0)
      };
      needle.source = this._findSourceIndex(needle.source);
      if (needle.source < 0) {
        return [];
      }
      var mappings = [];
      var index = this._findMapping(
        needle,
        this._originalMappings,
        "originalLine",
        "originalColumn",
        util.compareByOriginalPositions,
        binarySearch.LEAST_UPPER_BOUND
      );
      if (index >= 0) {
        var mapping = this._originalMappings[index];
        if (aArgs.column === void 0) {
          var originalLine = mapping.originalLine;
          while (mapping && mapping.originalLine === originalLine) {
            mappings.push({
              line: util.getArg(mapping, "generatedLine", null),
              column: util.getArg(mapping, "generatedColumn", null),
              lastColumn: util.getArg(mapping, "lastGeneratedColumn", null)
            });
            mapping = this._originalMappings[++index];
          }
        } else {
          var originalColumn = mapping.originalColumn;
          while (mapping && mapping.originalLine === line && mapping.originalColumn == originalColumn) {
            mappings.push({
              line: util.getArg(mapping, "generatedLine", null),
              column: util.getArg(mapping, "generatedColumn", null),
              lastColumn: util.getArg(mapping, "lastGeneratedColumn", null)
            });
            mapping = this._originalMappings[++index];
          }
        }
      }
      return mappings;
    };
    exports.SourceMapConsumer = SourceMapConsumer2;
    function BasicSourceMapConsumer(aSourceMap, aSourceMapURL) {
      var sourceMap = aSourceMap;
      if (typeof aSourceMap === "string") {
        sourceMap = util.parseSourceMapInput(aSourceMap);
      }
      var version = util.getArg(sourceMap, "version");
      var sources = util.getArg(sourceMap, "sources");
      var names = util.getArg(sourceMap, "names", []);
      var sourceRoot = util.getArg(sourceMap, "sourceRoot", null);
      var sourcesContent = util.getArg(sourceMap, "sourcesContent", null);
      var mappings = util.getArg(sourceMap, "mappings");
      var file = util.getArg(sourceMap, "file", null);
      if (version != this._version) {
        throw new Error("Unsupported version: " + version);
      }
      if (sourceRoot) {
        sourceRoot = util.normalize(sourceRoot);
      }
      sources = sources.map(String).map(util.normalize).map(function(source) {
        return sourceRoot && util.isAbsolute(sourceRoot) && util.isAbsolute(source) ? util.relative(sourceRoot, source) : source;
      });
      this._names = ArraySet.fromArray(names.map(String), true);
      this._sources = ArraySet.fromArray(sources, true);
      this._absoluteSources = this._sources.toArray().map(function(s) {
        return util.computeSourceURL(sourceRoot, s, aSourceMapURL);
      });
      this.sourceRoot = sourceRoot;
      this.sourcesContent = sourcesContent;
      this._mappings = mappings;
      this._sourceMapURL = aSourceMapURL;
      this.file = file;
    }
    BasicSourceMapConsumer.prototype = Object.create(SourceMapConsumer2.prototype);
    BasicSourceMapConsumer.prototype.consumer = SourceMapConsumer2;
    BasicSourceMapConsumer.prototype._findSourceIndex = function(aSource) {
      var relativeSource = aSource;
      if (this.sourceRoot != null) {
        relativeSource = util.relative(this.sourceRoot, relativeSource);
      }
      if (this._sources.has(relativeSource)) {
        return this._sources.indexOf(relativeSource);
      }
      var i;
      for (i = 0; i < this._absoluteSources.length; ++i) {
        if (this._absoluteSources[i] == aSource) {
          return i;
        }
      }
      return -1;
    };
    BasicSourceMapConsumer.fromSourceMap = function SourceMapConsumer_fromSourceMap(aSourceMap, aSourceMapURL) {
      var smc = Object.create(BasicSourceMapConsumer.prototype);
      var names = smc._names = ArraySet.fromArray(aSourceMap._names.toArray(), true);
      var sources = smc._sources = ArraySet.fromArray(aSourceMap._sources.toArray(), true);
      smc.sourceRoot = aSourceMap._sourceRoot;
      smc.sourcesContent = aSourceMap._generateSourcesContent(
        smc._sources.toArray(),
        smc.sourceRoot
      );
      smc.file = aSourceMap._file;
      smc._sourceMapURL = aSourceMapURL;
      smc._absoluteSources = smc._sources.toArray().map(function(s) {
        return util.computeSourceURL(smc.sourceRoot, s, aSourceMapURL);
      });
      var generatedMappings = aSourceMap._mappings.toArray().slice();
      var destGeneratedMappings = smc.__generatedMappings = [];
      var destOriginalMappings = smc.__originalMappings = [];
      for (var i = 0, length = generatedMappings.length; i < length; i++) {
        var srcMapping = generatedMappings[i];
        var destMapping = new Mapping();
        destMapping.generatedLine = srcMapping.generatedLine;
        destMapping.generatedColumn = srcMapping.generatedColumn;
        if (srcMapping.source) {
          destMapping.source = sources.indexOf(srcMapping.source);
          destMapping.originalLine = srcMapping.originalLine;
          destMapping.originalColumn = srcMapping.originalColumn;
          if (srcMapping.name) {
            destMapping.name = names.indexOf(srcMapping.name);
          }
          destOriginalMappings.push(destMapping);
        }
        destGeneratedMappings.push(destMapping);
      }
      quickSort(smc.__originalMappings, util.compareByOriginalPositions);
      return smc;
    };
    BasicSourceMapConsumer.prototype._version = 3;
    Object.defineProperty(BasicSourceMapConsumer.prototype, "sources", {
      get: function() {
        return this._absoluteSources.slice();
      }
    });
    function Mapping() {
      this.generatedLine = 0;
      this.generatedColumn = 0;
      this.source = null;
      this.originalLine = null;
      this.originalColumn = null;
      this.name = null;
    }
    var compareGenerated = util.compareByGeneratedPositionsDeflatedNoLine;
    function sortGenerated(array, start) {
      let l = array.length;
      let n = array.length - start;
      if (n <= 1) {
        return;
      } else if (n == 2) {
        let a = array[start];
        let b = array[start + 1];
        if (compareGenerated(a, b) > 0) {
          array[start] = b;
          array[start + 1] = a;
        }
      } else if (n < 20) {
        for (let i = start; i < l; i++) {
          for (let j = i; j > start; j--) {
            let a = array[j - 1];
            let b = array[j];
            if (compareGenerated(a, b) <= 0) {
              break;
            }
            array[j - 1] = b;
            array[j] = a;
          }
        }
      } else {
        quickSort(array, compareGenerated, start);
      }
    }
    BasicSourceMapConsumer.prototype._parseMappings = function SourceMapConsumer_parseMappings(aStr, aSourceRoot) {
      var generatedLine = 1;
      var previousGeneratedColumn = 0;
      var previousOriginalLine = 0;
      var previousOriginalColumn = 0;
      var previousSource = 0;
      var previousName = 0;
      var length = aStr.length;
      var index = 0;
      var cachedSegments = {};
      var temp = {};
      var originalMappings = [];
      var generatedMappings = [];
      var mapping, str, segment, end, value;
      let subarrayStart = 0;
      while (index < length) {
        if (aStr.charAt(index) === ";") {
          generatedLine++;
          index++;
          previousGeneratedColumn = 0;
          sortGenerated(generatedMappings, subarrayStart);
          subarrayStart = generatedMappings.length;
        } else if (aStr.charAt(index) === ",") {
          index++;
        } else {
          mapping = new Mapping();
          mapping.generatedLine = generatedLine;
          for (end = index; end < length; end++) {
            if (this._charIsMappingSeparator(aStr, end)) {
              break;
            }
          }
          str = aStr.slice(index, end);
          segment = [];
          while (index < end) {
            base64VLQ.decode(aStr, index, temp);
            value = temp.value;
            index = temp.rest;
            segment.push(value);
          }
          if (segment.length === 2) {
            throw new Error("Found a source, but no line and column");
          }
          if (segment.length === 3) {
            throw new Error("Found a source and line, but no column");
          }
          mapping.generatedColumn = previousGeneratedColumn + segment[0];
          previousGeneratedColumn = mapping.generatedColumn;
          if (segment.length > 1) {
            mapping.source = previousSource + segment[1];
            previousSource += segment[1];
            mapping.originalLine = previousOriginalLine + segment[2];
            previousOriginalLine = mapping.originalLine;
            mapping.originalLine += 1;
            mapping.originalColumn = previousOriginalColumn + segment[3];
            previousOriginalColumn = mapping.originalColumn;
            if (segment.length > 4) {
              mapping.name = previousName + segment[4];
              previousName += segment[4];
            }
          }
          generatedMappings.push(mapping);
          if (typeof mapping.originalLine === "number") {
            let currentSource = mapping.source;
            while (originalMappings.length <= currentSource) {
              originalMappings.push(null);
            }
            if (originalMappings[currentSource] === null) {
              originalMappings[currentSource] = [];
            }
            originalMappings[currentSource].push(mapping);
          }
        }
      }
      sortGenerated(generatedMappings, subarrayStart);
      this.__generatedMappings = generatedMappings;
      for (var i = 0; i < originalMappings.length; i++) {
        if (originalMappings[i] != null) {
          quickSort(originalMappings[i], util.compareByOriginalPositionsNoSource);
        }
      }
      this.__originalMappings = [].concat(...originalMappings);
    };
    BasicSourceMapConsumer.prototype._findMapping = function SourceMapConsumer_findMapping(aNeedle, aMappings, aLineName, aColumnName, aComparator, aBias) {
      if (aNeedle[aLineName] <= 0) {
        throw new TypeError("Line must be greater than or equal to 1, got " + aNeedle[aLineName]);
      }
      if (aNeedle[aColumnName] < 0) {
        throw new TypeError("Column must be greater than or equal to 0, got " + aNeedle[aColumnName]);
      }
      return binarySearch.search(aNeedle, aMappings, aComparator, aBias);
    };
    BasicSourceMapConsumer.prototype.computeColumnSpans = function SourceMapConsumer_computeColumnSpans() {
      for (var index = 0; index < this._generatedMappings.length; ++index) {
        var mapping = this._generatedMappings[index];
        if (index + 1 < this._generatedMappings.length) {
          var nextMapping = this._generatedMappings[index + 1];
          if (mapping.generatedLine === nextMapping.generatedLine) {
            mapping.lastGeneratedColumn = nextMapping.generatedColumn - 1;
            continue;
          }
        }
        mapping.lastGeneratedColumn = Infinity;
      }
    };
    BasicSourceMapConsumer.prototype.originalPositionFor = function SourceMapConsumer_originalPositionFor(aArgs) {
      var needle = {
        generatedLine: util.getArg(aArgs, "line"),
        generatedColumn: util.getArg(aArgs, "column")
      };
      var index = this._findMapping(
        needle,
        this._generatedMappings,
        "generatedLine",
        "generatedColumn",
        util.compareByGeneratedPositionsDeflated,
        util.getArg(aArgs, "bias", SourceMapConsumer2.GREATEST_LOWER_BOUND)
      );
      if (index >= 0) {
        var mapping = this._generatedMappings[index];
        if (mapping.generatedLine === needle.generatedLine) {
          var source = util.getArg(mapping, "source", null);
          if (source !== null) {
            source = this._sources.at(source);
            source = util.computeSourceURL(this.sourceRoot, source, this._sourceMapURL);
          }
          var name = util.getArg(mapping, "name", null);
          if (name !== null) {
            name = this._names.at(name);
          }
          return {
            source,
            line: util.getArg(mapping, "originalLine", null),
            column: util.getArg(mapping, "originalColumn", null),
            name
          };
        }
      }
      return {
        source: null,
        line: null,
        column: null,
        name: null
      };
    };
    BasicSourceMapConsumer.prototype.hasContentsOfAllSources = function BasicSourceMapConsumer_hasContentsOfAllSources() {
      if (!this.sourcesContent) {
        return false;
      }
      return this.sourcesContent.length >= this._sources.size() && !this.sourcesContent.some(function(sc) {
        return sc == null;
      });
    };
    BasicSourceMapConsumer.prototype.sourceContentFor = function SourceMapConsumer_sourceContentFor(aSource, nullOnMissing) {
      if (!this.sourcesContent) {
        return null;
      }
      var index = this._findSourceIndex(aSource);
      if (index >= 0) {
        return this.sourcesContent[index];
      }
      var relativeSource = aSource;
      if (this.sourceRoot != null) {
        relativeSource = util.relative(this.sourceRoot, relativeSource);
      }
      var url;
      if (this.sourceRoot != null && (url = util.urlParse(this.sourceRoot))) {
        var fileUriAbsPath = relativeSource.replace(/^file:\/\//, "");
        if (url.scheme == "file" && this._sources.has(fileUriAbsPath)) {
          return this.sourcesContent[this._sources.indexOf(fileUriAbsPath)];
        }
        if ((!url.path || url.path == "/") && this._sources.has("/" + relativeSource)) {
          return this.sourcesContent[this._sources.indexOf("/" + relativeSource)];
        }
      }
      if (nullOnMissing) {
        return null;
      } else {
        throw new Error('"' + relativeSource + '" is not in the SourceMap.');
      }
    };
    BasicSourceMapConsumer.prototype.generatedPositionFor = function SourceMapConsumer_generatedPositionFor(aArgs) {
      var source = util.getArg(aArgs, "source");
      source = this._findSourceIndex(source);
      if (source < 0) {
        return {
          line: null,
          column: null,
          lastColumn: null
        };
      }
      var needle = {
        source,
        originalLine: util.getArg(aArgs, "line"),
        originalColumn: util.getArg(aArgs, "column")
      };
      var index = this._findMapping(
        needle,
        this._originalMappings,
        "originalLine",
        "originalColumn",
        util.compareByOriginalPositions,
        util.getArg(aArgs, "bias", SourceMapConsumer2.GREATEST_LOWER_BOUND)
      );
      if (index >= 0) {
        var mapping = this._originalMappings[index];
        if (mapping.source === needle.source) {
          return {
            line: util.getArg(mapping, "generatedLine", null),
            column: util.getArg(mapping, "generatedColumn", null),
            lastColumn: util.getArg(mapping, "lastGeneratedColumn", null)
          };
        }
      }
      return {
        line: null,
        column: null,
        lastColumn: null
      };
    };
    exports.BasicSourceMapConsumer = BasicSourceMapConsumer;
    function IndexedSourceMapConsumer(aSourceMap, aSourceMapURL) {
      var sourceMap = aSourceMap;
      if (typeof aSourceMap === "string") {
        sourceMap = util.parseSourceMapInput(aSourceMap);
      }
      var version = util.getArg(sourceMap, "version");
      var sections = util.getArg(sourceMap, "sections");
      if (version != this._version) {
        throw new Error("Unsupported version: " + version);
      }
      this._sources = new ArraySet();
      this._names = new ArraySet();
      var lastOffset = {
        line: -1,
        column: 0
      };
      this._sections = sections.map(function(s) {
        if (s.url) {
          throw new Error("Support for url field in sections not implemented.");
        }
        var offset = util.getArg(s, "offset");
        var offsetLine = util.getArg(offset, "line");
        var offsetColumn = util.getArg(offset, "column");
        if (offsetLine < lastOffset.line || offsetLine === lastOffset.line && offsetColumn < lastOffset.column) {
          throw new Error("Section offsets must be ordered and non-overlapping.");
        }
        lastOffset = offset;
        return {
          generatedOffset: {
            // The offset fields are 0-based, but we use 1-based indices when
            // encoding/decoding from VLQ.
            generatedLine: offsetLine + 1,
            generatedColumn: offsetColumn + 1
          },
          consumer: new SourceMapConsumer2(util.getArg(s, "map"), aSourceMapURL)
        };
      });
    }
    IndexedSourceMapConsumer.prototype = Object.create(SourceMapConsumer2.prototype);
    IndexedSourceMapConsumer.prototype.constructor = SourceMapConsumer2;
    IndexedSourceMapConsumer.prototype._version = 3;
    Object.defineProperty(IndexedSourceMapConsumer.prototype, "sources", {
      get: function() {
        var sources = [];
        for (var i = 0; i < this._sections.length; i++) {
          for (var j = 0; j < this._sections[i].consumer.sources.length; j++) {
            sources.push(this._sections[i].consumer.sources[j]);
          }
        }
        return sources;
      }
    });
    IndexedSourceMapConsumer.prototype.originalPositionFor = function IndexedSourceMapConsumer_originalPositionFor(aArgs) {
      var needle = {
        generatedLine: util.getArg(aArgs, "line"),
        generatedColumn: util.getArg(aArgs, "column")
      };
      var sectionIndex = binarySearch.search(
        needle,
        this._sections,
        function(needle2, section2) {
          var cmp = needle2.generatedLine - section2.generatedOffset.generatedLine;
          if (cmp) {
            return cmp;
          }
          return needle2.generatedColumn - section2.generatedOffset.generatedColumn;
        }
      );
      var section = this._sections[sectionIndex];
      if (!section) {
        return {
          source: null,
          line: null,
          column: null,
          name: null
        };
      }
      return section.consumer.originalPositionFor({
        line: needle.generatedLine - (section.generatedOffset.generatedLine - 1),
        column: needle.generatedColumn - (section.generatedOffset.generatedLine === needle.generatedLine ? section.generatedOffset.generatedColumn - 1 : 0),
        bias: aArgs.bias
      });
    };
    IndexedSourceMapConsumer.prototype.hasContentsOfAllSources = function IndexedSourceMapConsumer_hasContentsOfAllSources() {
      return this._sections.every(function(s) {
        return s.consumer.hasContentsOfAllSources();
      });
    };
    IndexedSourceMapConsumer.prototype.sourceContentFor = function IndexedSourceMapConsumer_sourceContentFor(aSource, nullOnMissing) {
      for (var i = 0; i < this._sections.length; i++) {
        var section = this._sections[i];
        var content = section.consumer.sourceContentFor(aSource, true);
        if (content || content === "") {
          return content;
        }
      }
      if (nullOnMissing) {
        return null;
      } else {
        throw new Error('"' + aSource + '" is not in the SourceMap.');
      }
    };
    IndexedSourceMapConsumer.prototype.generatedPositionFor = function IndexedSourceMapConsumer_generatedPositionFor(aArgs) {
      for (var i = 0; i < this._sections.length; i++) {
        var section = this._sections[i];
        if (section.consumer._findSourceIndex(util.getArg(aArgs, "source")) === -1) {
          continue;
        }
        var generatedPosition = section.consumer.generatedPositionFor(aArgs);
        if (generatedPosition) {
          var ret = {
            line: generatedPosition.line + (section.generatedOffset.generatedLine - 1),
            column: generatedPosition.column + (section.generatedOffset.generatedLine === generatedPosition.line ? section.generatedOffset.generatedColumn - 1 : 0)
          };
          return ret;
        }
      }
      return {
        line: null,
        column: null
      };
    };
    IndexedSourceMapConsumer.prototype._parseMappings = function IndexedSourceMapConsumer_parseMappings(aStr, aSourceRoot) {
      this.__generatedMappings = [];
      this.__originalMappings = [];
      for (var i = 0; i < this._sections.length; i++) {
        var section = this._sections[i];
        var sectionMappings = section.consumer._generatedMappings;
        for (var j = 0; j < sectionMappings.length; j++) {
          var mapping = sectionMappings[j];
          var source = section.consumer._sources.at(mapping.source);
          if (source !== null) {
            source = util.computeSourceURL(section.consumer.sourceRoot, source, this._sourceMapURL);
          }
          this._sources.add(source);
          source = this._sources.indexOf(source);
          var name = null;
          if (mapping.name) {
            name = section.consumer._names.at(mapping.name);
            this._names.add(name);
            name = this._names.indexOf(name);
          }
          var adjustedMapping = {
            source,
            generatedLine: mapping.generatedLine + (section.generatedOffset.generatedLine - 1),
            generatedColumn: mapping.generatedColumn + (section.generatedOffset.generatedLine === mapping.generatedLine ? section.generatedOffset.generatedColumn - 1 : 0),
            originalLine: mapping.originalLine,
            originalColumn: mapping.originalColumn,
            name
          };
          this.__generatedMappings.push(adjustedMapping);
          if (typeof adjustedMapping.originalLine === "number") {
            this.__originalMappings.push(adjustedMapping);
          }
        }
      }
      quickSort(this.__generatedMappings, util.compareByGeneratedPositionsDeflated);
      quickSort(this.__originalMappings, util.compareByOriginalPositions);
    };
    exports.IndexedSourceMapConsumer = IndexedSourceMapConsumer;
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/source-node.js
var require_source_node = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/lib/source-node.js"(exports) {
    "use strict";
    var SourceMapGenerator = require_source_map_generator().SourceMapGenerator;
    var util = require_util();
    var REGEX_NEWLINE = /(\r?\n)/;
    var NEWLINE_CODE = 10;
    var isSourceNode = "$$$isSourceNode$$$";
    function SourceNode(aLine, aColumn, aSource, aChunks, aName) {
      this.children = [];
      this.sourceContents = {};
      this.line = aLine == null ? null : aLine;
      this.column = aColumn == null ? null : aColumn;
      this.source = aSource == null ? null : aSource;
      this.name = aName == null ? null : aName;
      this[isSourceNode] = true;
      if (aChunks != null)
        this.add(aChunks);
    }
    SourceNode.fromStringWithSourceMap = function SourceNode_fromStringWithSourceMap(aGeneratedCode, aSourceMapConsumer, aRelativePath) {
      var node = new SourceNode();
      var remainingLines = aGeneratedCode.split(REGEX_NEWLINE);
      var remainingLinesIndex = 0;
      var shiftNextLine = function() {
        var lineContents = getNextLine();
        var newLine = getNextLine() || "";
        return lineContents + newLine;
        function getNextLine() {
          return remainingLinesIndex < remainingLines.length ? remainingLines[remainingLinesIndex++] : void 0;
        }
      };
      var lastGeneratedLine = 1, lastGeneratedColumn = 0;
      var lastMapping = null;
      aSourceMapConsumer.eachMapping(function(mapping) {
        if (lastMapping !== null) {
          if (lastGeneratedLine < mapping.generatedLine) {
            addMappingWithCode(lastMapping, shiftNextLine());
            lastGeneratedLine++;
            lastGeneratedColumn = 0;
          } else {
            var nextLine = remainingLines[remainingLinesIndex] || "";
            var code = nextLine.substr(0, mapping.generatedColumn - lastGeneratedColumn);
            remainingLines[remainingLinesIndex] = nextLine.substr(mapping.generatedColumn - lastGeneratedColumn);
            lastGeneratedColumn = mapping.generatedColumn;
            addMappingWithCode(lastMapping, code);
            lastMapping = mapping;
            return;
          }
        }
        while (lastGeneratedLine < mapping.generatedLine) {
          node.add(shiftNextLine());
          lastGeneratedLine++;
        }
        if (lastGeneratedColumn < mapping.generatedColumn) {
          var nextLine = remainingLines[remainingLinesIndex] || "";
          node.add(nextLine.substr(0, mapping.generatedColumn));
          remainingLines[remainingLinesIndex] = nextLine.substr(mapping.generatedColumn);
          lastGeneratedColumn = mapping.generatedColumn;
        }
        lastMapping = mapping;
      }, this);
      if (remainingLinesIndex < remainingLines.length) {
        if (lastMapping) {
          addMappingWithCode(lastMapping, shiftNextLine());
        }
        node.add(remainingLines.splice(remainingLinesIndex).join(""));
      }
      aSourceMapConsumer.sources.forEach(function(sourceFile) {
        var content = aSourceMapConsumer.sourceContentFor(sourceFile);
        if (content != null) {
          if (aRelativePath != null) {
            sourceFile = util.join(aRelativePath, sourceFile);
          }
          node.setSourceContent(sourceFile, content);
        }
      });
      return node;
      function addMappingWithCode(mapping, code) {
        if (mapping === null || mapping.source === void 0) {
          node.add(code);
        } else {
          var source = aRelativePath ? util.join(aRelativePath, mapping.source) : mapping.source;
          node.add(new SourceNode(
            mapping.originalLine,
            mapping.originalColumn,
            source,
            code,
            mapping.name
          ));
        }
      }
    };
    SourceNode.prototype.add = function SourceNode_add(aChunk) {
      if (Array.isArray(aChunk)) {
        aChunk.forEach(function(chunk) {
          this.add(chunk);
        }, this);
      } else if (aChunk[isSourceNode] || typeof aChunk === "string") {
        if (aChunk) {
          this.children.push(aChunk);
        }
      } else {
        throw new TypeError(
          "Expected a SourceNode, string, or an array of SourceNodes and strings. Got " + aChunk
        );
      }
      return this;
    };
    SourceNode.prototype.prepend = function SourceNode_prepend(aChunk) {
      if (Array.isArray(aChunk)) {
        for (var i = aChunk.length - 1; i >= 0; i--) {
          this.prepend(aChunk[i]);
        }
      } else if (aChunk[isSourceNode] || typeof aChunk === "string") {
        this.children.unshift(aChunk);
      } else {
        throw new TypeError(
          "Expected a SourceNode, string, or an array of SourceNodes and strings. Got " + aChunk
        );
      }
      return this;
    };
    SourceNode.prototype.walk = function SourceNode_walk(aFn) {
      var chunk;
      for (var i = 0, len = this.children.length; i < len; i++) {
        chunk = this.children[i];
        if (chunk[isSourceNode]) {
          chunk.walk(aFn);
        } else {
          if (chunk !== "") {
            aFn(chunk, {
              source: this.source,
              line: this.line,
              column: this.column,
              name: this.name
            });
          }
        }
      }
    };
    SourceNode.prototype.join = function SourceNode_join(aSep) {
      var newChildren;
      var i;
      var len = this.children.length;
      if (len > 0) {
        newChildren = [];
        for (i = 0; i < len - 1; i++) {
          newChildren.push(this.children[i]);
          newChildren.push(aSep);
        }
        newChildren.push(this.children[i]);
        this.children = newChildren;
      }
      return this;
    };
    SourceNode.prototype.replaceRight = function SourceNode_replaceRight(aPattern, aReplacement) {
      var lastChild = this.children[this.children.length - 1];
      if (lastChild[isSourceNode]) {
        lastChild.replaceRight(aPattern, aReplacement);
      } else if (typeof lastChild === "string") {
        this.children[this.children.length - 1] = lastChild.replace(aPattern, aReplacement);
      } else {
        this.children.push("".replace(aPattern, aReplacement));
      }
      return this;
    };
    SourceNode.prototype.setSourceContent = function SourceNode_setSourceContent(aSourceFile, aSourceContent) {
      this.sourceContents[util.toSetString(aSourceFile)] = aSourceContent;
    };
    SourceNode.prototype.walkSourceContents = function SourceNode_walkSourceContents(aFn) {
      for (var i = 0, len = this.children.length; i < len; i++) {
        if (this.children[i][isSourceNode]) {
          this.children[i].walkSourceContents(aFn);
        }
      }
      var sources = Object.keys(this.sourceContents);
      for (var i = 0, len = sources.length; i < len; i++) {
        aFn(util.fromSetString(sources[i]), this.sourceContents[sources[i]]);
      }
    };
    SourceNode.prototype.toString = function SourceNode_toString() {
      var str = "";
      this.walk(function(chunk) {
        str += chunk;
      });
      return str;
    };
    SourceNode.prototype.toStringWithSourceMap = function SourceNode_toStringWithSourceMap(aArgs) {
      var generated = {
        code: "",
        line: 1,
        column: 0
      };
      var map = new SourceMapGenerator(aArgs);
      var sourceMappingActive = false;
      var lastOriginalSource = null;
      var lastOriginalLine = null;
      var lastOriginalColumn = null;
      var lastOriginalName = null;
      this.walk(function(chunk, original) {
        generated.code += chunk;
        if (original.source !== null && original.line !== null && original.column !== null) {
          if (lastOriginalSource !== original.source || lastOriginalLine !== original.line || lastOriginalColumn !== original.column || lastOriginalName !== original.name) {
            map.addMapping({
              source: original.source,
              original: {
                line: original.line,
                column: original.column
              },
              generated: {
                line: generated.line,
                column: generated.column
              },
              name: original.name
            });
          }
          lastOriginalSource = original.source;
          lastOriginalLine = original.line;
          lastOriginalColumn = original.column;
          lastOriginalName = original.name;
          sourceMappingActive = true;
        } else if (sourceMappingActive) {
          map.addMapping({
            generated: {
              line: generated.line,
              column: generated.column
            }
          });
          lastOriginalSource = null;
          sourceMappingActive = false;
        }
        for (var idx = 0, length = chunk.length; idx < length; idx++) {
          if (chunk.charCodeAt(idx) === NEWLINE_CODE) {
            generated.line++;
            generated.column = 0;
            if (idx + 1 === length) {
              lastOriginalSource = null;
              sourceMappingActive = false;
            } else if (sourceMappingActive) {
              map.addMapping({
                source: original.source,
                original: {
                  line: original.line,
                  column: original.column
                },
                generated: {
                  line: generated.line,
                  column: generated.column
                },
                name: original.name
              });
            }
          } else {
            generated.column++;
          }
        }
      });
      this.walkSourceContents(function(sourceFile, sourceContent) {
        map.setSourceContent(sourceFile, sourceContent);
      });
      return { code: generated.code, map };
    };
    exports.SourceNode = SourceNode;
  }
});

// ../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/source-map.js
var require_source_map = __commonJS({
  "../../node_modules/.pnpm/source-map-js@1.2.1_patch_hash=0354139cbc5dbd66e1bc59167ff8e42d3a9a2169038a35f1b7b1b4b843e08a6c/node_modules/source-map-js/source-map.js"(exports) {
    "use strict";
    exports.SourceMapGenerator = require_source_map_generator().SourceMapGenerator;
    exports.SourceMapConsumer = require_source_map_consumer().SourceMapConsumer;
    exports.SourceNode = require_source_node().SourceNode;
  }
});

// src/vm.ts
var import_source_map_js = __toESM(require_source_map(), 1);
import { newQuickJSWASMModuleFromVariant, shouldInterruptAfterDeadline } from "quickjs-emscripten-core";

// src/compiler/compiler.ts
import jsxPlugin from "@babel/plugin-transform-react-jsx";
import * as Babel from "@babel/standalone";

// src/compiler/plugins/async-iterator.ts
var Open = "async function* __fn__() {";
var Close = "}";
var USER_CODE_START_MARKER = "/* __LLMZ_USER_CODE_START__ */";
var USER_CODE_END_MARKER = "/* __LLMZ_USER_CODE_END__ */";
var AsyncIterator = {
  preProcessing: (code) => {
    return `"use strict";
${Open}
${USER_CODE_START_MARKER}
${code}
${USER_CODE_END_MARKER}
${Close}
`.trim();
  },
  postProcessing: (code) => {
    code = code.slice(code.indexOf(Open) + Open.length).trim();
    code = code.slice(0, code.lastIndexOf(Close));
    code = code.replace(USER_CODE_START_MARKER, "").replace(USER_CODE_END_MARKER, "");
    return code;
  }
};

// src/compiler/plugins/braces-tsx.ts
import { parse } from "@babel/parser";
function fixTSXUntilNoErrors(rawCode, maxAttempts = 50) {
  let code = rawCode;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      parse(code, {
        sourceType: "module",
        plugins: ["typescript", "jsx"]
      });
      return code;
    } catch (err) {
      if (err.code !== "BABEL_PARSER_SYNTAX_ERROR" || !err.loc) {
        throw err;
      }
      const fixed = fixOneBraceProblemAt(code, err.loc);
      if (!fixed) {
        throw err;
      }
      code = fixed;
    }
  }
  throw new Error("Hit maxAttempts while trying to fix TSX code");
}
function fixOneBraceProblemAt(code, loc) {
  const lines = code.split("\n");
  if (loc.line < 1 || loc.line > lines.length) {
    return null;
  }
  const lineStr = lines[loc.line - 1];
  if (loc.column < 0 || lineStr && loc.column > lineStr.length) {
    return null;
  }
  const absIndex = lines.slice(0, loc.line - 1).reduce((sum, l) => sum + l.length + 1, 0) + loc.column;
  const braceIndex = findNearestBrace(code, absIndex);
  if (braceIndex == null) {
    return null;
  }
  const braceChar = code[braceIndex];
  if (braceChar !== "{" && braceChar !== "}") {
    return null;
  }
  const entity = braceChar === "{" ? "&#123;" : "&#125;";
  return code.slice(0, braceIndex) + entity + code.slice(braceIndex + 1);
}
function findNearestBrace(code, startIndex) {
  for (let i = startIndex; i >= 0; i--) {
    if (code[i] === "{" || code[i] === "}") {
      return i;
    }
  }
  for (let i = startIndex; i < code.length; i++) {
    if (code[i] === "{" || code[i] === "}") {
      return i;
    }
  }
  return null;
}
var HTML_ENTITIES = {
  "{": "&#123;",
  "}": "&#125;",
  "<": "&lt;",
  ">": "&gt;",
  "&": "&amp;",
  '"': "&quot;",
  "'": "&#39;",
  "`": "&#96;"
};
function escapeBracesInCodeFences(tsx) {
  let i = 0;
  let output = "";
  let insideFencedBlock = false;
  let jsxDepth = 0;
  let braceDepth = 0;
  while (i < tsx.length) {
    const char = tsx[i];
    if (char === "<" && !insideFencedBlock) {
      if (tsx[i + 1] !== "/") {
        jsxDepth++;
      } else {
        jsxDepth--;
      }
      output += char;
      i++;
      continue;
    }
    if (jsxDepth > 0 && !insideFencedBlock) {
      if (char === "{") {
        braceDepth++;
        output += char;
        i++;
        continue;
      } else if (char === "}") {
        braceDepth--;
        output += char;
        i++;
        continue;
      }
    }
    if (tsx.startsWith("```", i)) {
      insideFencedBlock = !insideFencedBlock;
      output += "```";
      i += 3;
      continue;
    }
    if (jsxDepth > 0 && braceDepth === 0 && !insideFencedBlock && char === "`") {
      const closingIndex = tsx.indexOf("`", i + 1);
      if (closingIndex !== -1) {
        const inlineCode = tsx.slice(i + 1, closingIndex);
        let escaped = "`";
        for (const c of inlineCode) {
          escaped += HTML_ENTITIES[c] || c;
        }
        escaped += "`";
        output += escaped;
        i = closingIndex + 1;
        continue;
      }
    }
    if (insideFencedBlock && char in HTML_ENTITIES) {
      output += HTML_ENTITIES[char];
      i++;
      continue;
    }
    output += char;
    i++;
  }
  return output;
}
var JSXMarkdown = {
  preProcessing: (code) => {
    code = escapeBracesInCodeFences(code);
    code = fixTSXUntilNoErrors(code);
    return code;
  }
};

// src/compiler/plugins/html-to-markdown.ts
import { types as t } from "@babel/core";
function htmlToMarkdownPlugin() {
  return {
    name: "html-to-markdown",
    visitor: {
      JSXElement(path) {
        const { openingElement } = path.node;
        const tagName = t.isJSXIdentifier(openingElement.name) ? openingElement.name.name : null;
        if (!tagName)
          return;
        const lowerTagName = tagName.toLowerCase();
        if (lowerTagName === "strong" || lowerTagName === "b") {
          convertToMarkdown(path, "**", "**");
        } else if (lowerTagName === "em" || lowerTagName === "i") {
          convertToMarkdown(path, "*", "*");
        } else if (lowerTagName === "u") {
          convertToMarkdown(path, "__", "__");
        } else if (lowerTagName === "a") {
          convertLinkToMarkdown(path);
        } else if (lowerTagName === "br") {
          convertBrToNewline(path);
        } else if (lowerTagName === "p") {
          convertParagraphToMarkdown(path);
        } else if (lowerTagName === "ul") {
          convertUnorderedListToMarkdown(path);
        } else if (lowerTagName === "ol") {
          convertOrderedListToMarkdown(path);
        }
      }
    }
  };
}
function hasOnlyTextChildren(children) {
  return children.every(
    (child) => t.isJSXText(child) || t.isJSXExpressionContainer(child) && !t.isJSXElement(child.expression) || t.isJSXElement(child) && canConvertToMarkdown(child)
  );
}
function canConvertToMarkdown(element) {
  const { openingElement } = element;
  const tagName = t.isJSXIdentifier(openingElement.name) ? openingElement.name.name.toLowerCase() : null;
  if (!tagName)
    return false;
  const allowedTags = ["strong", "b", "em", "i", "u", "br", "p", "a", "li", "ul", "ol"];
  if (!allowedTags.includes(tagName))
    return false;
  if (tagName === "a") {
    return openingElement.attributes.length === 1 && t.isJSXAttribute(openingElement.attributes[0]) && t.isJSXIdentifier(openingElement.attributes[0].name) && openingElement.attributes[0].name.name === "href";
  }
  if (tagName === "br") {
    return openingElement.attributes.length === 0;
  }
  return openingElement.attributes.length === 0;
}
function convertToMarkdown(path, prefix, suffix) {
  const element = path.node;
  if (!canConvertToMarkdown(element))
    return;
  const children = element.children;
  if (!hasOnlyTextChildren(children))
    return;
  const textContent = extractTextContent(children);
  if (!textContent)
    return;
  const markdownText = `${prefix}${textContent}${suffix}`;
  path.replaceWith(t.jsxText(markdownText));
}
function convertLinkToMarkdown(path) {
  const element = path.node;
  if (!canConvertToMarkdown(element))
    return;
  const children = element.children;
  if (!hasOnlyTextChildren(children))
    return;
  const hrefAttr = element.openingElement.attributes.find(
    (attr) => t.isJSXAttribute(attr) && t.isJSXIdentifier(attr.name) && attr.name.name === "href"
  );
  if (!hrefAttr || !t.isJSXAttribute(hrefAttr))
    return;
  let href = "";
  if (t.isStringLiteral(hrefAttr.value)) {
    href = hrefAttr.value.value;
  } else if (t.isJSXExpressionContainer(hrefAttr.value) && t.isStringLiteral(hrefAttr.value.expression)) {
    href = hrefAttr.value.expression.value;
  } else {
    return;
  }
  const textContent = extractTextContent(children);
  if (!textContent)
    return;
  const markdownText = `[${textContent}](${href})`;
  path.replaceWith(t.jsxText(markdownText));
}
function convertBrToNewline(path) {
  const element = path.node;
  if (!canConvertToMarkdown(element))
    return;
  path.replaceWith(t.jsxText("\n"));
}
function convertParagraphToMarkdown(path) {
  const element = path.node;
  if (!canConvertToMarkdown(element))
    return;
  const children = element.children;
  if (!hasOnlyTextChildren(children))
    return;
  const textContent = extractTextContent(children);
  if (!textContent)
    return;
  const markdownText = `${textContent}

`;
  path.replaceWith(t.jsxText(markdownText));
}
function convertUnorderedListToMarkdown(path) {
  const element = path.node;
  if (element.openingElement.attributes.length > 0)
    return;
  const children = element.children;
  const listItems = [];
  for (const child of children) {
    if (t.isJSXElement(child)) {
      const tagName = t.isJSXIdentifier(child.openingElement.name) ? child.openingElement.name.name : null;
      if ((tagName == null ? void 0 : tagName.toLowerCase()) === "li" && canConvertToMarkdown(child)) {
        const itemText = extractTextContent(child.children);
        if (itemText) {
          listItems.push(`- ${itemText}`);
        }
      } else {
        return;
      }
    } else if (t.isJSXText(child) && child.value.trim()) {
      return;
    }
  }
  if (listItems.length === 0)
    return;
  const markdownText = listItems.join("\n") + "\n";
  path.replaceWith(t.jsxText(markdownText));
}
function convertOrderedListToMarkdown(path) {
  const element = path.node;
  if (element.openingElement.attributes.length > 0)
    return;
  const children = element.children;
  const listItems = [];
  for (const child of children) {
    if (t.isJSXElement(child)) {
      const tagName = t.isJSXIdentifier(child.openingElement.name) ? child.openingElement.name.name : null;
      if ((tagName == null ? void 0 : tagName.toLowerCase()) === "li" && canConvertToMarkdown(child)) {
        const itemText = extractTextContent(child.children);
        if (itemText) {
          listItems.push(`${listItems.length + 1}. ${itemText}`);
        }
      } else {
        return;
      }
    } else if (t.isJSXText(child) && child.value.trim()) {
      return;
    }
  }
  if (listItems.length === 0)
    return;
  const markdownText = listItems.join("\n") + "\n";
  path.replaceWith(t.jsxText(markdownText));
}
function extractTextContent(children) {
  let text = "";
  for (const child of children) {
    if (t.isJSXText(child)) {
      text += child.value;
    } else if (t.isJSXExpressionContainer(child)) {
      return "";
    } else if (t.isJSXElement(child)) {
      const nestedText = extractTextContent(child.children);
      if (!nestedText)
        return "";
      const tagName = t.isJSXIdentifier(child.openingElement.name) ? child.openingElement.name.name.toLowerCase() : null;
      switch (tagName) {
        case "strong":
        case "b":
          text += `**${nestedText}**`;
          break;
        case "em":
        case "i":
          text += `*${nestedText}*`;
          break;
        case "u":
          text += `__${nestedText}__`;
          break;
        case "br":
          text += "\n";
          break;
        default:
          return "";
      }
    }
  }
  return text;
}

// src/compiler/plugins/jsx-preserve-newlines.ts
import {
  isJSXIdentifier,
  jsxIdentifier,
  jsxText
} from "@babel/types";
var preserveJSXNewLines = {
  visitor: {
    JSXText(path) {
      const rawText = path.node.value;
      const modifiedText = rawText.replace(/\n/g, "\n\u200B");
      if (rawText !== modifiedText) {
        path.replaceWith(jsxText(modifiedText));
        path.skip();
      }
    },
    JSXOpeningElement(path) {
      if (isJSXIdentifier(path.node.name)) {
        path.node.name = jsxIdentifier(path.node.name.name.toLowerCase());
      }
    },
    JSXClosingElement(path) {
      if (isJSXIdentifier(path.node.name)) {
        path.node.name = jsxIdentifier(path.node.name.name.toLowerCase());
      }
    }
  }
};
var postProcessing = (code) => {
  return code.replaceAll("\\u200B", "\\n");
};
var JSXNewLines = {
  babelPlugin: preserveJSXNewLines,
  postProcessing
};

// src/compiler/plugins/jsx-undefined-vars.ts
import { types as t2 } from "@babel/core";
function jsxUndefinedVarsPlugin() {
  return {
    name: "jsx-undefined-vars",
    visitor: {
      JSXExpressionContainer(path) {
        const expression = path.node.expression;
        if (t2.isJSXEmptyExpression(expression)) {
          return;
        }
        if (!t2.isIdentifier(expression)) {
          return;
        }
        const varName = expression.name;
        const knownGlobals = /* @__PURE__ */ new Set([
          "undefined",
          "null",
          "true",
          "false",
          "NaN",
          "Infinity",
          "console",
          "Math",
          "JSON",
          "Object",
          "Array",
          "String",
          "Number",
          "Boolean",
          "Date",
          "RegExp",
          "Error",
          "Promise",
          "Symbol"
        ]);
        if (knownGlobals.has(varName)) {
          return;
        }
        const iife = t2.callExpression(
          t2.arrowFunctionExpression(
            [],
            t2.blockStatement([
              t2.tryStatement(
                t2.blockStatement([t2.returnStatement(t2.identifier(varName))]),
                t2.catchClause(
                  null,
                  // no error binding needed
                  t2.blockStatement([t2.returnStatement(t2.stringLiteral(varName))])
                )
              )
            ])
          ),
          []
        );
        path.node.expression = iife;
      }
    }
  };
}

// src/compiler/plugins/line-tracking.ts
import { callExpression, expressionStatement, identifier, numericLiteral } from "@babel/types";
var LineTrackingFnIdentifier = "__track__";
var lineTrackingBabelPlugin = function({ types: t3 }) {
  return {
    visitor: {
      Program(path) {
        const trackedLines = /* @__PURE__ */ new Set();
        const lineTrackingCall = (line) => {
          trackedLines.add(line);
          return expressionStatement(callExpression(identifier(LineTrackingFnIdentifier), [numericLiteral(line)]));
        };
        path.traverse({
          enter(path2) {
            if (path2.node.loc && !path2.isProgram()) {
              const { line: startLine } = path2.node.loc.start;
              const node = path2.node;
              const track = () => {
                try {
                  path2.insertBefore(lineTrackingCall(startLine));
                } catch (e) {
                  console.error(e);
                }
              };
              if (t3.isFunctionDeclaration(node) || t3.isClassMethod(node) || t3.isReturnStatement(node) || t3.isAwaitExpression(node) || t3.isForOfStatement(node) || t3.isCallExpression(node) || t3.isUnaryExpression(node) || t3.isVariableDeclaration(node) || t3.isBlockStatement(path2.parent) || t3.isAssignmentExpression(node)) {
                if (!trackedLines.has(startLine) && !t3.isArrowFunctionExpression(path2.parent) && !t3.isObjectProperty(path2.parent) && !t3.isAwaitExpression(path2.parent) && !t3.isUnaryExpression(path2.parent) && !t3.isReturnStatement(path2.parent) && !t3.isForOfStatement(path2.parent) && !t3.isForStatement(path2.parent) && !t3.isForInStatement(path2.parent) && !t3.isVariableDeclaration(path2.parent) && !t3.isBinaryExpression(path2.parent) && !t3.isLogicalExpression(path2.parent)) {
                  track();
                }
              }
            }
          }
        });
      }
    }
  };
};

// src/compiler/plugins/replace-comment.ts
var CommentFnIdentifier = "__comment__";
var replaceCommentBabelPlugin = function({
  types: t3
}) {
  return {
    visitor: {
      Program(path) {
        const processed = /* @__PURE__ */ new Set();
        path.traverse({
          enter(path2) {
            const node = path2.node;
            function processComments(comments, insertMethod) {
              if (!comments) {
                return;
              }
              comments.sort((a, b) => a.start > b.start ? -1 : 1).forEach((comment) => {
                var _a;
                if (processed.has(comment.loc)) {
                  return;
                }
                processed.add(comment.loc);
                const commentCall = t3.expressionStatement(
                  t3.callExpression(t3.identifier(CommentFnIdentifier), [
                    t3.stringLiteral(comment.value.trim()),
                    t3.numericLiteral(((_a = comment.loc) == null ? void 0 : _a.start.line) ?? 0)
                  ])
                );
                const isInsideObjectProperty = t3.isObjectProperty(node) || path2.findParent((path3) => t3.isObjectProperty(path3.node));
                if (!isInsideObjectProperty) {
                  insertMethod(commentCall);
                }
              });
              comments.length = 0;
            }
            processComments(node.trailingComments ?? [], (commentCall) => path2.insertAfter(commentCall));
            processComments(node.leadingComments ?? [], (commentCall) => path2.insertBefore(commentCall));
            processComments(node.innerComments ?? [], (commentCall) => path2.insertAfter(commentCall));
          }
        });
      }
    }
  };
};

// src/compiler/plugins/return-async.ts
var instrumentLastLinePlugin = function({
  types: t3
}) {
  return {
    visitor: {
      FunctionDeclaration(path) {
        var _a, _b;
        if (((_a = path.node.id) == null ? void 0 : _a.name) !== "__fn__") {
          return;
        }
        const statements = path.node.body.body;
        if (statements.length === 0) {
          return;
        }
        const lastStatement = statements[statements.length - 1];
        if (t3.isReturnStatement(lastStatement)) {
          if (t3.isExpression(lastStatement.argument)) {
            lastStatement.argument = t3.awaitExpression(lastStatement.argument);
          }
          return;
        }
        if (t3.isExpressionStatement(lastStatement) && (t3.isCallExpression(lastStatement.expression) || t3.isAwaitExpression(lastStatement.expression))) {
          const returnStatement3 = t3.returnStatement(t3.awaitExpression(lastStatement.expression));
          (_b = path.get("body").get("body")[statements.length - 1]) == null ? void 0 : _b.replaceWith(returnStatement3);
        }
      }
    }
  };
};

// src/compiler/plugins/track-tool-calls.ts
import {
  expressionStatement as expressionStatement2,
  identifier as identifier2,
  stringLiteral,
  callExpression as callExpression2,
  arrowFunctionExpression,
  blockStatement,
  returnStatement,
  tryStatement,
  numericLiteral as numericLiteral2,
  catchClause,
  throwStatement,
  awaitExpression,
  variableDeclaration,
  variableDeclarator,
  newExpression,
  memberExpression,
  assignmentExpression,
  binaryExpression
} from "@babel/types";
var ToolCallTrackerFnIdentifier = "__toolc__";
var ToolTrackerRetIdentifier = "__ret__";
var extractLVal = (path) => {
  let source = path.getSource();
  if (path.isIdentifier()) {
    return {
      type: "single",
      left: source,
      evalFn: `let ${source} = arguments[0]; return { ${source} };`
    };
  }
  if (path.isArrayPattern()) {
    source = path.get("elements").map((el) => el.getSource()).join(", ");
    return {
      type: "array",
      left: source,
      evalFn: `let [${source}] = arguments[0] ?? []; return { ${source} };`
    };
  }
  if (path.isObjectPattern()) {
    return {
      type: "object",
      left: source,
      evalFn: `let ${source} = arguments[0] ?? {}; return ${source};`
    };
  }
  return {
    type: "unsupported",
    evalFn: "",
    left: ""
  };
};
var toolCallTrackingPlugin = (calls = /* @__PURE__ */ new Map()) => function({}) {
  let callId = 0;
  const skip = /* @__PURE__ */ new Set();
  return {
    visitor: {
      Program() {
        callId = 0;
        skip.clear();
      },
      CallExpression(path) {
        var _a;
        if (skip.has(path.node) || path.findParent((p) => skip.has(p.node))) {
          return;
        }
        if (path.parentPath.isYieldExpression()) {
          return;
        }
        let lval = null;
        const declaration = path.findParent(
          (p) => p.isVariableDeclaration()
        );
        const assignment = path.findParent(
          (p) => p.isAssignmentExpression()
        );
        if (declaration) {
          lval = (_a = declaration.get("declarations")[0]) == null ? void 0 : _a.get("id");
        }
        if (assignment) {
          const left = assignment.get("left");
          if (left.isLVal()) {
            lval = left;
          }
        }
        const endStatement = (value) => expressionStatement2(
          callExpression2(identifier2(ToolCallTrackerFnIdentifier), [
            numericLiteral2(callId),
            stringLiteral("end"),
            value
          ])
        );
        const isAsync = path.parentPath.isAwaitExpression();
        const tryBodyStatement = blockStatement([
          expressionStatement2(
            callExpression2(identifier2(ToolCallTrackerFnIdentifier), [numericLiteral2(callId), stringLiteral("start")])
          ),
          variableDeclaration("const", [
            variableDeclarator(
              identifier2(ToolTrackerRetIdentifier),
              isAsync ? awaitExpression(path.node) : path.node
            )
          ]),
          endStatement(identifier2(ToolTrackerRetIdentifier)),
          returnStatement(identifier2(ToolTrackerRetIdentifier))
        ]);
        if (!lval) {
          if (path.getSource().trim().length) {
            const newError = identifier2("__newError");
            const newCall2 = callExpression2(
              arrowFunctionExpression(
                [],
                // params
                blockStatement([
                  tryStatement(
                    tryBodyStatement,
                    catchClause(
                      identifier2("err"),
                      blockStatement([
                        endStatement(identifier2("err")),
                        variableDeclaration("const", [
                          variableDeclarator(
                            newError,
                            newExpression(identifier2("Error"), [
                              memberExpression(identifier2("err"), identifier2("message"))
                            ])
                          )
                        ]),
                        expressionStatement2(
                          assignmentExpression(
                            "=",
                            memberExpression(newError, identifier2("stack")),
                            binaryExpression(
                              "+",
                              memberExpression(identifier2("err"), identifier2("stack")),
                              binaryExpression(
                                "+",
                                stringLiteral("\n"),
                                memberExpression(newError, identifier2("stack"))
                              )
                            )
                          )
                        ),
                        throwStatement(newError)
                      ])
                    )
                  )
                ]),
                isAsync
              ),
              []
              // args
            );
            callId++;
            skip.add(newCall2);
            path.replaceWith(newCall2);
          }
          return;
        }
        callId++;
        const parts = path.get("callee").getSource().split(".");
        const object = parts.length === 1 ? "global" : parts[0];
        const tool = parts.length === 1 ? parts[0] : parts[1];
        const assign = extractLVal(lval);
        if (assign.type === "unsupported") {
          return;
        }
        const newCall = callExpression2(
          arrowFunctionExpression(
            [],
            // params
            blockStatement([
              tryStatement(
                tryBodyStatement,
                catchClause(
                  identifier2("err"),
                  blockStatement([
                    endStatement(identifier2("err")),
                    throwStatement(
                      newExpression(identifier2("Error"), [memberExpression(identifier2("err"), identifier2("message"))])
                    )
                  ])
                )
              )
            ]),
            isAsync
          ),
          []
          // args
        );
        calls.set(callId, { object, tool, assignment: assign });
        skip.add(newCall);
        path.replaceWith(newCall);
      }
    }
  };
};

// src/compiler/plugins/variable-extraction.ts
import {
  expressionStatement as expressionStatement3,
  identifier as identifier3,
  stringLiteral as stringLiteral2,
  callExpression as callExpression3,
  arrowFunctionExpression as arrowFunctionExpression2,
  blockStatement as blockStatement2,
  returnStatement as returnStatement2
} from "@babel/types";
var VariableTrackingFnIdentifier = "__var__";
var variableTrackingPlugin = (variables = /* @__PURE__ */ new Set()) => function({ types: t3 }) {
  let trackingStatements = [];
  const trackVariable = (variableName) => {
    if (variableName.startsWith("__")) {
      return;
    }
    variables.add(variableName);
    trackingStatements.push(
      expressionStatement3(
        callExpression3(identifier3(VariableTrackingFnIdentifier), [
          stringLiteral2(variableName),
          arrowFunctionExpression2([], callExpression3(identifier3("eval"), [stringLiteral2(variableName)]))
        ])
      )
    );
  };
  return {
    visitor: {
      FunctionDeclaration(path) {
        path.node.params.forEach((param) => {
          if (t3.isIdentifier(param)) {
            trackVariable(param.name);
          } else if (t3.isAssignmentPattern(param) && t3.isIdentifier(param.left)) {
            trackVariable(param.left.name);
          } else if (t3.isObjectPattern(param)) {
            param.properties.forEach((prop) => {
              if (t3.isObjectProperty(prop) && t3.isIdentifier(prop.value)) {
                trackVariable(prop.value.name);
              }
            });
          } else if (t3.isArrayPattern(param)) {
            param.elements.forEach((element) => {
              if (t3.isIdentifier(element)) {
                trackVariable(element.name);
              }
            });
          }
        });
        if (trackingStatements.length) {
          path.get("body").unshiftContainer("body", trackingStatements);
          trackingStatements = [];
        }
      },
      ArrowFunctionExpression(path) {
        path.node.params.forEach((param) => {
          if (t3.isIdentifier(param)) {
            trackVariable(param.name);
          } else if (t3.isAssignmentPattern(param) && t3.isIdentifier(param.left)) {
            trackVariable(param.left.name);
          } else if (t3.isObjectPattern(param)) {
            param.properties.forEach((prop) => {
              if (t3.isObjectProperty(prop) && t3.isIdentifier(prop.value)) {
                trackVariable(prop.value.name);
              }
            });
          } else if (t3.isArrayPattern(param)) {
            param.elements.forEach((element) => {
              if (t3.isIdentifier(element)) {
                trackVariable(element.name);
              }
            });
          }
        });
        if (trackingStatements.length) {
          const body = path.get("body");
          if (body.isBlockStatement()) {
            body.unshiftContainer("body", trackingStatements);
          } else if (body.isExpression() && body.getSource().trim().length) {
            const newBody = blockStatement2([...trackingStatements, returnStatement2(body.node)]);
            path.get("body").replaceWith(newBody);
          }
          trackingStatements = [];
        }
      },
      VariableDeclaration(path) {
        const parent = path.parentPath;
        const handleObjectPattern = (pattern) => {
          pattern.properties.forEach((prop) => {
            if (t3.isObjectProperty(prop) && t3.isIdentifier(prop.value)) {
              trackVariable(prop.value.name);
            } else if (t3.isObjectProperty(prop) && t3.isObjectPattern(prop.value)) {
              handleObjectPattern(prop.value);
            } else if (t3.isObjectProperty(prop) && t3.isAssignmentPattern(prop.value) && t3.isIdentifier(prop.value.left)) {
              trackVariable(prop.value.left.name);
            }
          });
        };
        path.node.declarations.forEach((declarator) => {
          if (parent.isForXStatement() || parent.isForStatement()) {
            return;
          }
          if (t3.isIdentifier(declarator.id)) {
            trackVariable(declarator.id.name);
          }
          if (t3.isObjectPattern(declarator.id)) {
            handleObjectPattern(declarator.id);
          }
          if (t3.isArrayPattern(declarator.id)) {
            declarator.id.elements.forEach((element) => {
              if (t3.isIdentifier(element)) {
                trackVariable(element.name);
              } else if (t3.isRestElement(element) && t3.isIdentifier(element.argument)) {
                trackVariable(element.argument.name);
              } else if (t3.isAssignmentPattern(element) && t3.isIdentifier(element.left)) {
                trackVariable(element.left.name);
              }
            });
          }
        });
        if (trackingStatements.length) {
          path.insertAfter(trackingStatements);
          trackingStatements = [];
        }
      }
    }
  };
};

// src/compiler/compiler.ts
var DEFAULT_TRANSFORM_OPTIONS = {
  parserOpts: {
    allowReturnOutsideFunction: true,
    allowAwaitOutsideFunction: true,
    strictMode: true,
    startLine: 1
  },
  sourceFileName: "<anonymous>",
  sourceMaps: true,
  minified: false,
  retainLines: true
};
var Identifiers = {
  JSXFnIdentifier: "__jsx__",
  AsyncIterYieldFnIdentifier: "__async_iter_yield__",
  ConsoleObjIdentifier: "console",
  LineTrackingFnIdentifier,
  VariableTrackingFnIdentifier,
  ToolCallTrackerFnIdentifier,
  ToolTrackerRetIdentifier,
  CommentFnIdentifier
};
function compile(code) {
  code = AsyncIterator.preProcessing(code);
  code = JSXMarkdown.preProcessing(code);
  let output = Babel.transform(code, {
    parserOpts: {
      allowReturnOutsideFunction: true,
      allowAwaitOutsideFunction: true
    },
    presets: ["typescript"],
    plugins: [
      JSXNewLines.babelPlugin,
      htmlToMarkdownPlugin,
      // Convert simple HTML to markdown first
      jsxUndefinedVarsPlugin,
      // Must run BEFORE JSX transform
      [
        jsxPlugin,
        {
          throwIfNamespace: false,
          runtime: "classic",
          pragma: Identifiers.JSXFnIdentifier
        }
      ],
      instrumentLastLinePlugin
    ],
    filename: "<anonymous>.tsx",
    sourceFileName: "<anonymous>",
    sourceMaps: true,
    minified: false,
    retainLines: true
  });
  const variables = /* @__PURE__ */ new Set();
  const toolCalls = /* @__PURE__ */ new Map();
  const codeWithMarkers = output.code;
  output = Babel.transform(output.code, {
    ...DEFAULT_TRANSFORM_OPTIONS,
    parserOpts: {
      ...DEFAULT_TRANSFORM_OPTIONS.parserOpts
    },
    plugins: [
      lineTrackingBabelPlugin,
      replaceCommentBabelPlugin,
      variableTrackingPlugin(variables),
      toolCallTrackingPlugin(toolCalls)
    ],
    retainLines: true
  });
  let outputCode = output.code;
  outputCode = AsyncIterator.postProcessing(outputCode);
  outputCode = JSXNewLines.postProcessing(outputCode);
  return {
    code: outputCode,
    codeWithMarkers,
    // Code from before second transform, still has literal markers
    map: output.map,
    variables,
    toolCalls
  };
}

// src/quickjs-variant.ts
var getVariant = async () => {
  const module = await import("@jitl/quickjs-singlefile-browser-release-sync");
  return module.default;
};
var BundledReleaseSyncVariant = {
  type: "sync",
  importFFI: async () => {
    const variant = await getVariant();
    return variant.importFFI();
  },
  importModuleLoader: async () => {
    const variant = await getVariant();
    return variant.importModuleLoader();
  }
};

// src/vm.ts
var MAX_VM_EXECUTION_TIME = 6e4;
var NO_TRACKING = [
  Identifiers.CommentFnIdentifier,
  Identifiers.ToolCallTrackerFnIdentifier,
  Identifiers.ToolTrackerRetIdentifier,
  Identifiers.VariableTrackingFnIdentifier,
  Identifiers.JSXFnIdentifier,
  Identifiers.ConsoleObjIdentifier
];
function getCompiledCode(code, traces = []) {
  try {
    return compile(code);
  } catch (err) {
    traces.push({
      type: "invalid_code_exception",
      message: (err == null ? void 0 : err.message) ?? "Unknown error",
      code,
      started_at: Date.now()
    });
    throw new InvalidCodeError(err.message, code);
  }
}
async function runAsyncFunction(context, code, traces = [], signal = null, timeout = MAX_VM_EXECUTION_TIME) {
  var _a, _b, _c, _d, _e, _f, _g;
  const transformed = getCompiledCode(code, traces);
  const lines_executed = /* @__PURE__ */ new Map();
  const variables = {};
  const consumer = new import_source_map_js.SourceMapConsumer({
    version: transformed.map.version.toString(),
    mappings: transformed.map.mappings,
    names: transformed.map.names,
    sources: [transformed.map.file],
    sourcesContent: [transformed.code],
    file: transformed.map.file,
    sourceRoot: transformed.map.sourceRoot
  });
  context ??= {};
  for (const name of Array.from(transformed.variables)) {
    delete context[name];
  }
  let DRIVER = "quickjs";
  if (typeof process !== "undefined" && ((_a = process == null ? void 0 : process.env) == null ? void 0 : _a.USE_QUICKJS) === "false") {
    DRIVER = "node";
  }
  if (DRIVER === "quickjs") {
    try {
      context[Identifiers.CommentFnIdentifier] = (comment, line) => {
        if (comment.includes("__LLMZ_USER_CODE_START__") || comment.includes("__LLMZ_USER_CODE_END__")) {
          return;
        }
        traces.push({ type: "comment", comment, line, started_at: Date.now() });
      };
      const codeWithMarkers = transformed.codeWithMarkers || transformed.code;
      const markerLines = codeWithMarkers.split("\n");
      const USER_CODE_START_MARKER2 = "/* __LLMZ_USER_CODE_START__ */";
      let userCodeStartLine = -1;
      for (let i = 0; i < markerLines.length; i++) {
        if ((_b = markerLines[i]) == null ? void 0 : _b.includes(USER_CODE_START_MARKER2)) {
          userCodeStartLine = i + 1;
          break;
        }
      }
      context[Identifiers.LineTrackingFnIdentifier] = (line) => {
        const originalLine = consumer.originalPositionFor({
          line,
          column: 0
        });
        const mappedLine = originalLine.line ?? line;
        const userCodeLine = Math.max(1, mappedLine - userCodeStartLine);
        lines_executed.set(userCodeLine, (lines_executed.get(userCodeLine) ?? 0) + 1);
      };
      context[Identifiers.JSXFnIdentifier] = (tool, props, ...children) => createJsxComponent({
        type: tool,
        props,
        children
      });
      context[Identifiers.VariableTrackingFnIdentifier] = (name, getter) => {
        if (NO_TRACKING.includes(name)) {
          return;
        }
        variables[name] = () => {
          try {
            const value = getter();
            if (typeof value === "function") {
              return "[[non-primitive]]";
            }
            return value;
          } catch {
            return "[[non-primitive]]";
          }
        };
      };
      let currentToolCall;
      context[Identifiers.ToolCallTrackerFnIdentifier] = (callId, type, outputOrError) => {
        var _a2;
        const temp = Signals.maybeDeserializeError(outputOrError == null ? void 0 : outputOrError.message);
        if (type === "end" && temp instanceof SnapshotSignal && (temp == null ? void 0 : temp.toolCall)) {
          currentToolCall = {
            ...temp.toolCall,
            assignment: (_a2 = transformed.toolCalls.get(callId)) == null ? void 0 : _a2.assignment
          };
        }
      };
      context[Identifiers.ConsoleObjIdentifier] = {
        log: (...args) => {
          const message = args.shift();
          traces.push({ type: "log", message, args, started_at: Date.now() });
        }
      };
      context[Identifiers.AsyncIterYieldFnIdentifier] = async function(value) {
        const startedAt = Date.now();
        try {
          if (typeof value.type !== "string" || value.type.trim().length === 0) {
            throw new Error("A yield statement must yield a valid tool");
          }
          const toolName = Object.keys(context).find((x) => x.toUpperCase() === value.type.toUpperCase());
          if (!toolName) {
            throw new Error(`Yield tool "${value.type}", but tool is not found`);
          }
          await context[toolName](value);
        } finally {
          traces.push({ type: "yield", value, started_at: startedAt, ended_at: Date.now() });
        }
      };
      const QuickJS = await newQuickJSWASMModuleFromVariant(BundledReleaseSyncVariant);
      const runtime = QuickJS.newRuntime();
      runtime.setMemoryLimit(128 * 1024 * 1024);
      const startTime = Date.now();
      const timeoutHandler = shouldInterruptAfterDeadline(startTime + timeout);
      runtime.setInterruptHandler(() => {
        if (signal == null ? void 0 : signal.aborted) {
          return true;
        }
        return timeoutHandler(runtime);
      });
      const vm = runtime.newContext();
      const trackedProperties = /* @__PURE__ */ new Set();
      const referenceProperties = /* @__PURE__ */ new Set();
      const pendingPromises = [];
      const toVmValue = (value) => {
        var _a2;
        if (typeof value === "string") {
          return vm.newString(value);
        } else if (typeof value === "number") {
          return vm.newNumber(value);
        } else if (typeof value === "boolean") {
          return value ? vm.true : vm.false;
        } else if (value === null) {
          return vm.null;
        } else if (value === void 0) {
          return vm.undefined;
        } else if (Array.isArray(value)) {
          const items = value.map((item) => {
            if (typeof item === "string") {
              return JSON.stringify(item);
            } else if (typeof item === "number" || typeof item === "boolean") {
              return String(item);
            } else if (item === null) {
              return "null";
            } else if (item === void 0) {
              return "undefined";
            } else if (typeof item === "object") {
              return JSON.stringify(item);
            }
            return "undefined";
          });
          const arrayLiteral = `[${items.join(",")}]`;
          const result = vm.evalCode(arrayLiteral);
          if ("error" in result) {
            (_a2 = result.error) == null ? void 0 : _a2.dispose();
            return vm.undefined;
          }
          const arrHandle = result.value;
          return arrHandle;
        } else if (typeof value === "object") {
          const obj = vm.newObject();
          for (const [k, v] of Object.entries(value)) {
            if (typeof v !== "function") {
              const propHandle = toVmValue(v);
              vm.setProp(obj, k, propHandle);
              disposeIfNeeded(propHandle);
            }
          }
          return obj;
        }
        return vm.undefined;
      };
      const disposeIfNeeded = (handle) => {
        if (handle !== vm.true && handle !== vm.false && handle !== vm.null && handle !== vm.undefined) {
          handle.dispose();
        }
      };
      const bridgeFunction = (fn, _fnName = "anonymous") => {
        return (...argHandles) => {
          const args = argHandles.map((h) => vm.dump(h));
          try {
            const result = fn(...args);
            if (result && typeof result.then === "function") {
              const promise = vm.newPromise();
              pendingPromises.push({
                hostPromise: result,
                deferredPromise: promise
              });
              void promise.settled.then(() => {
                if (runtime.alive) {
                  runtime.executePendingJobs();
                }
              });
              return promise.handle;
            }
            return toVmValue(result);
          } catch (err) {
            const serialized = err instanceof Error ? err.message : String(err);
            throw new Error(serialized);
          }
        };
      };
      try {
        for (const [key, value] of Object.entries(context)) {
          const descriptor = Object.getOwnPropertyDescriptor(context, key);
          if (descriptor && (descriptor.get || descriptor.set)) {
            referenceProperties.add(key);
            trackedProperties.add(key);
            let getterCode = "undefined";
            if (descriptor.get) {
              const getterBridge = vm.newFunction(`get_${key}`, () => {
                try {
                  const hostValue = context[key];
                  return toVmValue(hostValue);
                } catch (err) {
                  const serialized = err instanceof Error ? err.message : String(err);
                  throw new Error(serialized);
                }
              });
              const getterName = `__getter_${key}__`;
              vm.setProp(vm.global, getterName, getterBridge);
              getterBridge.dispose();
              getterCode = getterName;
            }
            let setterCode = "undefined";
            if (descriptor.set) {
              const setterBridge = vm.newFunction(`set_${key}`, (valueHandle) => {
                try {
                  const jsValue = vm.dump(valueHandle);
                  context[key] = jsValue;
                  return vm.undefined;
                } catch (err) {
                  const serialized = err instanceof Error ? err.message : String(err);
                  throw new Error(serialized);
                }
              });
              const setterName = `__setter_${key}__`;
              vm.setProp(vm.global, setterName, setterBridge);
              setterBridge.dispose();
              setterCode = setterName;
            }
            const definePropertyCode = `
          Object.defineProperty(globalThis, '${key}', {
            enumerable: true,
            configurable: ${descriptor.configurable !== false},
            get: ${getterCode},
            set: ${setterCode}
          });
        `;
            const result = vm.evalCode(definePropertyCode);
            if ("error" in result) {
              (_c = result.error) == null ? void 0 : _c.dispose();
            } else {
              result.value.dispose();
            }
            continue;
          }
          if (typeof value === "function") {
            const fnHandle = vm.newFunction(key, bridgeFunction(value, key));
            vm.setProp(vm.global, key, fnHandle);
            fnHandle.dispose();
          } else if (Array.isArray(value)) {
            trackedProperties.add(key);
            const arrayHandle = toVmValue(value);
            vm.setProp(vm.global, key, arrayHandle);
            disposeIfNeeded(arrayHandle);
          } else if (typeof value === "object" && value !== null) {
            trackedProperties.add(key);
            const objHandle = vm.newObject();
            const props = /* @__PURE__ */ new Set([...Object.keys(value), ...Object.getOwnPropertyNames(value)]);
            const getterSetterProps = [];
            for (const prop of props) {
              const propDescriptor = Object.getOwnPropertyDescriptor(value, prop);
              if (propDescriptor && (propDescriptor.get || propDescriptor.set)) {
                referenceProperties.add(`${key}.${prop}`);
                getterSetterProps.push({ prop, descriptor: propDescriptor });
              } else if (typeof value[prop] === "function") {
                const propFnHandle = vm.newFunction(prop, bridgeFunction(value[prop], `${key}.${prop}`));
                vm.setProp(objHandle, prop, propFnHandle);
                propFnHandle.dispose();
              } else {
                const propHandle = toVmValue(value[prop]);
                vm.setProp(objHandle, prop, propHandle);
                disposeIfNeeded(propHandle);
              }
            }
            vm.setProp(vm.global, key, objHandle);
            objHandle.dispose();
            for (const { prop, descriptor: descriptor2 } of getterSetterProps) {
              let getterCode = "undefined";
              if (descriptor2.get) {
                const getterBridge = vm.newFunction(`get_${prop}`, () => {
                  try {
                    const hostValue = context[key][prop];
                    return toVmValue(hostValue);
                  } catch (err) {
                    const serialized = err instanceof Error ? err.message : String(err);
                    throw new Error(serialized);
                  }
                });
                const getterName = `__getter_${key}_${prop}__`;
                vm.setProp(vm.global, getterName, getterBridge);
                getterBridge.dispose();
                getterCode = getterName;
              }
              let setterCode = "undefined";
              if (descriptor2.set) {
                const setterBridge = vm.newFunction(`set_${prop}`, (valueHandle) => {
                  try {
                    const jsValue = vm.dump(valueHandle);
                    context[key][prop] = jsValue;
                    return vm.undefined;
                  } catch (err) {
                    const serialized = err instanceof Error ? err.message : String(err);
                    throw new Error(serialized);
                  }
                });
                const setterName = `__setter_${key}_${prop}__`;
                vm.setProp(vm.global, setterName, setterBridge);
                setterBridge.dispose();
                setterCode = setterName;
              }
              const definePropertyCode = `
            Object.defineProperty(${key}, '${prop}', {
              enumerable: true,
              configurable: ${descriptor2.configurable !== false},
              get: ${getterCode},
              set: ${setterCode}
            });
          `;
              const result = vm.evalCode(definePropertyCode);
              if ("error" in result) {
                (_d = result.error) == null ? void 0 : _d.dispose();
              } else {
                result.value.dispose();
              }
            }
            if (Object.isSealed(value)) {
              const sealResult = vm.evalCode(`Object.seal(globalThis['${key}']);`);
              if ("error" in sealResult) {
                (_e = sealResult.error) == null ? void 0 : _e.dispose();
              } else {
                sealResult.value.dispose();
              }
            }
            if (!Object.isExtensible(value)) {
              const preventResult = vm.evalCode(`Object.preventExtensions(globalThis['${key}']);`);
              if ("error" in preventResult) {
                (_f = preventResult.error) == null ? void 0 : _f.dispose();
              } else {
                preventResult.value.dispose();
              }
            }
          } else {
            trackedProperties.add(key);
            const valueHandle = toVmValue(value);
            vm.setProp(vm.global, key, valueHandle);
            disposeIfNeeded(valueHandle);
          }
        }
        const varTrackFnHandle = vm.newFunction(
          Identifiers.VariableTrackingFnIdentifier,
          (nameHandle, getterHandle) => {
            var _a2;
            const name = vm.getString(nameHandle);
            if (NO_TRACKING.includes(name)) {
              return;
            }
            try {
              const valueResult = vm.callFunction(getterHandle, vm.undefined);
              if ("error" in valueResult) {
                variables[name] = "[[non-primitive]]";
                (_a2 = valueResult.error) == null ? void 0 : _a2.dispose();
                return;
              }
              const value = vm.dump(valueResult.value);
              valueResult.value.dispose();
              if (typeof value === "function" || typeof value === "string" && value.includes("=>")) {
                variables[name] = "[[non-primitive]]";
              } else {
                variables[name] = value;
              }
            } catch {
              variables[name] = "[[non-primitive]]";
            }
          }
        );
        vm.setProp(vm.global, Identifiers.VariableTrackingFnIdentifier, varTrackFnHandle);
        varTrackFnHandle.dispose();
        const scriptCode = `
"use strict";
globalThis.__llmz_result = undefined;
globalThis.__llmz_result_set = false;
globalThis.__llmz_error = null;
globalThis.__llmz_error_stack = null;
globalThis.__llmz_yields = [];

(async () => {
  try {
    async function* __fn__() {
${transformed.code}
    }

    const fn = __fn__();
    let iteration = 0;
    const maxIterations = 10000; // Safety limit

    while (iteration < maxIterations) {
      const { value, done } = await fn.next();

      if (done) {
        globalThis.__llmz_result = value;
        globalThis.__llmz_result_set = true;
        break;
      }

      // Store yielded value
      globalThis.__llmz_yields.push(value);

      // Call yield handler
      await ${Identifiers.AsyncIterYieldFnIdentifier}(value);

      iteration++;
    }

    if (iteration >= maxIterations) {
      throw new Error('Maximum iterations exceeded');
    }
  } catch (err) {
    // Store both the error message (which may contain serialized signal data)
    // and the stack trace from QuickJS
    // If err is a string (as thrown from promise rejection), use it directly
    // Otherwise extract the message property
    globalThis.__llmz_error = typeof err === 'string' ? err : String(err.message || err || '');
    // Force the stack to be converted to a string in QuickJS context
    globalThis.__llmz_error_stack = '' + (err.stack || '');
  }
})();
`.trim();
        const copyBackContextFromVM = () => {
          for (const key of trackedProperties) {
            if (referenceProperties.has(key)) {
              continue;
            }
            try {
              const valueResult = vm.evalCode(`globalThis['${key}']`);
              const vmValue = vm.dump(valueResult.unwrap());
              valueResult.unwrap().dispose();
              try {
                context[key] = vmValue;
              } catch {
              }
            } catch {
            }
          }
        };
        const execResult = vm.evalCode(scriptCode, "<quickjs>");
        if ("error" in execResult) {
          if (execResult.error) {
            const err = vm.dump(execResult.error);
            execResult.error.dispose();
            throw new Error((err == null ? void 0 : err.message) || "Execution failed");
          }
          throw new Error("Execution failed");
        }
        execResult.value.dispose();
        const maxIterations = 1e3;
        let iteration = 0;
        while (iteration < maxIterations) {
          let hasJobs = false;
          const maxJobs = 1e4;
          for (let i = 0; i < maxJobs; i++) {
            const pending = (_g = runtime.executePendingJobs) == null ? void 0 : _g.call(runtime, -1);
            const jobCount = pending === void 0 ? 0 : pending.unwrap();
            if (jobCount <= 0)
              break;
            hasJobs = true;
          }
          const currentPromises = [...pendingPromises];
          pendingPromises.length = 0;
          if (currentPromises.length > 0) {
            if (signal == null ? void 0 : signal.aborted) {
              const reason = signal.reason;
              const abortMessage = reason instanceof Error ? `${reason.name}: ${reason.message}` : reason ? String(reason) : "Execution was aborted";
              for (const { deferredPromise } of currentPromises) {
                const errValue = vm.newString(abortMessage);
                deferredPromise.reject(errValue);
                errValue.dispose();
              }
              runtime.executePendingJobs();
              break;
            }
            let abortListener = null;
            if (signal) {
              abortListener = () => {
                const reason = signal.reason;
                const abortMessage = reason instanceof Error ? `${reason.name}: ${reason.message}` : reason ? String(reason) : "Execution was aborted";
                for (const { deferredPromise } of currentPromises) {
                  const errValue = vm.newString(abortMessage);
                  deferredPromise.reject(errValue);
                  errValue.dispose();
                }
                runtime.executePendingJobs();
              };
              signal.addEventListener("abort", abortListener);
            }
            try {
              await Promise.all(
                currentPromises.map(async ({ hostPromise, deferredPromise }) => {
                  if (signal == null ? void 0 : signal.aborted) {
                    return;
                  }
                  try {
                    const value = await hostPromise;
                    if (signal == null ? void 0 : signal.aborted) {
                      return;
                    }
                    const vmValue = toVmValue(value);
                    deferredPromise.resolve(vmValue);
                    disposeIfNeeded(vmValue);
                  } catch (err) {
                    if (signal == null ? void 0 : signal.aborted) {
                      return;
                    }
                    const serialized = err instanceof Error ? err.message : String(err);
                    const createErrorResult = vm.evalCode(`new Error(${JSON.stringify(serialized)})`);
                    if ("error" in createErrorResult) {
                      const errValue = vm.newString(serialized);
                      deferredPromise.reject(errValue);
                      errValue.dispose();
                    } else {
                      const errorHandle = createErrorResult.value;
                      deferredPromise.reject(errorHandle);
                      errorHandle.dispose();
                    }
                  }
                })
              );
            } finally {
              if (signal && abortListener) {
                signal.removeEventListener("abort", abortListener);
              }
            }
            runtime.executePendingJobs();
            if (signal == null ? void 0 : signal.aborted) {
              break;
            }
          }
          if (!hasJobs && pendingPromises.length === 0) {
            break;
          }
          iteration++;
        }
        if (iteration >= maxIterations) {
          throw new Error("Maximum event loop iterations exceeded");
        }
        const errorResult = vm.evalCode("globalThis.__llmz_error");
        const errorValue = vm.dump(errorResult.unwrap());
        errorResult.unwrap().dispose();
        if (signal == null ? void 0 : signal.aborted) {
          const reason = signal.reason;
          if (reason instanceof Error) {
            throw reason;
          }
          throw new Error(reason ? String(reason) : "Execution was aborted");
        }
        if (errorValue !== null && errorValue !== "") {
          try {
            copyBackContextFromVM();
          } catch {
          }
          const errorStackResult = vm.evalCode("globalThis.__llmz_error_stack");
          const errorStack = vm.dump(errorStackResult.unwrap()) || "";
          errorStackResult.unwrap().dispose();
          const deserializedError = Signals.maybeDeserializeError(errorValue);
          if (deserializedError instanceof VMSignal) {
            deserializedError.stack = errorStack;
            throw deserializedError;
          }
          const error = new Error(errorValue);
          error.stack = errorStack;
          throw error;
        }
        copyBackContextFromVM();
        const resultSetResult = vm.evalCode("globalThis.__llmz_result_set");
        const resultSet = vm.dump(resultSetResult.unwrap());
        resultSetResult.unwrap().dispose();
        let returnValue = void 0;
        if (resultSet) {
          const resultResult = vm.evalCode("globalThis.__llmz_result");
          returnValue = vm.dump(resultResult.unwrap());
          resultResult.unwrap().dispose();
        }
        returnValue = Signals.maybeDeserializeError(returnValue);
        return {
          success: true,
          variables: mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter),
          signal: returnValue instanceof VMSignal ? returnValue : void 0,
          lines_executed: Array.from(lines_executed),
          return_value: returnValue
        };
      } catch (err) {
        if (signal == null ? void 0 : signal.aborted) {
          const reason = signal.reason;
          const abortError = reason instanceof Error ? reason : new Error(reason ? String(reason) : "Execution was aborted");
          return handleErrorQuickJS(
            abortError,
            code,
            consumer,
            traces,
            variables,
            lines_executed,
            userCodeStartLine,
            currentToolCall
          );
        }
        await Promise.all(
          pendingPromises.map(async ({ hostPromise, deferredPromise }) => {
            try {
              const value = await hostPromise;
              const vmValue = toVmValue(value);
              deferredPromise.resolve(vmValue);
              disposeIfNeeded(vmValue);
            } catch (err2) {
              const serialized = err2 instanceof Error ? err2.message : String(err2);
              const errValue = vm.newString(serialized);
              deferredPromise.reject(errValue);
              errValue.dispose();
            }
          })
        ).catch(() => {
        });
        return handleErrorQuickJS(
          err,
          code,
          consumer,
          traces,
          variables,
          lines_executed,
          userCodeStartLine,
          currentToolCall
        );
      } finally {
        try {
          vm.dispose();
        } catch {
        }
        try {
          runtime.dispose();
        } catch {
        }
      }
    } catch (quickjsError) {
      const debugInfo = {
        error: (quickjsError == null ? void 0 : quickjsError.message) || String(quickjsError),
        errorStack: quickjsError == null ? void 0 : quickjsError.stack,
        wasmSource: BundledReleaseSyncVariant._wasmSource,
        wasmLoadedSuccessfully: BundledReleaseSyncVariant._wasmLoadedSuccessfully,
        wasmSize: BundledReleaseSyncVariant._wasmSize,
        wasmLoadError: BundledReleaseSyncVariant._wasmLoadError,
        nodeVersion: typeof process !== "undefined" && process.version ? process.version : "undefined",
        platform: typeof process !== "undefined" && process.platform ? process.platform : "undefined"
      };
      console.warn("QuickJS failed to load, falling back to node driver.");
      console.warn("Error:", (quickjsError == null ? void 0 : quickjsError.message) || quickjsError);
      console.warn("Debug info:", JSON.stringify(debugInfo, null, 2));
      DRIVER = "node";
    }
  }
  if (DRIVER === "node") {
    context[Identifiers.CommentFnIdentifier] = (comment, line) => traces.push({ type: "comment", comment, line, started_at: Date.now() });
    context[Identifiers.LineTrackingFnIdentifier] = (line) => {
      lines_executed.set(line, (lines_executed.get(line) ?? 0) + 1);
    };
    context[Identifiers.JSXFnIdentifier] = (tool, props, ...children) => createJsxComponent({
      type: tool,
      props,
      children
    });
    context[Identifiers.VariableTrackingFnIdentifier] = (name, getter) => {
      if (NO_TRACKING.includes(name)) {
        return;
      }
      variables[name] = () => {
        try {
          const value = getter();
          if (typeof value === "function") {
            return "[[non-primitive]]";
          }
          return value;
        } catch {
          return "[[non-primitive]]";
        }
      };
    };
    let currentToolCall;
    context[Identifiers.ToolCallTrackerFnIdentifier] = (callId, type, outputOrError) => {
      var _a2;
      const temp = Signals.maybeDeserializeError(outputOrError == null ? void 0 : outputOrError.message);
      if (type === "end" && temp instanceof SnapshotSignal && (temp == null ? void 0 : temp.toolCall)) {
        currentToolCall = {
          ...temp.toolCall,
          assignment: (_a2 = transformed.toolCalls.get(callId)) == null ? void 0 : _a2.assignment
        };
      }
    };
    context[Identifiers.ConsoleObjIdentifier] = {
      log: (...args) => {
        const message = args.shift();
        traces.push({ type: "log", message, args, started_at: Date.now() });
      }
    };
    context[Identifiers.AsyncIterYieldFnIdentifier] = async function(value) {
      const startedAt = Date.now();
      try {
        if (typeof value.type !== "string" || value.type.trim().length === 0) {
          throw new Error("A yield statement must yield a valid tool");
        }
        const toolName = Object.keys(context).find((x) => x.toUpperCase() === value.type.toUpperCase());
        if (!toolName) {
          throw new Error(`Yield tool "${value.type}", but tool is not found`);
        }
        await context[toolName](value);
      } finally {
        traces.push({ type: "yield", value, started_at: startedAt, ended_at: Date.now() });
      }
    };
    const AsyncFunction = async function* () {
    }.constructor;
    return await (async () => {
      const descriptors = Object.getOwnPropertyDescriptors(context);
      const topLevelProperties = Object.keys(descriptors).filter(
        (x) => !NO_TRACKING.includes(x) && descriptors[x] && typeof descriptors[x].value !== "function" && typeof descriptors[x].value !== "object"
      );
      const __report = (name, value) => {
        if (context[name] !== void 0 && context[name] !== value) {
          context[name] = value;
        }
      };
      context.__report = __report;
      const reportAll = topLevelProperties.map((x) => `__report("${x}", ${x})`).join(";");
      const assigner = `let __${Identifiers.LineTrackingFnIdentifier} = ${Identifiers.LineTrackingFnIdentifier}; ${Identifiers.LineTrackingFnIdentifier} = function(line) { ${reportAll}; __${Identifiers.LineTrackingFnIdentifier}(line);}`;
      const wrapper = `"use strict"; try { ${assigner};${transformed.code} } finally { ${reportAll} };`;
      const fn = AsyncFunction(...Object.keys(context), wrapper);
      const res = fn(...Object.values(context));
      do {
        const { value, done } = await res.next();
        if (done) {
          return value;
        }
        await context[Identifiers.AsyncIterYieldFnIdentifier](value);
      } while (true);
    })().then((res) => {
      res = Signals.maybeDeserializeError(res);
      return {
        success: true,
        variables: mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter),
        signal: res instanceof VMSignal ? res : void 0,
        lines_executed: Array.from(lines_executed),
        return_value: res
      };
    }).catch((err) => handleErrorNode(err, code, consumer, traces, variables, lines_executed, currentToolCall)).catch((err) => handleCatch(err, traces, variables, lines_executed));
  }
  throw new Error(`Unknown driver: ${DRIVER}`);
}
var handleErrorQuickJS = (err, code, _consumer, traces, variables, lines_executed, userCodeStartLine, currentToolCall) => {
  var _a, _b, _c;
  err = Signals.maybeDeserializeError(err);
  const lines = code.split("\n");
  const stackTrace = err.stack || "";
  const LINE_OFFSET = 1;
  const regex = /<quickjs>:(\d+)/g;
  const QUICKJS_WRAPPER_OFFSET = 10;
  const matches = Array.from(stackTrace.matchAll(regex)).map((x) => {
    const quickjsLine = Number(x[1]);
    const transformedCodeLine = quickjsLine - QUICKJS_WRAPPER_OFFSET;
    const line = Math.max(1, transformedCodeLine - userCodeStartLine + 1);
    const actualLine = lines[line - LINE_OFFSET] ?? "";
    const whiteSpacesCount = actualLine.length - actualLine.trimStart().length;
    const minColumn = whiteSpacesCount;
    return {
      line,
      column: minColumn
    };
  });
  if (matches.length === 0 && lines_executed.size > 0) {
    const lastLine2 = Math.max(...Array.from(lines_executed.keys()));
    const actualLine = lines[lastLine2 - LINE_OFFSET] ?? "";
    const whiteSpacesCount = actualLine.length - actualLine.trimStart().length;
    matches.push({
      line: lastLine2,
      column: whiteSpacesCount
    });
  }
  const lastLine = ((_a = maxBy_default(matches, (x) => x.line ?? 0)) == null ? void 0 : _a.line) ?? 0;
  let debugUserCode = "";
  let truncatedCode = "";
  let truncated = false;
  const appendCode = (line) => {
    debugUserCode += line;
    if (!truncated) {
      truncatedCode += line;
    }
  };
  for (let i = 0; i < lines.length; i++) {
    const VM_OFFSET = 2;
    const DISPLAY_OFFSET = 0;
    const line = lines[i];
    const correctedStackLineIndex = i + LINE_OFFSET + VM_OFFSET;
    const match = matches.find((x) => x.line + VM_OFFSET === correctedStackLineIndex);
    const paddedLineNumber = String(correctedStackLineIndex - VM_OFFSET - DISPLAY_OFFSET).padStart(3, "0");
    if (match) {
      appendCode(`> ${paddedLineNumber} | ${line}
`);
      appendCode(`    ${" ".repeat(paddedLineNumber.length + match.column)}^^^^^^^^^^
`);
      if (match.line >= lastLine) {
        truncated = true;
      }
    } else {
      appendCode(`  ${paddedLineNumber} | ${line}
`);
    }
  }
  debugUserCode = cleanStackTrace(debugUserCode).trim();
  truncatedCode = cleanStackTrace(truncatedCode).trim();
  if (err instanceof VMSignal) {
    const signalError = err;
    signalError.stack = debugUserCode;
    signalError.truncatedCode = truncatedCode;
    signalError.variables = mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter);
    signalError.toolCall = currentToolCall;
    return {
      success: true,
      variables: mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter),
      signal: err,
      lines_executed: Array.from(lines_executed)
    };
  } else {
    traces.push({
      type: "code_execution_exception",
      position: [((_b = matches[0]) == null ? void 0 : _b.line) ?? 0, ((_c = matches[0]) == null ? void 0 : _c.column) ?? 0],
      message: err.message,
      stackTrace: debugUserCode,
      started_at: Date.now()
    });
    const codeError = new CodeExecutionError(err.message, code, debugUserCode);
    const deserializedError = Signals.maybeDeserializeError(codeError);
    return {
      success: false,
      variables: mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter),
      error: deserializedError,
      traces,
      lines_executed: Array.from(lines_executed)
    };
  }
};
var handleErrorNode = (err, code, consumer, traces, variables, _lines_executed, currentToolCall) => {
  var _a, _b, _c;
  err = Signals.maybeDeserializeError(err);
  const lines = code.split("\n");
  const stackTrace = err.stack || "";
  const LINE_OFFSET = 1;
  const regex = /<anonymous>:(\d+):(\d+)/g;
  const matches = [...stackTrace.matchAll(regex)].map((x) => {
    const originalLine = consumer.originalPositionFor({
      line: Number(x[1]),
      column: Number(x[2])
    });
    const line = originalLine.line ?? Number(x[1]);
    const actualLine = lines[line - LINE_OFFSET] ?? "";
    const whiteSpacesCount = actualLine.length - actualLine.trimStart().length;
    const minColumn = Math.max(whiteSpacesCount, originalLine.column);
    return {
      line,
      column: Math.min(minColumn, Number(x[2]))
    };
  });
  const lastLine = ((_a = maxBy_default(matches, (x) => x.line ?? 0)) == null ? void 0 : _a.line) ?? 0;
  let debugUserCode = "";
  let truncatedCode = "";
  let truncated = false;
  const appendCode = (line) => {
    debugUserCode += line;
    if (!truncated) {
      truncatedCode += line;
    }
  };
  for (let i = 0; i < lines.length; i++) {
    const VM_OFFSET = 2;
    const DISPLAY_OFFSET = 0;
    const line = lines[i];
    const correctedStackLineIndex = i + LINE_OFFSET + VM_OFFSET;
    const match = matches.find((x) => x.line + VM_OFFSET === correctedStackLineIndex);
    const paddedLineNumber = String(correctedStackLineIndex - VM_OFFSET - DISPLAY_OFFSET).padStart(3, "0");
    if (match) {
      appendCode(`> ${paddedLineNumber} | ${line}
`);
      appendCode(`    ${" ".repeat(paddedLineNumber.length + match.column)}^^^^^^^^^^
`);
      if (match.line >= lastLine) {
        truncated = true;
      }
    } else {
      appendCode(`  ${paddedLineNumber} | ${line}
`);
    }
  }
  debugUserCode = cleanStackTrace(debugUserCode).trim();
  truncatedCode = cleanStackTrace(truncatedCode).trim();
  if (err instanceof VMSignal) {
    err.stack = debugUserCode;
    err.truncatedCode = truncatedCode;
    err.variables = mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter);
    err.toolCall = currentToolCall;
    throw err;
  } else {
    traces.push({
      type: "code_execution_exception",
      position: [((_b = matches[0]) == null ? void 0 : _b.line) ?? 0, ((_c = matches[0]) == null ? void 0 : _c.column) ?? 0],
      message: err.message,
      stackTrace: debugUserCode,
      started_at: Date.now()
    });
    throw new CodeExecutionError(err.message, code, debugUserCode);
  }
};
var handleCatch = (err, traces, variables, lines_executed) => {
  err = Signals.maybeDeserializeError(err);
  return {
    success: err instanceof VMSignal ? true : false,
    variables: mapValues_default(variables, (getter) => isFunction_default(getter) ? getter() : getter),
    error: err,
    signal: err instanceof VMSignal ? err : void 0,
    traces,
    lines_executed: Array.from(lines_executed)
  };
};

export {
  runAsyncFunction
};
