// llmz/dist/chunk-EKXZXQJB.js
import {
  camelCase_default,
  deburr_default,
  isPlainObject_default,
  pickBy_default,
  startCase_default
} from "./chunk-5XC6ZMS5.js";

// src/utils.ts
import { getWasmTokenizer } from "@bpinternal/thicktoken";
import { z, transforms } from "@bpinternal/zui";
var tokenizer = null;
var init = async () => {
  if (tokenizer) {
    return;
  }
  while (typeof getWasmTokenizer !== "function") {
    await new Promise((resolve) => setTimeout(resolve, 50));
  }
  tokenizer = await getWasmTokenizer();
};
var getTokenizer = () => {
  if (!tokenizer) {
    throw new Error("Tokenizer not initialized, make sure to call init() first and await it");
  }
  return tokenizer;
};
var Tokens = (min, max) => {
  return z.string().superRefine((value, ctx) => {
    const tokens = tokenizer.count(value);
    if (value.length < min) {
      ctx.addIssue({
        code: "too_small",
        minimum: min,
        type: "string",
        message: `Text is ${tokens} but expected at least ${min} tokens`,
        inclusive: true
      });
    }
    if (value.length > max) {
      ctx.addIssue({
        code: "too_big",
        maximum: max,
        type: "string",
        message: `Text is ${tokens} tokens but expected at most ${max} tokens`,
        inclusive: true
      });
    }
    return value;
  });
};
var Identifier = z.string().min(1).max(100).regex(
  /^([a-zA-Z_][a-zA-Z0-9_]*)$/,
  "Identifiers must start with a letter and can only contain letters, numbers and underscores"
);
function escapeString(str) {
  if (typeof str !== "string") {
    return "";
  }
  const rawStr = String.raw`${str}`;
  const escapedStr = rawStr.replace(/\\/g, "\\\\").replace(/\n/g, "\\n").replace(/"/g, '\\"').replace(/'/g, "\\'");
  if (escapedStr.includes("`")) {
    return `"${escapedStr}"`;
  } else if (escapedStr.includes("'")) {
    return `"${escapedStr}"`;
  } else {
    return `'${escapedStr}'`;
  }
}
var toPropertyKey = (key) => {
  if (/^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key)) {
    return key;
  }
  return escapeString(key);
};
var trimEmptyLines = (lines) => {
  var _a;
  while (lines.length && !((_a = lines[0]) == null ? void 0 : _a.trim())) {
    lines.shift();
  }
  while (lines.length && !lines[lines.length - 1].trim()) {
    lines.pop();
  }
};
var getMultilineComment = (description) => {
  description = description == null ? void 0 : description.replace(/(\n(\s*)?){3,}/g, "\n\n");
  const ensureLineStartsWithAsterisk = (line) => line.startsWith("* ") ? ` ${line}` : ` * ${line}`;
  const escapeCommentEnd = (line) => line.replace(/\*\//g, "*\\/");
  const descLines = (description ?? "").split("\n").map((line) => line.trim());
  trimEmptyLines(descLines);
  if (descLines.length) {
    descLines[0] = descLines[0].replace(/^\/\*\*?/, "");
    descLines[descLines.length - 1] = descLines[descLines.length - 1].replace(/\*\/$/, "");
  }
  trimEmptyLines(descLines);
  return descLines.length === 0 ? "" : descLines.length === 1 ? `/** ${escapeCommentEnd(descLines[0])} */` : `/**
${descLines.map(ensureLineStartsWithAsterisk).map(escapeCommentEnd).join("\n")}
 */`;
};
var toValidFunctionName = (str) => {
  let name = deburr_default(str);
  name = name.replace(/[^a-zA-Z0-9_$]/g, "");
  if (!/^[a-zA-Z_$]/.test(name)) {
    name = `_${name}`;
  }
  return camelCase_default(name);
};
var awaitObject = async (obj) => {
  const newObj = { ...obj };
  const promises = [];
  const setProp = async (key, value) => {
    newObj[key] = await value;
  };
  for (const [key, value] of Object.entries(obj)) {
    if (value instanceof Promise) {
      promises.push(setProp(key, value));
    } else if (isPlainObject_default(value)) {
      promises.push(setProp(key, await awaitObject(value)));
    }
  }
  await Promise.all(promises);
  return newObj;
};
var toValidObjectName = (str) => {
  let name = deburr_default(str);
  name = name.split(/[^a-zA-Z0-9_$]/g).map(startCase_default).join("");
  if (!/^[a-zA-Z_$]/.test(name)) {
    name = `_${name}`;
  }
  return name.replaceAll(" ", "");
};
var stripInvalidIdentifiers = (object) => {
  if (typeof object !== "object") {
    return object;
  }
  if (Array.isArray(object)) {
    return object.map(stripInvalidIdentifiers);
  }
  return pickBy_default(object, (__, key) => Identifier.safeParse(key).success);
};
var convertObjectToZuiLiterals = (obj, nested = false) => {
  if (typeof obj === "string") {
    return z.literal(obj).catch(() => obj);
  }
  if (typeof obj === "number") {
    return z.literal(obj).catch(() => obj);
  }
  if (typeof obj === "boolean") {
    return z.literal(obj).catch(() => obj);
  }
  if (Array.isArray(obj)) {
    if (obj.length === 0) {
      return z.tuple([]).catch(() => obj);
    }
    return z.tuple([
      // the tuple needs to have at least one element
      convertObjectToZuiLiterals(obj[0]),
      ...obj.slice(1).map((child) => convertObjectToZuiLiterals(child))
    ]).catch(() => obj);
  }
  if (obj !== null && typeof obj === "object") {
    const shape = {};
    for (const key in obj) {
      shape[key] = convertObjectToZuiLiterals(obj[key], true);
    }
    if (nested) {
      return z.object(shape).catch(() => shape);
    }
    if (Object.keys(shape).length === 0) {
      return z.object({}).catch(() => shape);
    }
    return shape;
  }
  if (obj === void 0) {
    return z.undefined().catch(() => void 0);
  }
  if (obj === null) {
    return z.null().catch(() => null);
  }
  throw new Error(`Unsupported object type ${typeof obj}, ${obj})`);
};
var isValidSchema = (schema) => {
  try {
    transforms.fromJSONSchemaLegacy(schema);
    return typeof schema.type === "string";
  } catch {
    return false;
  }
};
function isZuiSchema(schema) {
  return !!schema && typeof schema === "object" && schema.__type__ === "ZuiType";
}
function isJsonSchema(schema) {
  return !!schema && typeof schema === "object" && ("$schema" in schema || "type" in schema || "properties" in schema);
}
function isValidIdentifier(name) {
  if (typeof name !== "string") {
    return false;
  }
  return /^[A-Z]{1,}[A-Z0-9_]{0,50}$/i.test(name);
}
function isValidMessageName(name) {
  if (typeof name !== "string") {
    return false;
  }
  return /^[A-Z]{1,50}$/i.test(name);
}

export {
  init,
  getTokenizer,
  Tokens,
  Identifier,
  escapeString,
  toPropertyKey,
  getMultilineComment,
  toValidFunctionName,
  awaitObject,
  toValidObjectName,
  stripInvalidIdentifiers,
  convertObjectToZuiLiterals,
  isValidSchema,
  isZuiSchema,
  isJsonSchema,
  isValidIdentifier,
  isValidMessageName
};
