// llmz/dist/chunk-PK72FAKD.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } var _class;

var _chunkKIN7Y247cjs = require('./chunk-KIN7Y247.cjs');

// src/errors.ts
var errorClasses = {};
function registerErrorClass(name, errorClass) {
  errorClasses[name] = errorClass;
}
var tryParseMessage = (str) => {
  try {
    return JSON.parse(str);
  } catch (e) {
    return str;
  }
};
var Signals;
((Signals2) => {
  function isWrappedError(error) {
    var _a, _b;
    const isAlreadyWrapped = error.name === "Error" && ((_a = error == null ? void 0 : error.message) == null ? void 0 : _a.startsWith("{")) && ((_b = error == null ? void 0 : error.message) == null ? void 0 : _b.endsWith("}"));
    return isAlreadyWrapped;
  }
  Signals2.isWrappedError = isWrappedError;
  function serializeError(error) {
    if (isWrappedError(error)) {
      const msg = tryParseMessage(error.message);
      return JSON.stringify({
        ...typeof msg === "object" ? msg : { message: msg },
        properties: { ...error }
      });
    }
    return JSON.stringify({
      name: error.constructor.name,
      message: error.message,
      stack: _chunkKIN7Y247cjs.cleanStackTrace.call(void 0, _nullishCoalesce(error.stack, () => ( ""))),
      properties: { ...error }
    });
  }
  Signals2.serializeError = serializeError;
  function maybeDeserializeError(error) {
    const errorIsAlreadyDeserialized = error instanceof Error && error.name in errorClasses;
    if (errorIsAlreadyDeserialized) {
      return error;
    }
    const serializedError = error instanceof Error ? error.message : typeof error === "string" ? error : _nullishCoalesce((error == null ? void 0 : error.toString()), () => ( ""));
    try {
      const parsed = JSON.parse(serializedError);
      if (parsed && parsed.name && parsed.message) {
        const { name, message, properties = {} } = parsed;
        const ErrorClass = errorClasses[name] || Error;
        const errorInstance = new ErrorClass(message);
        errorInstance.message = message;
        errorInstance.name = name;
        errorInstance.stack = _chunkKIN7Y247cjs.cleanStackTrace.call(void 0, _nullishCoalesce((error == null ? void 0 : error.stack), () => ( "")));
        Object.assign(errorInstance, properties);
        if (isWrappedError(errorInstance)) {
          return maybeDeserializeError(errorInstance);
        }
        return errorInstance;
      }
    } catch (e2) {
    }
    return error;
  }
  Signals2.maybeDeserializeError = maybeDeserializeError;
})(Signals || (Signals = exports.Signals = {}));
var VMSignal = (_class = class extends Error {
  constructor(message) {
    super(message);_class.prototype.__init.call(this);_class.prototype.__init2.call(this);;
    this.message = message;
    this.message = Signals.serializeError(this);
  }
  /**
   * The code that was executed by the VM up to the point of the signal
   */
  __init() {this.truncatedCode = ""}
  /** The current tool call, if any */
  
  /**
   * Contains all the declared and executed variables during the VM execution
   * See file plugins/variable-extraction.ts for more details
   */
  __init2() {this.variables = {}}
}, _class);
var SnapshotSignal = class extends VMSignal {
  constructor(message) {
    super(message);
    this.message = Signals.serializeError(this);
  }
};
var VMLoopSignal = class extends VMSignal {
  constructor(message) {
    super(message);
    this.message = Signals.serializeError(this);
  }
};
var ThinkSignal = class extends VMLoopSignal {
  constructor(reason, context, metadata) {
    super("Think signal received: " + reason);
    this.reason = reason;
    this.context = context;
    this.metadata = metadata;
    this.message = Signals.serializeError(this);
    this.stack = "";
  }
  toString() {
    return Signals.serializeError(this);
  }
};
var CodeExecutionError = class extends Error {
  constructor(message, code, stacktrace) {
    super(message);
    this.code = code;
    this.stacktrace = stacktrace;
    this.message = Signals.serializeError(this);
  }
};
var InvalidCodeError = class extends Error {
  constructor(message, code) {
    super(message);
    this.code = code;
    this.message = Signals.serializeError(this);
    this.stack = "";
  }
};
var LoopExceededError = class extends Error {
  constructor() {
    super("Loop exceeded error");
    this.message = Signals.serializeError(this);
    this.stack = "";
  }
};
var CodeFormattingError = class extends Error {
  constructor(message, code) {
    super(message, {
      cause: "Code formatting error"
    });
    this.code = code;
    this.stack = "";
  }
};
var AssignmentError = class extends Error {
  constructor(message) {
    super(message);
    this.message = Signals.serializeError(this);
  }
};
var CognitiveError = class extends Error {
  constructor(message) {
    super(message);
    this.message = Signals.serializeError(this);
    this.stack = "";
  }
};
registerErrorClass("VMSignal", VMSignal);
registerErrorClass("SnapshotSignal", SnapshotSignal);
registerErrorClass("VMLoopSignal", VMLoopSignal);
registerErrorClass("ThinkSignal", ThinkSignal);
registerErrorClass("CodeExecutionError", CodeExecutionError);
registerErrorClass("AssignmentError", AssignmentError);
registerErrorClass("CognitiveError", CognitiveError);












exports.Signals = Signals; exports.VMSignal = VMSignal; exports.SnapshotSignal = SnapshotSignal; exports.ThinkSignal = ThinkSignal; exports.CodeExecutionError = CodeExecutionError; exports.InvalidCodeError = InvalidCodeError; exports.LoopExceededError = LoopExceededError; exports.CodeFormattingError = CodeFormattingError; exports.AssignmentError = AssignmentError; exports.CognitiveError = CognitiveError;
