// llmz/dist/chunk-54G63644.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _chunkRKKRSBX3cjs = require('./chunk-RKKRSBX3.cjs');

// src/jsx.ts
var _zui = require('@bpinternal/zui');
function isAnyJsxComponent(component) {
  return typeof component === "object" && component !== null && "__jsx" in component && component.__jsx === true && typeof component.type === "string";
}
function isJsxComponent(type, component) {
  return typeof component === "object" && component !== null && "__jsx" in component && component.__jsx === true && typeof component.type === "string" && typeof type === "string" && component.type.toUpperCase().trim() === type.toUpperCase().trim();
}
var createJsxComponent = (props) => ({
  __jsx: true,
  type: _chunkRKKRSBX3cjs.isString_default.call(void 0, props.type) ? props.type.toUpperCase() : "__unknown__",
  children: Array.isArray(props.children) ? props.children : props.children ? [props.children] : [],
  props: _chunkRKKRSBX3cjs.isPlainObject_default.call(void 0, props.props) ? props.props : {}
});
var Jsx = _zui.z.custom(
  (value) => isAnyJsxComponent(value),
  (value) => ({
    params: { value },
    message: `Invalid JSX component: ${value}`
  })
);






exports.isAnyJsxComponent = isAnyJsxComponent; exports.isJsxComponent = isJsxComponent; exports.createJsxComponent = createJsxComponent; exports.Jsx = Jsx;
