// llmz/dist/index.js
import {
  DefaultExit,
  ErrorExecutionResult,
  ExecutionResult,
  ListenExit,
  PartialExecutionResult,
  Snapshot,
  SuccessExecutionResult,
  ThinkExit,
  getValue,
  parseExit
} from "./chunk-E6WEQRZF.js";
import "./chunk-6EAURSDP.js";
import {
  Tool
} from "./chunk-G5INQ5LJ.js";
import {
  formatTypings,
  getTypings
} from "./chunk-LR6KTFHO.js";
import {
  CodeFormattingError,
  LoopExceededError,
  SnapshotSignal,
  ThinkSignal
} from "./chunk-WSVDMGMR.js";
import "./chunk-YEAWWJSJ.js";
import {
  Exit
} from "./chunk-E5YW5YQN.js";
import {
  Component,
  assertValidComponent,
  isAnyComponent,
  isComponent,
  renderToTsx
} from "./chunk-AMQFJRRX.js";
import "./chunk-UOPTKXLE.js";
import {
  truncateWrappedContent,
  wrapContent
} from "./chunk-M4WTTARE.js";
import {
  escapeString,
  getMultilineComment,
  isValidIdentifier,
  toValidFunctionName,
  toValidObjectName
} from "./chunk-EKXZXQJB.js";
import {
  cloneDeep_default,
  upperFirst_default
} from "./chunk-5XC6ZMS5.js";

// package.json
var version = "0.0.74";

// src/objects.ts
import { z } from "@bpinternal/zui";

// src/hoist.ts
import { CodeGenerator } from "@babel/generator";
import { parse } from "@babel/parser";
import Traverse from "@babel/traverse";
import * as t from "@babel/types";
var traverse = Traverse;
function getTypingsWithoutComments(type) {
  const typeClone = cloneDeep_default(type);
  removeBabelComments(typeClone);
  return new CodeGenerator(typeClone, { comments: false }).generate().code;
}
function removeBabelComments(node) {
  node.leadingComments = null;
  node.trailingComments = null;
  node.innerComments = null;
  Object.keys(node).forEach((key) => {
    const val = node[key];
    if (Array.isArray(val)) {
      val.forEach((child) => child && typeof child === "object" && removeBabelComments(child));
    } else if (val && typeof val === "object" && val.type) {
      removeBabelComments(val);
    }
  });
}
function extractAndHoistTypes(ast) {
  const typeMap = /* @__PURE__ */ new Map();
  const typeCounts = /* @__PURE__ */ new Map();
  const hoistedTypes = [];
  function addTypeToMap(typeNode) {
    const typeString = getTypingsWithoutComments(typeNode.typeAnnotation);
    if (!typeMap.has(typeString)) {
      typeMap.set(typeString, typeNode);
    }
  }
  function generateUniqueTypeName(baseName) {
    let typeName = baseName;
    let counter = 1;
    while ([...typeMap.values()].some((typeNode) => typeNode.id.name === typeName)) {
      typeName = `${baseName}${counter++}`;
    }
    return typeName;
  }
  function createTypeAlias(name, type) {
    return t.tsTypeAliasDeclaration(t.identifier(name), null, type);
  }
  function getTypePaths(path) {
    var _a, _b, _c;
    let currPath = path;
    const parts = /* @__PURE__ */ new Set();
    while (currPath) {
      const { node, parentPath } = currPath;
      if (t.isIdentifier(node) && node.name) {
        parts.add(node.name);
      }
      if (t.isTSMethodSignature(node) && currPath.key === "parameters") {
        parts.add("Input");
      }
      if (currPath.key === "returnType") {
        parts.add("Output");
        const methodName = "id" in parentPath.node && "name" in parentPath.node.id && ((_b = (_a = parentPath == null ? void 0 : parentPath.node) == null ? void 0 : _a.id) == null ? void 0 : _b.name);
        if (methodName) {
          parts.add(methodName);
        }
      }
      if (((_c = node == null ? void 0 : node.key) == null ? void 0 : _c.type) === "Identifier") {
        parts.add(node.key.name);
      }
      if (t.isTSParameterProperty(node) || t.isTSFunctionType(node)) {
        parts.add("Input");
      }
      if (t.isTSFunctionType(node)) {
        parts.add("Output");
      }
      currPath = parentPath;
    }
    return parts.size ? Array.from(parts).reverse().map(upperFirst_default) : ["UnnamedType"];
  }
  function generateTypeName(path) {
    return getTypePaths(path).join("");
  }
  function findNestedTypes(path) {
    const node = path.node;
    if (t.isTSTypeLiteral(node)) {
      const typeString = getTypingsWithoutComments(node);
      typeCounts.set(typeString, (typeCounts.get(typeString) || 0) + 1);
      if (typeMap.has(typeString) && (typeCounts.get(typeString) || 0) > 1) {
        path.replaceWith(t.tsTypeReference(typeMap.get(typeString).id));
      } else if (!typeMap.has(typeString) && (typeCounts.get(typeString) || 0) > 1) {
        const typeName = generateUniqueTypeName(generateTypeName(path));
        const typeAlias = createTypeAlias(typeName, node);
        addTypeToMap(typeAlias);
        path.replaceWith(t.tsTypeReference(t.identifier(typeName)));
      }
      const members = path.get("members");
      const memberPaths = Array.isArray(members) ? members : [members];
      memberPaths.forEach((memberPath) => {
        findNestedTypes(memberPath);
      });
    }
    if (t.isTSPropertySignature(node) && path.get("typeAnnotation")) {
      const annPath = path.get("typeAnnotation.typeAnnotation");
      if (annPath == null ? void 0 : annPath.node) {
        findNestedTypes(annPath);
      }
    }
    if (t.isTSArrayType(node)) {
      findNestedTypes(path.get("elementType"));
    }
    if (t.isTSUnionType(node) || t.isTSIntersectionType(node)) {
      ;
      path.get("types").forEach((p) => findNestedTypes(p));
    }
    if (t.isTSTypeReference(node) && node.typeParameters) {
      ;
      path.get("typeParameters.params").forEach((param) => findNestedTypes(param));
    }
    if (t.isTSFunctionType(node)) {
      const returnPath = path.get("typeAnnotation");
      if (returnPath == null ? void 0 : returnPath.node)
        findNestedTypes(returnPath);
      const params = path.get("parameters");
      params.forEach((paramPath) => {
        const ta = paramPath.get("typeAnnotation.typeAnnotation");
        if (ta == null ? void 0 : ta.node)
          findNestedTypes(ta);
      });
    }
    if (t.isTSMethodSignature(node)) {
      const returnPath = path.get("typeAnnotation");
      if (returnPath == null ? void 0 : returnPath.node) {
        findNestedTypes(returnPath);
      }
      const params = path.get("parameters");
      params.forEach((paramPath) => {
        const ta = paramPath.get("typeAnnotation.typeAnnotation");
        if (ta == null ? void 0 : ta.node)
          findNestedTypes(ta);
      });
    }
    if (t.isTSDeclareFunction(node)) {
      const params = path.get("params");
      params.forEach((p) => {
        const ta = p.get("typeAnnotation.typeAnnotation");
        if (ta == null ? void 0 : ta.node)
          findNestedTypes(ta);
      });
      const returnType = path.get("returnType.typeAnnotation");
      if (returnType == null ? void 0 : returnType.node)
        findNestedTypes(returnType);
    }
  }
  traverse(ast, {
    TSTypeAliasDeclaration(path) {
      const typeString = getTypingsWithoutComments(path.node.typeAnnotation);
      typeCounts.set(typeString, (typeCounts.get(typeString) || 0) + 1);
      if (typeMap.has(typeString) && (typeCounts.get(typeString) || 0) > 1) {
        const existing = typeMap.get(typeString);
        path.replaceWith(
          t.tsTypeAliasDeclaration(path.node.id, null, t.tsTypeReference(t.identifier(existing.id.name)))
        );
      } else {
        addTypeToMap(path.node);
      }
    },
    TSPropertySignature(path) {
      const annPath = path.get("typeAnnotation.typeAnnotation");
      if (annPath == null ? void 0 : annPath.node)
        findNestedTypes(annPath);
    },
    TSFunctionType(path) {
      const returnPath = path.get("typeAnnotation");
      if (returnPath.node)
        findNestedTypes(returnPath);
      const params = path.get("parameters");
      params.forEach((param) => {
        const ta = param.get("typeAnnotation.typeAnnotation");
        if (ta == null ? void 0 : ta.node)
          findNestedTypes(ta);
      });
    },
    TSMethodSignature(path) {
      const returnPath = path.get("typeAnnotation.typeAnnotation");
      if (returnPath == null ? void 0 : returnPath.node) {
        findNestedTypes(returnPath);
      }
      const params = path.get("parameters");
      params.forEach((param) => {
        const ta = param.get("typeAnnotation.typeAnnotation");
        if (ta == null ? void 0 : ta.node)
          findNestedTypes(ta);
      });
    },
    TSDeclareFunction(path) {
      const params = path.get("params");
      params.forEach((p) => {
        const ta = p.get("typeAnnotation.typeAnnotation");
        if (ta == null ? void 0 : ta.node)
          findNestedTypes(ta);
      });
      const returnPath = path.get("returnType.typeAnnotation");
      if (returnPath == null ? void 0 : returnPath.node)
        findNestedTypes(returnPath);
    }
  });
  typeMap.forEach((typeNode, typeString) => {
    if ((typeCounts.get(typeString) || 0) > 1) {
      if (!hoistedTypes.some((ht) => {
        return getTypingsWithoutComments(ht.typeAnnotation) === getTypingsWithoutComments(typeNode.typeAnnotation);
      })) {
        hoistedTypes.push(typeNode);
      }
    }
  });
  ast.program.body = [
    ...hoistedTypes,
    ...ast.program.body.filter(
      (node) => !hoistedTypes.some((ht) => t.isTSTypeAliasDeclaration(node) && node.id.name === ht.id.name)
    )
  ];
  return ast;
}
async function hoistTypings(code, formatOptions) {
  formatOptions ??= {};
  formatOptions.throwOnError ??= true;
  for (let i = 1; i <= 5; i++) {
    try {
      const initialCode = code;
      const ast = parse(code, {
        sourceType: "module",
        plugins: ["typescript"]
      });
      const transformedAst = extractAndHoistTypes(ast);
      code = new CodeGenerator(transformedAst, {
        // quotes: 'single',
        compact: false
      }).generate().code;
      if (initialCode === code) {
        break;
      }
    } catch (err) {
      console.error(err);
      if (formatOptions.throwOnError) {
        throw new CodeFormattingError(err instanceof Error ? err.message : String(err ?? "Unknown Error"), code);
      }
      break;
    }
  }
  return formatTypings(code, formatOptions);
}

// src/objects.ts
var LARGE_OBJECT_LINES_OF_CODE = 10;
var ObjectInstance = class {
  name;
  description;
  properties;
  tools;
  metadata;
  /**
   * Creates a new ObjectInstance.
   *
   * @param props - Object configuration
   * @param props.name - Unique object name (must be valid TypeScript identifier)
   * @param props.description - Human-readable description of the object
   * @param props.tools - Array of tools to group under this object namespace
   * @param props.properties - Array of stateful properties for this object
   * @param props.metadata - Additional metadata for the object
   *
   * @throws Error if name is not a valid identifier
   * @throws Error if description is not a string
   * @throws Error if metadata is not an object
   * @throws Error if properties/tools are not arrays
   * @throws Error if properties exceed 100 limit
   * @throws Error if property names are duplicated or invalid
   * @throws Error if property descriptions exceed 5000 characters
   *
   * @example
   * ```typescript
   * const userProfile = new ObjectInstance({
   *   name: 'user',
   *   description: 'User profile management',
   *   properties: [
   *     {
   *       name: 'name',
   *       value: 'John Doe',
   *       type: z.string().min(1),
   *       description: 'User full name',
   *       writable: true,
   *     },
   *     {
   *       name: 'email',
   *       value: null,
   *       type: z.string().email().nullable(),
   *       description: 'User email address',
   *       writable: true,
   *     },
   *   ],
   *   tools: [
   *     new Tool({
   *       name: 'updateProfile',
   *       input: z.object({ name: z.string(), email: z.string() }),
   *       handler: async ({ name, email }) => {
   *         // Update external system
   *         await updateUserInDatabase({ name, email })
   *       },
   *     }),
   *   ],
   *   metadata: {
   *     version: '1.0',
   *     category: 'user-management',
   *   },
   * })
   * ```
   */
  constructor(props) {
    var _a;
    if (!isValidIdentifier(props.name)) {
      throw new Error(
        `Invalid name for tool ${props.name}. A tool name must start with a letter and contain only letters, numbers, and underscores. It must be 1-50 characters long.`
      );
    }
    if (props.description !== void 0 && typeof props.description !== "string") {
      throw new Error(
        `Invalid description for tool ${props.name}. Expected a string, but got type "${typeof props.description}"`
      );
    }
    if (props.metadata !== void 0 && typeof props.metadata !== "object") {
      throw new Error(
        `Invalid metadata for tool ${props.name}. Expected an object, but got type "${typeof props.metadata}"`
      );
    }
    if (props.properties !== void 0 && !Array.isArray(props.properties)) {
      throw new Error(
        `Invalid properties for tool ${props.name}. Expected an array, but got type "${typeof props.properties}"`
      );
    }
    if (props.tools !== void 0 && !Array.isArray(props.tools)) {
      throw new Error(`Invalid tools for tool ${props.name}. Expected an array, but got type "${typeof props.tools}"`);
    }
    if ((_a = props.properties) == null ? void 0 : _a.length) {
      if (props.properties.length > 100) {
        throw new Error(
          `Too many properties for tool ${props.name}. Expected at most 100 properties, but got ${props.properties.length}`
        );
      }
      for (const prop of props.properties) {
        if (props.properties.filter((p) => p.name === prop.name).length > 1) {
          throw new Error(`Duplicate property name "${prop.name}" in tool ${props.name}`);
        }
        if (!isValidIdentifier(prop.name)) {
          throw new Error(
            `Invalid name for property ${prop.name}. A property name must start with a letter and contain only letters, numbers, and underscores. It must be 1-50 characters long.`
          );
        }
        if (prop.description !== void 0 && typeof prop.description !== "string") {
          throw new Error(
            `Invalid description for property ${prop.name}. Expected a string, but got type "${typeof prop.description}"`
          );
        }
        if (props.description && props.description.length >= 5e3) {
          throw new Error(
            `Description for property ${prop.name} is too long. Expected at most 5000 characters, but got ${props.description.length}`
          );
        }
        if (typeof prop.writable !== "boolean") {
          prop.writable = false;
        }
      }
    }
    this.name = props.name;
    this.description = props.description;
    this.metadata = props.metadata ?? {};
    this.properties = props.properties;
    this.tools = Tool.withUniqueNames(props.tools ?? []);
  }
  /**
   * Generates TypeScript namespace declarations for this object.
   *
   * This method creates TypeScript definitions that are included in the LLM context
   * to help it understand the available properties and tools. Properties become
   * const declarations with appropriate Readonly/Writable types, and tools become
   * function signatures.
   *
   * @returns Promise resolving to TypeScript namespace declaration
   *
   * @example
   * ```typescript
   * const obj = new ObjectInstance({
   *   name: 'user',
   *   properties: [
   *     { name: 'name', value: 'John', writable: true },
   *     { name: 'id', value: 123, writable: false },
   *   ],
   *   tools: [
   *     new Tool({
   *       name: 'save',
   *       input: z.object({ data: z.string() }),
   *       handler: async ({ data }) => { /* ... *\/ },
   *     }),
   *   ],
   * })
   *
   * const typings = await obj.getTypings()
   * console.log(typings)
   * // Output:
   * // export namespace user {
   * //   const name: Writable<string> = "John"
   * //   const id: Readonly<number> = 123
   * //   function save(args: { data: string }): Promise<void>
   * // }
   * ```
   */
  async getTypings() {
    return getObjectTypings(this).withProperties().withTools().build();
  }
  /**
   * Converts this ObjectInstance to its JSON representation.
   *
   * This method serializes the object into a JSON format that includes its name,
   * description, properties, tools, and metadata. It is used for serialization
   * and transmission of the object state.
   *
   * @returns JSON representation of the ObjectInstance
   */
  toJSON() {
    return {
      name: this.name,
      description: this.description,
      properties: this.properties,
      tools: (this.tools ?? []).map((tool) => tool.toJSON()),
      metadata: this.metadata
    };
  }
};
function getObjectTypings(obj) {
  let includeProperties = false;
  let includeTools = false;
  let hoisting = false;
  const typings = [];
  const addProperties = async () => {
    var _a;
    if (includeProperties && ((_a = obj.properties) == null ? void 0 : _a.length)) {
      typings.push("");
      typings.push("// ---------------- //");
      typings.push("//    Properties    //");
      typings.push("// ---------------- //");
      typings.push("");
      for (const prop of obj.properties ?? []) {
        const description = prop.description ?? "";
        if (description == null ? void 0 : description.trim().length) {
          typings.push(getMultilineComment(description));
        }
        let type = "unknown";
        if (prop.type) {
          type = await getTypings(prop.type, {});
        } else if (prop.value !== void 0) {
          type = typeof prop.value;
        }
        type = prop.writable ? `Writable<${type}>` : `Readonly<${type}>`;
        const value = embedPropertyValue(prop);
        typings.push(`const ${prop.name}: ${type} = ${value}`);
      }
    }
  };
  const addTools = async () => {
    var _a;
    if (includeTools && ((_a = obj.tools) == null ? void 0 : _a.length)) {
      typings.push("");
      typings.push("// ---------------- //");
      typings.push("//       Tools      //");
      typings.push("// ---------------- //");
      typings.push("");
      for (const tool of obj.tools) {
        const fnType = z.function(tool.zInput, tool.zOutput).title(tool.name).describe(tool.description ?? "");
        let temp = await getTypings(fnType, {
          declaration: true
        });
        temp = temp.replace("declare function ", "function ");
        typings.push(temp);
      }
    }
  };
  const finalize = async () => {
    var _a;
    let closingBracket = "";
    if (typings.length >= LARGE_OBJECT_LINES_OF_CODE) {
      closingBracket = ` // end namespace "${obj.name}"`;
    }
    let body = typings.join("\n");
    if (hoisting) {
      body = await hoistTypings(body, { throwOnError: false });
    }
    typings.push("}" + closingBracket);
    let header = "";
    if ((_a = obj.description) == null ? void 0 : _a.trim().length) {
      header = getMultilineComment(obj.description);
    }
    return formatTypings(
      `${header}
      export namespace ${obj.name} {
      ${body}
      } ${closingBracket}`.trim(),
      { throwOnError: false }
    );
  };
  const api = {
    withProperties: () => {
      includeProperties = true;
      return api;
    },
    withTools: () => {
      includeTools = true;
      return api;
    },
    withHoisting: () => {
      hoisting = true;
      return api;
    },
    async build() {
      await addProperties();
      await addTools();
      return finalize();
    }
  };
  return api;
}
function embedPropertyValue(property) {
  if (typeof property.value === "string") {
    return escapeString(property.value);
  }
  if (Number.isNaN(property.value)) {
    return "NaN";
  }
  if (typeof property.value === "number" && Number.isInteger(property.value)) {
    return property.value.toString();
  }
  if (typeof property.value === "boolean") {
    return property.value.toString();
  }
  if (Array.isArray(property.value) || typeof property.value === "object") {
    return JSON.stringify(property.value);
  }
  if (property.value instanceof Date) {
    return `new Date('${property.value.toISOString()}')`;
  }
  if (property.value instanceof RegExp) {
    return `new RegExp(${escapeString(property.value.source)}, ${escapeString(property.value.flags)})`;
  }
  if (property.value === null) {
    return "null";
  }
  if (property.value === void 0) {
    return "undefined";
  }
  if (typeof property.value === "function") {
    return "function() {}";
  }
  if (typeof property.value === "symbol") {
    return "Symbol()";
  }
  if (typeof property.value === "bigint") {
    return `${property.value}n`;
  }
  if (property.value instanceof Error) {
    return `Error(${escapeString(property.value.message)})`;
  }
  if (property.value instanceof Map) {
    return `new Map(${JSON.stringify(Array.from(property.value.entries()))})`;
  }
  if (property.value instanceof Set) {
    return `new Set(${JSON.stringify(Array.from(property.value.values()))})`;
  }
  return "unknown";
}

// src/citations.ts
var RARE_SYMBOLS = {
  ARROW_UP: "\u2191",
  CIRCLE_BULLET: "\u30FB",
  STAR_BULLET_FULL: "\u2605",
  STAR_BULLET_EMPTY: "\u2606",
  ARROW_BULLET: "\u2192",
  SQUARE_BULLET: "\u25A0",
  TRIANGLE_BULLET: "\u25BA",
  OPENING_TAG: "\u3010",
  CLOSING_TAG: "\u3011",
  SS: "\xA7",
  CROSS: "\u2020"
};
var CitationsManager = class {
  _citations = /* @__PURE__ */ new Map();
  _nextId = 0;
  /**
   * Registers a new source and returns the created Citation.
   * @param source An object representing the source of the citation.
   * @returns The created Citation.
   */
  registerSource(source) {
    const id = this._nextId++;
    const tag = `${RARE_SYMBOLS.OPENING_TAG}${id}${RARE_SYMBOLS.CLOSING_TAG}`;
    const citation = {
      id,
      source,
      tag
    };
    this._citations.set(id, citation);
    return citation;
  }
  /**
   * Extracts citations from the provided content and cleans the content of citation tags.
   * Non-found citations are replaced with a special "Not Found" citation.
   * @param content The string content containing citation tags.
   * @returns An object containing the cleaned content and an array of found citations.
   */
  extractCitations(content, replace) {
    const citations = [];
    const notFoundCitation = {
      id: -1,
      source: "Not Found",
      tag: ""
    };
    const regex = new RegExp(`${RARE_SYMBOLS.OPENING_TAG}([\\d|\\w|\\s|,]{0,})${RARE_SYMBOLS.CLOSING_TAG}`, "ig");
    let match;
    const offsets = [];
    while ((match = regex.exec(content)) !== null) {
      const offset = match.index;
      const length = match[0].length;
      offsets.push({ start: offset, length });
      const multi = (match[1] ?? "").split(/\D/g).map((s) => s.trim()).filter(Boolean).map((s) => parseInt(s, 10)).filter((s) => !isNaN(s) && s >= 0);
      for (const citationId of multi) {
        const citation = this._citations.get(citationId);
        if (citation) {
          citations.push({ ...citation, offset });
        } else {
          citations.push({ ...notFoundCitation, offset });
        }
      }
    }
    const entries = offsets.map((o) => ({
      start: o.start,
      length: o.length,
      citations: citations.filter((x) => x.offset === o.start && x.id !== -1)
    })).sort((a, b) => a.start - b.start);
    let result = "";
    let cursor = 0;
    for (const { start, length, citations: citations2 } of entries) {
      result += content.slice(cursor, start);
      const replacement = citations2.map((citation) => replace ? replace(citation) : "").join("");
      result += replacement;
      cursor = start + length;
    }
    result += content.slice(cursor);
    return { cleaned: result, citations };
  }
  /**
   * Strips citation tags from the provided content.
   * @param content The string content containing citation tags.
   * @returns The cleaned content without citation tags.
   */
  static stripCitationTags(content) {
    const regex = new RegExp(`${RARE_SYMBOLS.OPENING_TAG}([\\d|\\w|\\s|,]{0,})${RARE_SYMBOLS.CLOSING_TAG}?`, "g");
    return content.replace(regex, "");
  }
  /**
   * Removes citations from a deeply nested plain object and returns a new object with citations removed.
   * @param obj The plain object to process.
   * @returns A tuple containing the new object and an array of extracted citations with paths.
   */
  removeCitationsFromObject(obj) {
    const result = [];
    const processObject = (current, path) => {
      if (typeof current === "string") {
        const extraction = this.extractCitations(current);
        if (extraction.citations.length > 0) {
          result.push(...extraction.citations.map((citation) => ({ path, citation })));
        }
        return extraction.cleaned;
      } else if (typeof current === "object" && current !== null) {
        const newObject = Array.isArray(current) ? [] : {};
        for (const key of Object.keys(current)) {
          newObject[key] = processObject(current[key], `${path}.${key}`);
        }
        return newObject;
      }
      return current;
    };
    const newObj = processObject(obj, "root");
    return [newObj, result];
  }
  /**
   * Re-adds citations to the cleaned content based on their offsets.
   * @param cleaned The cleaned string without citation tags.
   * @param citations The array of citations with offsets.
   * @returns The string with citation tags re-added.
   */
  reAddCitations(cleaned, citations) {
    let content = cleaned;
    citations.sort((a, b) => (a.offset ?? 0) - (b.offset ?? 0));
    const adjustment = 0;
    for (const citation of citations) {
      if (citation.offset != null) {
        const position = citation.offset + adjustment;
        content = content.slice(0, position) + citation.tag + content.slice(position);
      }
    }
    return content;
  }
};

// src/component.default.ts
import { z as z2 } from "@bpinternal/zui";
var Button = new Component({
  type: "leaf",
  description: "A button component that can perform actions when clicked",
  name: "Button",
  aliases: ["btn"],
  examples: [
    {
      name: "Say action",
      description: "A button that triggers a say action",
      code: `yield <Message>
  <Button action="say" label="Hello" />
</Message>`
    },
    {
      name: "Postback action",
      description: "A button that sends a postback value",
      code: `yield <Message>
  Choose an option:
  <Button action="postback" label="Buy" value="buy_product" />
  <Button action="postback" label="Buy" value="buy_product2" />
</Message>`
    }
  ],
  leaf: {
    props: z2.object({
      action: z2.enum(["say", "url", "postback"]).default("say").describe('The action to perform when the button is clicked. Can be "say", "url", or "postback"'),
      label: z2.string().describe("The text displayed on the button (min 1 character, max 250 characters)"),
      value: z2.string().optional().describe('The postback value to send when the button is clicked. Required if action is "postback"'),
      url: z2.string().optional().describe('The URL to open when the button is clicked. Required if action is "url"')
    })
  }
});
var Image = new Component({
  type: "leaf",
  name: "Image",
  description: "Displays an image from a URL.",
  aliases: [],
  examples: [
    {
      name: "Basic image",
      description: "A simple image with alt text",
      code: `yield <Message>
  An example image:
  <Image url="https://example.com/photo.jpg" alt="Example image" />
</Message>`
    }
  ],
  leaf: {
    props: z2.object({
      url: z2.string().describe("The URL of the image (must be valid)"),
      alt: z2.string().optional().describe("Alternative text describing the image")
    })
  }
});
var File = new Component({
  type: "leaf",
  name: "File",
  description: "Sends a downloadable file to the user.",
  aliases: [],
  examples: [
    {
      name: "PDF download",
      description: "Send a PDF file with a name",
      code: `yield <Message>
  Here is your report:
  <File url="https://example.com/report.pdf" name="Report.pdf" />
</Message>`
    }
  ],
  leaf: {
    props: z2.object({
      url: z2.string().describe("The URL of the file (must be valid)"),
      name: z2.string().optional().describe("The display name of the file")
    })
  }
});
var Video = new Component({
  type: "leaf",
  name: "Video",
  description: "Embeds a video from a URL.",
  aliases: [],
  examples: [
    {
      name: "Intro video",
      description: "A video with a title",
      code: `yield <Message>
  Watch this video:
  <Video url="https://example.com/intro.mp4" title="Welcome" />
</Message>`
    }
  ],
  leaf: {
    props: z2.object({
      url: z2.string().describe("The URL of the video (must be valid)"),
      title: z2.string().optional().describe("Title for the video")
    })
  }
});
var Audio = new Component({
  type: "leaf",
  name: "Audio",
  description: "Plays an audio clip from a URL.",
  aliases: [],
  examples: [
    {
      name: "Sample audio",
      description: "Play a short audio clip with a title",
      code: `yield <Message>
  Listen to this audio:
  <Audio url="https://example.com/audio.mp3" title="Sample" />
</Message>`
    }
  ],
  leaf: {
    props: z2.object({
      url: z2.string().describe("The URL of the audio clip (must be valid)"),
      title: z2.string().optional().describe("Title for the audio clip")
    })
  }
});
var Card = new Component({
  type: "container",
  name: "Card",
  description: "A visual card component that can include an image and buttons.",
  aliases: [],
  examples: [
    {
      name: "Product card",
      description: "A card with an image and two buttons",
      code: `yield <Message>
  Featured product:
  <Card title="Product Name" subtitle="Limited offer">
    <Image url="https://example.com/product.jpg" alt="Product image" />
    <Button action="postback" label="Buy" value="buy_product" />
    <Button action="postback" label="Wishlist" value="wishlist" />
  </Card>
</Message>`
    }
  ],
  container: {
    props: z2.object({
      title: z2.string().min(1).max(250).describe("Title text (1\u2013250 characters)"),
      subtitle: z2.string().optional().describe("Optional subtitle for the card")
    }),
    children: [
      {
        description: "Image (optional, max 1)",
        component: Image.definition
      },
      {
        description: "Button (optional, up to 5)",
        component: Button.definition
      }
    ]
  }
});
var Carousel = new Component({
  type: "container",
  name: "Carousel",
  description: "A virtual container for displaying 1 to 10 Card components as a carousel.",
  aliases: [],
  examples: [
    {
      name: "Product carousel",
      description: "A carousel with multiple cards",
      code: `yield <Message>
  Here are some products you might like:
  <Carousel>
    <Card title="Item 1" subtitle="First product">
      <Image url="https://example.com/item1.jpg" alt="Item 1" />
      <Button action="postback" label="Buy" value="buy_1" />
    </Card>
    <Card title="Item 2" subtitle="Second product">
      <Image url="https://example.com/item2.jpg" alt="Item 2" />
      <Button action="postback" label="Buy" value="buy_2" />
    </Card>
  </Carousel>
</Message>`
    }
  ],
  container: {
    props: z2.object({}),
    children: [
      {
        description: "Card component (required, 1\u201310 allowed)",
        component: Card.definition
      }
    ]
  }
});
var Text = new Component({
  type: "default",
  name: "Message",
  aliases: ["Text", "Markdown"],
  description: "Markdown-formatted text that appears directly inside components like <Message>.",
  examples: [
    {
      name: "Basic Markdown",
      description: "Simple markdown content inside a message",
      code: `yield <Message>
  **Hello**, welcome to our service!
</Message>`
    }
  ],
  default: {
    props: z2.object({}),
    children: []
  }
});
var DefaultComponents = {
  Button,
  Image,
  File,
  Video,
  Audio,
  Card,
  Carousel,
  Text
};

// src/chat.ts
var Chat = class {
  handler;
  transcript;
  components;
  /**
   * Creates a new Chat instance.
   *
   * @param props - Chat configuration
   * @param props.handler - Function to handle agent messages (called for every agent output)
   * @param props.components - Available UI components (static array or dynamic function)
   * @param props.transcript - Conversation history (static array or dynamic function, defaults to empty)
   *
   * @example
   * ```typescript
   * // Basic chat with static configuration
   * const chat = new Chat({
   *   handler: async (input) => {
   *     if (isComponent(input, DefaultComponents.Text)) {
   *       console.log('Agent:', input.children.join(''))
   *     }
   *   },
   *   components: [DefaultComponents.Text, DefaultComponents.Button],
   *   transcript: [
   *     { role: 'user', content: 'Hello', timestamp: Date.now() }
   *   ],
   * })
   * ```
   *
   * @example
   * ```typescript
   * // Dynamic chat with functions for real-time updates
   * class MyChat extends Chat {
   *   private messages: Transcript.Message[] = []
   *
   *   constructor() {
   *     super({
   *       handler: (input) => this.handleMessage(input),
   *
   *       // Dynamic components - can change during execution
   *       components: () => [
   *         DefaultComponents.Text,
   *         DefaultComponents.Button,
   *         ...this.getCustomComponents()
   *       ],
   *
   *       // Dynamic transcript - always reflects current state
   *       transcript: () => this.messages,
   *     })
   *   }
   * }
   * ```
   */
  constructor(props) {
    this.handler = props.handler;
    this.components = props.components;
    this.transcript = props.transcript || [];
  }
  /**
   * Called when an execution cycle completes, regardless of the outcome.
   *
   * Override this method to handle execution results, manage conversation flow,
   * or perform cleanup tasks. This is called after each `execute()` call completes,
   * whether it succeeds, fails, or is interrupted.
   *
   * @param result - The execution result containing status, iterations, and exit information
   *
   * @example
   * ```typescript
   * class MyChat extends Chat {
   *   public result?: ExecutionResult
   *
   *   onExecutionDone(result: ExecutionResult) {
   *     // Store result for conversation flow control
   *     this.result = result
   *
   *     // Handle different result types
   *     if (result.isSuccess()) {
   *       console.log('✅ Execution completed successfully')
   *       console.log('Exit:', result.output.exit_name)
   *     } else if (result.isError()) {
   *       console.error('❌ Execution failed:', result.output.error)
   *     } else if (result.isInterrupted()) {
   *       console.log('⏸️ Execution interrupted (partial result)')
   *     }
   *   }
   *
   *   // Use stored result for conversation flow
   *   isWaitingForInput(): boolean {
   *     return this.result?.is(ListenExit) ?? false
   *   }
   * }
   * ```
   *
   * @example
   * ```typescript
   * // CLIChat implementation example
   * class CLIChat extends Chat {
   *   public status?: IterationStatus
   *   public result?: ExecutionResult
   *
   *   onExecutionDone(result: ExecutionResult) {
   *     this.result = result
   *     this.status = result.iterations.at(-1)?.status
   *   }
   *
   *   // Check if agent exited with specific exit type
   *   hasExitedWith<R>(exit: Exit<R>): boolean {
   *     return this.status?.type === 'exit_success' &&
   *            this.status.exit_success.exit_name === exit.name
   *   }
   * }
   * ```
   */
  onExecutionDone(_result) {
  }
};

// src/index.ts
var utils = {
  toValidObjectName,
  toValidFunctionName,
  wrapContent,
  truncateWrappedContent
};
var execute = async (props) => {
  const { executeContext } = await import("./llmz-7OZC2RFA.js");
  return executeContext(props);
};
var init = async () => {
  await import("./llmz-7OZC2RFA.js");
  await import("./component-EZAROCAF.js");
  await import("./tool-ZP6BC3P2.js");
  await import("./exit-56DISPLS.js");
  await import("./jsx-PCXQVMLU.js");
  await import("./vm-GDB4GW4S.js");
  await import("./utils-D3J443FA.js");
  await import("./truncator-2FSONJM2.js");
  await import("./typings-DVMU3KVB.js");
  await import("./dual-modes-ILL7OVJG.js");
};
export {
  Chat,
  CitationsManager,
  Component,
  DefaultComponents,
  DefaultExit,
  ErrorExecutionResult,
  ExecutionResult,
  Exit,
  ListenExit,
  LoopExceededError,
  ObjectInstance,
  PartialExecutionResult,
  Snapshot,
  SnapshotSignal,
  SuccessExecutionResult,
  ThinkExit,
  ThinkSignal,
  Tool,
  assertValidComponent,
  execute,
  getValue,
  init,
  isAnyComponent,
  isComponent,
  parseExit,
  renderToTsx,
  utils,
  version
};
