// @bpinternal/const/dist/billing-versions.js
"use strict";
/**
 * Billing versions for v4 billing system
 * Each version represents a complete billing configuration
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBillingVersion = exports.getActiveBillingVersion = exports.billingDefinitions = exports.billingV4Definition = exports.BILLING_VERSIONS = void 0;
const billing_plans_1 = require("./billing-plans");
const billing_addons_1 = require("./billing-addons");
const billing_meters_1 = require("./billing-meters");
exports.BILLING_VERSIONS = ['v4'];
exports.billingV4Definition = {
    version: 'v4',
    name: 'Billing Version 4',
    description: 'New billing system with credit grants and entitlements',
    plans: billing_plans_1.billingPlansDefs,
    addons: billing_addons_1.billingAddonsDefs,
    meters: billing_meters_1.billingMetersDefs,
    isActive: true,
    createdAt: new Date().toISOString(),
    // stripeMapping will be populated by Pulumi
};
exports.billingDefinitions = {
    v4: exports.billingV4Definition,
};
const getActiveBillingVersion = () => {
    const activeVersion = Object.values(exports.billingDefinitions).find((v) => v.isActive);
    if (!activeVersion) {
        throw new Error('No active billing version found');
    }
    return activeVersion;
};
exports.getActiveBillingVersion = getActiveBillingVersion;
const getBillingVersion = (version) => {
    return exports.billingDefinitions[version];
};
exports.getBillingVersion = getBillingVersion;
