// @bpinternal/const/dist/billing-features.js
"use strict";
/**
 * Billing features for v4 billing system
 * Based on the requirements from the billing refactor document
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.billingFeatureDefs = exports.CREDIT_GRANT_BILLING_FEATURE_IDS = exports.NON_ATOMIC_BILLING_FEATURE_IDS = exports.ATOMIC_BILLING_FEATURE_IDS = exports.BILLING_FEATURE_IDS = void 0;
exports.BILLING_FEATURE_IDS = [
    'human_in_the_loop',
    'ratelimit',
    'incoming_messages_events',
    'integration_spend',
    'integration_subscription',
    'table_rows',
    'bot_count',
    'always_alive_count',
    'always_alive_concurrency',
    'collaborator_count',
    'file_storage',
    'vector_db_storage',
    'saved_versions',
];
exports.ATOMIC_BILLING_FEATURE_IDS = [
    'integration_subscription',
    'bot_count',
    'always_alive_count',
    'always_alive_concurrency',
    'collaborator_count',
    'saved_versions',
];
exports.NON_ATOMIC_BILLING_FEATURE_IDS = [
    'incoming_messages_events',
    'integration_spend',
    'table_rows',
    'file_storage',
    'vector_db_storage',
];
exports.CREDIT_GRANT_BILLING_FEATURE_IDS = ['integration_spend'];
exports.billingFeatureDefs = {
    human_in_the_loop: {
        name: 'Human in the Loop',
        description: 'Ability to enable human intervention in bot conversations',
        type: 'boolean',
    },
    ratelimit: {
        name: 'Rate Limit',
        description: 'API rate limiting configuration',
        type: 'static',
    },
    incoming_messages_events: {
        name: 'Incoming Messages/Events',
        description: 'Number of incoming messages and events processed per month',
        type: 'usage',
        resetPeriod: 'monthly',
        dimensions: {
            botId: 'Bot ID',
            type: 'Message or Event type',
        },
    },
    integration_spend: {
        name: 'Integration Spend',
        description: 'Credits spent on integrations (AI, third-party services)',
        type: 'usage',
        resetPeriod: 'monthly',
        dimensions: {
            botId: 'Bot ID',
            integrationName: 'Integration name',
            integrationVersionId: 'Integration version ID',
            actionName: 'Action name',
        },
    },
    integration_subscription: {
        name: 'Integration Subscription',
        description: 'Monthly subscription to premium integrations',
        type: 'subscription',
        resetPeriod: 'monthly',
    },
    table_rows: {
        name: 'Table Rows',
        description: 'Number of table rows that can be stored',
        type: 'static',
    },
    bot_count: {
        name: 'Bot Count',
        description: 'Number of bots that can be created',
        type: 'static',
    },
    always_alive_count: {
        name: 'Always Alive Count',
        description: 'Number of bots that can be kept always alive',
        type: 'static',
    },
    always_alive_concurrency: {
        name: 'Always Alive Concurrency',
        description: 'Number of concurrent instances for always alive bots',
        type: 'static',
    },
    collaborator_count: {
        name: 'Collaborator Count',
        description: 'Number of collaborators that can be added to workspace',
        type: 'static',
    },
    file_storage: {
        name: 'File Storage',
        description: 'Amount of file storage available',
        type: 'static',
    },
    vector_db_storage: {
        name: 'Vector DB Storage',
        description: 'Amount of vector database storage available',
        type: 'static',
    },
    saved_versions: {
        name: 'Saved Versions',
        description: 'Number of bot versions that can be saved',
        type: 'static',
    },
};
