// @bpinternal/const/dist/billing-plans.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.billingPlansDefs = exports.BILLING_ADDON_IDS = exports.BILLING_PLAN_IDS = exports.ALLOWED_TRIAL_PLANS = exports.ALLOWED_TRIAL_LENGTHS_IN_DAYS = exports.FREE_PRICE = void 0;
/**
 * Special price value indicating a free product/plan.
 * This will be converted to $0 during deployment.
 */
exports.FREE_PRICE = -1;
exports.ALLOWED_TRIAL_LENGTHS_IN_DAYS = 7;
exports.ALLOWED_TRIAL_PLANS = ['plus', 'team'];
// keep this in order of lowest to highest tier
exports.BILLING_PLAN_IDS = ['community', 'plus', 'team', 'enterprise'];
exports.BILLING_ADDON_IDS = [
    'bot_addon',
    'collaborator_addon',
    'always_alive_addon',
    'messages_events_addon',
    'table_rows_addon',
    'file_storage_addon',
    'vector_db_storage_addon',
    'web_search_addon',
    'ai_credits_addon',
    'enterprise_ai_token',
    'enterprise_additional_workspace',
];
/* IMPORTANT: the `name` field is used to match existing products in Stripe DO NOT CHANGE THE NAME OF AN EXISTING PRODUCT */
exports.billingPlansDefs = {
    community: {
        id: 'community',
        name: 'Plan - Pay as you go',
        description: 'Free plan for developers',
        tierIndex: 0,
        prices: [
            {
                price: exports.FREE_PRICE,
                currency: 'usd',
                interval: 'month',
            },
        ],
        features: {
            human_in_the_loop: false,
            ratelimit: 100,
            incoming_messages_events: 500,
            integration_spend: 0,
            integration_subscription: false,
            table_rows: 1000,
            bot_count: 1,
            always_alive_count: 0,
            always_alive_concurrency: 1,
            collaborator_count: 1,
            file_storage: 100_000_000, // 100 MB
            vector_db_storage: 100_000_000, // 100 MB
            saved_versions: 3,
        },
    },
    plus: {
        id: 'plus',
        name: 'Plan - Plus',
        description: 'For growing businesses',
        tierIndex: 1,
        prices: [
            {
                price: 89,
                currency: 'usd',
                interval: 'month',
            },
            {
                price: 948,
                currency: 'usd',
                interval: 'year',
            },
        ],
        features: {
            human_in_the_loop: true,
            ratelimit: 100,
            incoming_messages_events: 5000,
            integration_spend: 5_000_000_000, // $5 in nano-dollars
            integration_subscription: true,
            table_rows: 100000,
            bot_count: 2,
            always_alive_count: 1,
            always_alive_concurrency: 1,
            collaborator_count: 2,
            file_storage: 10_000_000_000, // 10 GB
            vector_db_storage: 1_000_000_000, // 1 GB
            saved_versions: 50,
        },
    },
    team: {
        id: 'team',
        name: 'Plan - Team',
        description: 'For small teams',
        tierIndex: 2,
        prices: [
            {
                price: 495,
                currency: 'usd',
                interval: 'month',
            },
            {
                price: 5346,
                currency: 'usd',
                interval: 'year',
            },
        ],
        features: {
            human_in_the_loop: true,
            ratelimit: 1000,
            incoming_messages_events: 50000,
            integration_spend: 5_000_000_000, // $5 in nano-dollars
            integration_subscription: true,
            table_rows: 100000,
            bot_count: 3,
            always_alive_count: 3,
            always_alive_concurrency: 3,
            collaborator_count: 3,
            file_storage: 10_000_000_000, // 10 GB
            vector_db_storage: 2_000_000_000, // 2 GB
            saved_versions: 100,
        },
    },
    enterprise: {
        id: 'enterprise',
        name: 'Plan - Enterprise',
        description: 'Enterprise license subscription',
        tierIndex: 3,
        prices: [
            {
                price: exports.FREE_PRICE, // Custom pricing
                currency: 'usd',
                interval: 'month',
            },
        ],
        features: {
            human_in_the_loop: true,
            ratelimit: 10000,
            incoming_messages_events: 1000000,
            integration_spend: 5_000_000_000, // $5 in nano-dollars
            integration_subscription: true,
            table_rows: 10000000,
            bot_count: 100,
            always_alive_count: 50,
            always_alive_concurrency: 3,
            collaborator_count: 100,
            file_storage: 1000_000_000_000, // 1000 GB
            vector_db_storage: 500_000_000_000, // 500 GB
            saved_versions: 1000,
        },
    },
};
