// @bpinternal/const/dist/billing-addons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.billingAddonsDefs = void 0;
const billing_plans_1 = require("./billing-plans");
/* IMPORTANT: the `name` field is used to match existing products in Stripe DO NOT CHANGE THE NAME OF AN EXISTING PRODUCT */
exports.billingAddonsDefs = {
    bot_addon: {
        id: 'bot_addon',
        name: 'Add-on - Bot',
        description: 'Add one more bot to your workspace',
        prices: [
            {
                price: 10,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'bot_count',
        increment: 1,
    },
    collaborator_addon: {
        id: 'collaborator_addon',
        name: 'Add-on - Collaborators',
        description: 'Add one more collaborator to your workspace',
        prices: [
            {
                price: 25,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'collaborator_count',
        increment: 1,
    },
    always_alive_addon: {
        id: 'always_alive_addon',
        name: 'Add-on - Always Alive',
        description: 'Keep one more bot always alive',
        prices: [
            {
                price: 10,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'always_alive_count',
        increment: 1,
    },
    messages_events_addon: {
        id: 'messages_events_addon',
        name: 'Add-on - Messages & Events',
        description: 'Adds 5,000 messages and events',
        prices: [
            {
                price: 20,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'incoming_messages_events',
        increment: 5000,
    },
    table_rows_addon: {
        id: 'table_rows_addon',
        name: 'Add-on - Table Rows',
        description: 'Adds 100,000 additional table rows',
        prices: [
            {
                price: 25,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'table_rows',
        increment: 100000,
    },
    file_storage_addon: {
        id: 'file_storage_addon',
        name: 'Add-on - File Storage',
        description: 'Adds 10GB of File Storage',
        prices: [
            {
                price: 10,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'file_storage',
        increment: 10_000_000_000, // 10 GB
    },
    vector_db_storage_addon: {
        id: 'vector_db_storage_addon',
        name: 'Add-on - Vector DB Storage',
        description: 'Add 1 GB of vector database storage',
        prices: [
            {
                price: 20,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'vector_db_storage',
        increment: 1_000_000_000, // 1 GB
    },
    web_search_addon: {
        id: 'web_search_addon',
        name: 'Web Search and Crawl',
        description: 'Web search and crawl functionality',
        prices: [
            {
                price: 10,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'integration_spend',
        increment: 10_000_000_000, // $10 in nano-dollars
    },
    ai_credits_addon: {
        id: 'ai_credits_addon',
        name: 'AI Tokens',
        description: 'Add $25 worth of AI credits',
        prices: [
            {
                price: 25,
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'integration_spend',
        increment: 25_000_000_000, // $25 in nano-dollars
    },
    enterprise_ai_token: {
        id: 'enterprise_ai_token',
        name: 'Enterprise - AI Token',
        description: 'Enterprise AI token credits',
        prices: [
            {
                price: billing_plans_1.FREE_PRICE, // Custom pricing
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'integration_spend',
        increment: 100_000_000_000, // $100 in nano-dollars
        hidden: true,
    },
    enterprise_additional_workspace: {
        id: 'enterprise_additional_workspace',
        name: 'Enterprise - Additional Workspace',
        description: 'Enterprise additional workspace',
        prices: [
            {
                price: billing_plans_1.FREE_PRICE, // Custom pricing
                currency: 'usd',
                interval: 'month',
            },
        ],
        feature: 'collaborator_count',
        increment: 10,
        hidden: true,
    },
};
