// @bpinternal/const/dist/billing-meters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.billingMetersDefs = exports.BILLING_METERS_IDS = void 0;
exports.BILLING_METERS_IDS = ['ai_spend_credit_grant', 'ai_spend_pay_as_you_go'];
exports.billingMetersDefs = {
    ai_spend_credit_grant: {
        id: 'ai_spend_credit_grant',
        name: 'AI Spend - Credit grant',
        description: 'Pre-purchased credits applied toward AI usage',
        prices: [
            {
                unitAmount: 1,
                currency: 'usd',
                interval: 'month',
                aggregation: 'last_during_period',
            },
        ],
        meterType: 'creditGrant',
    },
    ai_spend_pay_as_you_go: {
        id: 'ai_spend_pay_as_you_go',
        name: 'AI Spend - Pay as you go',
        description: 'Usage-based charges for AI spend, billed monthly',
        prices: [
            {
                unitAmount: 1,
                currency: 'usd',
                interval: 'month',
                aggregation: 'last_during_period',
            },
        ],
        meterType: 'payAsYouGo',
    },
};
