// @bpinternal/const/dist/prefixes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.objectToPrefixMap = exports.prefixToObjectMap = void 0;
exports.prefixToObjectMap = {
    accnt: 'account',
    accntpf: 'accountPreference',
    action: 'action',
    activty: 'activity',
    anlytic: 'analytics',
    audit: 'audit',
    bak: 'botApiKey',
    bot: 'bot',
    card: 'card',
    cfg: 'configuration',
    channel: 'channel',
    conv: 'conversation',
    devbot: 'devBot',
    devint: 'devIntegration',
    evt: 'event',
    file: 'file',
    flow: 'flow',
    iak: 'integrationApiKey',
    int: 'integration',
    iface: 'interface',
    ifver: 'interfaceVersion',
    intver: 'integrationVersion',
    iss: 'issue',
    issevt: 'issueEvent',
    kb: 'knowledgeBase',
    limit: 'limit',
    media: 'media',
    msg: 'message',
    node: 'node',
    notif: 'notification',
    pat: 'personalAccessToken',
    plugin: 'plugin',
    plugver: 'pluginVersion',
    quota: 'quota',
    recevt: 'recurringEvent',
    report: 'report',
    sandbox: 'sandbox',
    schema: 'schema',
    script: 'script',
    state: 'state',
    table: 'table',
    tag: 'tag',
    task: 'task',
    usage: 'usage',
    user: 'user',
    webhook: 'webhook',
    wkspace: 'workspace',
    wrkflow: 'workflow',
    wkspacepf: 'workspacePreference',
    trial: 'trial',
};
exports.objectToPrefixMap = {
    account: 'accnt',
    accountPreference: 'accntpf',
    action: 'action',
    activity: 'activty',
    analytics: 'anlytic',
    audit: 'audit',
    bot: 'bot',
    botApiKey: 'bak',
    card: 'card',
    channel: 'channel',
    configuration: 'cfg',
    conversation: 'conv',
    devBot: 'devbot',
    devIntegration: 'devint',
    event: 'evt',
    file: 'file',
    flow: 'flow',
    integration: 'int',
    integrationApiKey: 'iak',
    integrationVersion: 'intver',
    interface: 'iface',
    interfaceVersion: 'ifver',
    issue: 'iss',
    issueEvent: 'issevt',
    knowledgeBase: 'kb',
    limit: 'limit',
    media: 'media',
    message: 'msg',
    node: 'node',
    notification: 'notif',
    personalAccessToken: 'pat',
    plugin: 'plugin',
    pluginVersion: 'plugver',
    quota: 'quota',
    recurringEvent: 'recevt',
    report: 'report',
    sandbox: 'sandbox',
    schema: 'schema',
    script: 'script',
    state: 'state',
    table: 'table',
    tag: 'tag',
    task: 'task',
    usage: 'usage',
    user: 'user',
    webhook: 'webhook',
    workflow: 'wrkflow',
    workspace: 'wkspace',
    workspacePreference: 'wkspacepf',
    trial: 'trial',
};
