// @bpinternal/const/dist/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.billingConfigSchema = exports.billingVersionConfigSchema = exports.meterSchema = exports.addOnSchema = exports.planSchema = exports.featureSchema = exports.meteredPriceSchema = exports.priceSchema = exports.meterTypeSchema = exports.aggregationSchema = exports.unitAmountSchema = exports.intervalSchema = void 0;
const zod_1 = require("zod");
const billing_versions_1 = require("./billing-versions");
const billing_features_1 = require("./billing-features");
exports.intervalSchema = zod_1.z.enum(['month', 'year']);
exports.unitAmountSchema = zod_1.z.number().int().min(1);
exports.aggregationSchema = zod_1.z.enum(['sum', 'last_during_period', 'last_ever', 'max']);
exports.meterTypeSchema = zod_1.z.enum(['payAsYouGo', 'creditGrant']);
exports.priceSchema = zod_1.z.object({
    providerPriceId: zod_1.z.string(),
    amount: zod_1.z.number(),
    currency: zod_1.z.enum(['usd']),
    interval: exports.intervalSchema,
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
});
exports.meteredPriceSchema = zod_1.z.object({
    providerPriceId: zod_1.z.string(),
    unitAmount: exports.unitAmountSchema,
    currency: zod_1.z.enum(['usd']),
    interval: exports.intervalSchema,
    aggregation: exports.aggregationSchema,
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
});
exports.featureSchema = zod_1.z.enum(billing_features_1.BILLING_FEATURE_IDS);
exports.planSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    tierIndex: zod_1.z.number(),
    providerProductId: zod_1.z.string(),
    prices: zod_1.z.record(exports.intervalSchema, exports.priceSchema),
    features: zod_1.z.record(exports.featureSchema, zod_1.z.any()),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
});
exports.addOnSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    providerProductId: zod_1.z.string(),
    prices: zod_1.z.record(exports.intervalSchema, exports.priceSchema),
    feature: exports.featureSchema,
    increment: zod_1.z.number(),
    prorationDisabled: zod_1.z.boolean().optional(),
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
    autoRechargeable: zod_1.z.boolean().optional(),
    hidden: zod_1.z.boolean().optional(),
});
exports.meterSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    providerProductId: zod_1.z.string(),
    prices: zod_1.z.record(exports.intervalSchema, exports.meteredPriceSchema),
    meterType: exports.meterTypeSchema,
    metadata: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
});
exports.billingVersionConfigSchema = zod_1.z.object({
    version: zod_1.z.enum(billing_versions_1.BILLING_VERSIONS),
    name: zod_1.z.string(),
    description: zod_1.z.string(),
    plans: zod_1.z.record(zod_1.z.string(), exports.planSchema),
    addons: zod_1.z.record(zod_1.z.string(), exports.addOnSchema),
    meters: zod_1.z.record(zod_1.z.string(), exports.meterSchema),
    isActive: zod_1.z.boolean(),
    createdAt: zod_1.z.string(),
});
exports.billingConfigSchema = zod_1.z.record(zod_1.z.enum(billing_versions_1.BILLING_VERSIONS), exports.billingVersionConfigSchema);
