// @bpinternal/const/dist/usage-metadata.js
"use strict";
/**
 * Usage metadata definitions for v4 billing system
 * These define the dimensions we want to track for each feature
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateUsageEvent = exports.usageMetadataConfigs = void 0;
exports.usageMetadataConfigs = {
    incoming_messages_events: {
        feature: 'incoming_messages_events',
        dimensions: {
            botId: {
                description: 'The ID of the bot that processed the message/event',
                required: true,
                type: 'string',
            },
            type: {
                description: 'Whether this is a message or an event',
                required: true,
                type: 'string',
            },
            channel: {
                description: 'The channel where the message/event originated',
                required: false,
                type: 'string',
            },
        },
    },
    integration_spend: {
        feature: 'integration_spend',
        dimensions: {
            botId: {
                description: 'The ID of the bot that used the integration',
                required: true,
                type: 'string',
            },
            integrationVersionId: {
                description: 'The version ID of the integration used',
                required: true,
                type: 'string',
            },
            actionName: {
                description: 'The specific action that was called',
                required: true,
                type: 'string',
            },
            provider: {
                description: 'The provider of the integration (e.g., openai, anthropic)',
                required: false,
                type: 'string',
            },
            model: {
                description: 'The specific model used (for AI integrations)',
                required: false,
                type: 'string',
            },
        },
    },
    table_rows: {
        feature: 'table_rows',
        dimensions: {
            botId: {
                description: 'The ID of the bot that owns the table',
                required: true,
                type: 'string',
            },
            tableId: {
                description: 'The ID of the table',
                required: true,
                type: 'string',
            },
            operation: {
                description: 'The operation performed (create, update, delete)',
                required: true,
                type: 'string',
            },
        },
    },
    file_storage: {
        feature: 'file_storage',
        dimensions: {
            botId: {
                description: 'The ID of the bot that owns the file',
                required: true,
                type: 'string',
            },
            fileId: {
                description: 'The ID of the file',
                required: true,
                type: 'string',
            },
            operation: {
                description: 'The operation performed (upload, delete)',
                required: true,
                type: 'string',
            },
            fileType: {
                description: 'The type/extension of the file',
                required: false,
                type: 'string',
            },
        },
    },
    vector_db_storage: {
        feature: 'vector_db_storage',
        dimensions: {
            botId: {
                description: 'The ID of the bot that owns the vectors',
                required: true,
                type: 'string',
            },
            collectionId: {
                description: 'The ID of the vector collection',
                required: true,
                type: 'string',
            },
            operation: {
                description: 'The operation performed (insert, delete)',
                required: true,
                type: 'string',
            },
        },
    },
};
const validateUsageEvent = (event) => {
    const config = exports.usageMetadataConfigs[event.feature];
    if (!config) {
        return false;
    }
    // Check required dimensions
    for (const [dimName, dimConfig] of Object.entries(config.dimensions)) {
        if (dimConfig.required && !(dimName in event.meta)) {
            return false;
        }
    }
    return true;
};
exports.validateUsageEvent = validateUsageEvent;
