// @a24z/mermaid-parser/dist/index.js
// src/dom-stubs.ts
if (typeof globalThis !== "undefined") {
  if (typeof window === "undefined") {
    globalThis.window = {
      scrollX: 0,
      scrollY: 0,
      location: { href: "" },
      navigator: { userAgent: "mermaid-validator" },
      addEventListener: () => {},
      removeEventListener: () => {},
      getComputedStyle: () => ({
        getPropertyValue: () => ""
      })
    };
  }
  if (typeof document === "undefined") {
    globalThis.document = {
      querySelector: () => null,
      querySelectorAll: () => [],
      getElementById: () => null,
      getElementsByClassName: () => [],
      getElementsByTagName: () => [],
      createElement: () => ({
        tagName: "div",
        style: {},
        setAttribute: () => {},
        getAttribute: () => null,
        removeAttribute: () => {},
        addEventListener: () => {},
        removeEventListener: () => {},
        appendChild: () => {},
        removeChild: () => {},
        getBoundingClientRect: () => ({
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          width: 100,
          height: 20,
          x: 0,
          y: 0
        }),
        innerHTML: "",
        textContent: "",
        childNodes: [],
        classList: {
          add: () => {},
          remove: () => {},
          contains: () => false
        }
      }),
      createElementNS: () => ({
        setAttribute: () => {},
        getAttribute: () => null,
        getBBox: () => ({ x: 0, y: 0, width: 100, height: 20 })
      }),
      createTextNode: () => ({}),
      body: { appendChild: () => {}, style: {} }
    };
  }
  if (typeof Element === "undefined") {
    globalThis.Element = class Element2 {
      setAttribute() {}
      getAttribute() {
        return null;
      }
      removeAttribute() {}
    };
  }
}

// src/index.ts
import * as yaml from "js-yaml";
var DIAGRAM_PATTERNS = {
  flowchart: /^\s*(graph|flowchart)(\s+TD|TB|BT|RL|LR)?/i,
  sequence: /^\s*sequenceDiagram/i,
  class: /^\s*classDiagram/i,
  state: /^\s*stateDiagram(-v2)?/i,
  er: /^\s*erDiagram/i,
  journey: /^\s*journey/i,
  gantt: /^\s*gantt/i,
  pie: /^\s*pie(\s+title)?/i,
  gitgraph: /^\s*gitGraph/i,
  mindmap: /^\s*mindmap/i,
  timeline: /^\s*timeline/i,
  quadrant: /^\s*quadrantChart/i,
  requirement: /^\s*requirementDiagram/i,
  c4context: /^\s*C4Context/i,
  c4container: /^\s*C4Container/i,
  c4component: /^\s*C4Component/i,
  c4dynamic: /^\s*C4Dynamic/i,
  c4deployment: /^\s*C4Deployment/i,
  sankey: /^\s*sankey-beta/i,
  block: /^\s*block-beta/i,
  packet: /^\s*packet-beta/i,
  architecture: /^\s*architecture-beta/i,
  xyChart: /^\s*xychart-beta/i
};
function preprocessDiagram(text) {
  const lines = text.split(`
`);
  let config = {};
  let codeLines = [];
  let inDirective = false;
  let directiveText = "";
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.startsWith("%%{")) {
      inDirective = true;
      directiveText = trimmed.slice(3);
      if (trimmed.endsWith("}%%")) {
        directiveText = directiveText.slice(0, -3);
        try {
          const parsed = JSON.parse(directiveText);
          config = { ...config, ...parsed };
        } catch {
          try {
            const yamlParsed = yaml.load(directiveText);
            config = { ...config, ...yamlParsed };
          } catch {}
        }
        inDirective = false;
        directiveText = "";
      }
      continue;
    }
    if (inDirective) {
      directiveText += line;
      if (trimmed.endsWith("}%%")) {
        directiveText = directiveText.slice(0, -3);
        try {
          const parsed = JSON.parse(directiveText);
          config = { ...config, ...parsed };
        } catch {}
        inDirective = false;
        directiveText = "";
      }
      continue;
    }
    if (trimmed.startsWith("%%") || trimmed.startsWith("#")) {
      continue;
    }
    codeLines.push(line);
  }
  return {
    code: codeLines.join(`
`),
    config: Object.keys(config).length > 0 ? config : undefined
  };
}
function detectType(text) {
  const cleaned = text.trim();
  for (const [type, pattern] of Object.entries(DIAGRAM_PATTERNS)) {
    if (pattern.test(cleaned)) {
      return type;
    }
  }
  return "unknown";
}
var SYNTAX_VALIDATORS = {
  flowchart: (text) => {
    const hasNodes = /\w+\s*(\[|\(|\{)/.test(text);
    const hasArrows = /\w+\s*-+>+\s*\w+/.test(text);
    const hasValidSyntax = !/[<>]{3,}/.test(text);
    return (hasNodes || hasArrows) && hasValidSyntax;
  },
  sequence: (text) => {
    const hasParticipants = /participant\s+\w+|actor\s+\w+|\w+\s*-[->]+\s*\w+/.test(text);
    return hasParticipants;
  },
  class: (text) => {
    const hasClasses = /class\s+\w+|[\w\s]+\s*[<|>-]+\s*[\w\s]+/.test(text);
    return hasClasses;
  },
  state: (text) => {
    const hasStates = /\[\*\]|\w+\s*-->\s*\w+|state\s+\w+/.test(text);
    return hasStates;
  },
  er: (text) => {
    const hasEntities = /\w+\s+\|\|?--[o{]?\{?\s*\w+/.test(text);
    return hasEntities;
  },
  gantt: (text) => {
    const hasGanttElements = /section\s+|title\s+|\w+\s*:\s*[\w,\s-]+/.test(text);
    return hasGanttElements;
  },
  pie: (text) => {
    const hasData = /"[^"]+"\s*:\s*\d+/.test(text);
    return hasData;
  },
  journey: (text) => {
    const hasJourneyElements = /section\s+|\w+:\s*\d+/.test(text);
    return hasJourneyElements;
  },
  default: () => true
};
async function validate(text, parseOptions) {
  try {
    if (!text || typeof text !== "string") {
      throw new Error("Invalid input: text must be a non-empty string");
    }
    const processed = preprocessDiagram(text);
    const type = detectType(processed.code);
    if (type === "unknown") {
      throw new Error("Unknown diagram type");
    }
    const validator = SYNTAX_VALIDATORS[type] || SYNTAX_VALIDATORS.default;
    const isValid = validator(processed.code);
    if (!isValid) {
      throw new Error(`Invalid ${type} diagram syntax`);
    }
    return {
      diagramType: type,
      config: processed.config
    };
  } catch (error) {
    if (parseOptions?.suppressErrors) {
      return false;
    }
    throw error;
  }
}
async function parse(text) {
  try {
    const result = await validate(text, { suppressErrors: false });
    if (result) {
      return {
        type: result.diagramType,
        config: result.config,
        valid: true
      };
    }
    return {
      type: "unknown",
      valid: false
    };
  } catch (error) {
    return {
      type: "unknown",
      valid: false,
      error: error instanceof Error ? error.message : String(error)
    };
  }
}
function isSupported(diagramType) {
  return Object.keys(DIAGRAM_PATTERNS).includes(diagramType);
}
function getSupportedDiagrams() {
  return Object.keys(DIAGRAM_PATTERNS);
}
function getDiagramType(text) {
  const processed = preprocessDiagram(text);
  return detectType(processed.code);
}
var mermaidValidator = {
  validate,
  parse,
  isSupported,
  getSupportedDiagrams,
  getDiagramType
};
var src_default = mermaidValidator;
export {
  validate,
  parse,
  isSupported,
  getSupportedDiagrams,
  getDiagramType,
  src_default as default
};

//# debugId=362C6C5B5AD52A4664756E2164756E21
