// @botpress/zai/dist/micropatch.js
export class Micropatch {
  _text;
  _eol;
  /**
   * Create a Micropatch instance.
   * @param source The file contents.
   * @param eol Line ending style. If omitted, it is auto-detected from `source` (CRLF if any CRLF is present; otherwise LF).
   */
  constructor(source, eol) {
    this._text = source;
    this._eol = eol ?? Micropatch.detectEOL(source);
  }
  /** Get current text. */
  getText() {
    return this._text;
  }
  /**
   * Replace current text.
   * Useful if you want to "load" a new snapshot without reconstructing the class.
   */
  setText(source, eol) {
    this._text = source;
    this._eol = eol ?? Micropatch.detectEOL(source);
  }
  /**
   * Apply ops text to current buffer.
   * @param opsText One or more operations in the v0.3 syntax.
   * @returns The updated text (also stored internally).
   * @throws If the patch contains invalid syntax (e.g., range on insert).
   */
  apply(opsText) {
    const ops = Micropatch.parseOps(opsText);
    this._text = Micropatch._applyOps(this._text, ops, this._eol);
    return this._text;
  }
  /**
   * Render a numbered view of the current buffer (token-cheap preview for models).
   * Format: `NNN|<line>`, starting at 001.
   */
  renderNumberedView() {
    const NL = this._eol === "lf" ? "\n" : "\r\n";
    const lines = Micropatch._splitEOL(this._text);
    return lines.map((l, i) => `${String(i + 1).padStart(3, "0")}|${l}`).join(NL);
  }
  // ---------------------- Static helpers ----------------------
  /** Detect EOL style from content. */
  static detectEOL(source) {
    return /\r\n/.test(source) ? "crlf" : "lf";
  }
  /** Split text into lines, preserving empty last line if present. */
  static _splitEOL(text) {
    const parts = text.split(/\r?\n/);
    return parts;
  }
  /** Join lines with the chosen EOL. */
  static _joinEOL(lines, eol) {
    const NL = eol === "lf" ? "\n" : "\r\n";
    return lines.join(NL);
  }
  /** Unescape payload text: `\◼︎` → `◼︎`. */
  static _unescapeMarker(s) {
    return s.replace(/\\◼︎/g, "\u25FC\uFE0E");
  }
  /**
   * Parse ops text (v0.3).
   * - Ignores blank lines and lines not starting with `◼︎` (you can keep comments elsewhere).
   * - Validates ranges for allowed ops.
   */
  static parseOps(opsText) {
    const lines = opsText.split(/\r?\n/);
    const ops = [];
    const headerRe = /^◼︎([<>=-])(\d+)(?:-(\d+))?(?:\|(.*))?$/;
    let i = 0;
    while (i < lines.length) {
      const line = lines[i];
      if (!line) {
        i++;
        continue;
      }
      if (!line.startsWith("\u25FC\uFE0E")) {
        i++;
        continue;
      }
      const m = headerRe.exec(line);
      if (!m || !m[1] || !m[2]) {
        throw new Error(`Invalid op syntax at line ${i + 1}: ${line}`);
      }
      const op = m[1];
      const aNum = parseInt(m[2], 10);
      const bNum = m[3] ? parseInt(m[3], 10) : void 0;
      const firstPayload = m[4] ?? "";
      if (aNum < 1 || bNum !== void 0 && bNum < aNum) {
        throw new Error(`Invalid line/range at line ${i + 1}: ${line}`);
      }
      if (op === "<" || op === ">") {
        if (bNum !== void 0) {
          throw new Error(`Insert cannot target a range (line ${i + 1})`);
        }
        const text = Micropatch._unescapeMarker(firstPayload);
        ops.push({ k: op, n: aNum, s: text });
        i++;
        continue;
      }
      if (op === "-") {
        if (firstPayload !== "") {
          throw new Error(`Delete must not have a payload (line ${i + 1})`);
        }
        if (bNum === void 0) {
          ops.push({ k: "-", n: aNum });
        } else {
          ops.push({ k: "--", a: aNum, b: bNum });
        }
        i++;
        continue;
      }
      if (op === "=") {
        const payload = [Micropatch._unescapeMarker(firstPayload)];
        let j = i + 1;
        while (j < lines.length) {
          const nextLine = lines[j];
          if (!nextLine || nextLine.startsWith("\u25FC\uFE0E")) break;
          payload.push(Micropatch._unescapeMarker(nextLine));
          j++;
        }
        if (bNum === void 0) {
          ops.push({ k: "=", n: aNum, s: payload });
        } else {
          ops.push({ k: "=-", a: aNum, b: bNum, s: payload });
        }
        i = j;
        continue;
      }
      i++;
    }
    return Micropatch._canonicalizeOrder(ops);
  }
  /** Order ops deterministically according to the spec. */
  static _canonicalizeOrder(ops) {
    const delS = [];
    const delR = [];
    const eqS = [];
    const eqR = [];
    const insB = [];
    const insA = [];
    for (const o of ops) {
      switch (o.k) {
        case "-":
          delS.push(o);
          break;
        case "--":
          delR.push(o);
          break;
        case "=":
          eqS.push(o);
          break;
        case "=-":
          eqR.push(o);
          break;
        case "<":
          insB.push(o);
          break;
        case ">":
          insA.push(o);
          break;
        default:
          break;
      }
    }
    delS.sort((a, b) => b.n - a.n);
    delR.sort((a, b) => b.a - a.a);
    eqS.sort((a, b) => a.n - b.n);
    eqR.sort((a, b) => a.a - b.a);
    insB.sort((a, b) => a.n - b.n);
    insA.sort((a, b) => a.n - b.n);
    return [].concat(delS, delR, eqS, eqR, insB, insA);
  }
  /**
   * Apply normalized ops to given source.
   * - Uses a live index map from ORIGINAL 1-based addresses → current positions.
   * - Skips ops whose targets can no longer be mapped (idempotency-friendly).
   */
  static _applyOps(source, ops, eol) {
    const lines = Micropatch._splitEOL(source);
    const idx = Array.from({ length: lines.length }, (_, i) => i);
    const map = (n) => idx[n - 1] ?? -1;
    const bump = (from, delta) => {
      for (let i = 0; i < idx.length; i++) {
        const current = idx[i];
        if (current !== void 0 && current >= from) {
          idx[i] = current + delta;
        }
      }
    };
    for (const o of ops) {
      switch (o.k) {
        case "-": {
          const i = map(o.n);
          if (i >= 0 && i < lines.length) {
            lines.splice(i, 1);
            bump(i, -1);
          }
          break;
        }
        case "--": {
          const a = map(o.a);
          const b = map(o.b);
          if (a >= 0 && b >= a && b < lines.length) {
            lines.splice(a, b - a + 1);
            bump(a, -(b - a + 1));
          }
          break;
        }
        case "=": {
          const i = map(o.n);
          if (i >= 0 && i < lines.length) {
            const rep = o.s;
            lines.splice(i, 1, ...rep);
            bump(i + 1, rep.length - 1);
          }
          break;
        }
        case "=-": {
          const a = map(o.a);
          const b = map(o.b);
          if (a >= 0 && b >= a && b < lines.length) {
            const rep = o.s;
            lines.splice(a, b - a + 1, ...rep);
            bump(a + 1, rep.length - (b - a + 1));
          }
          break;
        }
        case "<": {
          const i = Math.max(0, Math.min(map(o.n), lines.length));
          if (i >= 0) {
            lines.splice(i, 0, o.s);
            bump(i, 1);
          }
          break;
        }
        case ">": {
          const i = Math.max(0, Math.min(map(o.n) + 1, lines.length));
          if (i >= 0) {
            lines.splice(i, 0, o.s);
            bump(i, 1);
          }
          break;
        }
        default:
          break;
      }
    }
    return Micropatch._joinEOL(lines, eol);
  }
  // ---------------------- Convenience APIs ----------------------
  /**
   * Convenience: one-shot apply.
   * @param source Text to patch.
   * @param opsText Operations text.
   * @param eol EOL style (auto-detected if omitted).
   */
  static applyText(source, opsText, eol) {
    const inst = new Micropatch(source, eol);
    return inst.apply(opsText);
  }
  /**
   * Convenience: parse only.
   * Useful for validation without applying.
   */
  static validate(opsText) {
    const ops = Micropatch.parseOps(opsText);
    return { ok: true, count: ops.length };
  }
}
